{
 "metadata": {
  "name": "2.chlamy_Nitro_data_analysis"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pybedtools\n",
      "import math\n",
      "import sys\n",
      "import multiprocessing\n",
      "import cStringIO\n",
      "import tempfile\n",
      "import matplotlib.pyplot as plt\n",
      "import numpy as np\n",
      "import pandas\n",
      "import itertools\n",
      "import collections\n",
      "import re\n",
      "\n",
      "\n",
      "#in-house libs\n",
      "sys.path.append('/global/homes/a/apratap/dev/eclipse_workspace/python_scripts/lib/');\n",
      "import utils"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#function to create chlamy genes phytozome function df\n",
      "def get_chlamy_v5_3_phytozome_function_df():\n",
      "    \n",
      "    phytozome_biomart_query_dump = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/phytozome_function/martquery_1212100847_412.txt'\n",
      "    df = pandas.io.parsers.read_csv(phytozome_biomart_query_dump,\n",
      "                                    sep='\\t',\n",
      "                                    skiprows = 0,\n",
      "                                    names = [ 'gene_name','transcript_name','PFAM_ID','PFAM_description','Panther_ID',\n",
      "                                              'Panther_description','KOG_desctiption','KOG_ID','EC_ID','EC_description','KO_description',\n",
      "                                              'KO_ID','GO_description','GO_ID','Synonyms']\n",
      "                                    )\n",
      "    \n",
      "    \n",
      "    return df\n",
      "\n",
      "\n",
      "\n",
      "#get the raw data in pandas df \n",
      "df = get_chlamy_v5_3_phytozome_function_df()\n",
      "\n",
      "#group by gene name\n",
      "grouped = df.groupby('gene_name')\n",
      "\n",
      "#creating one functional annotation line per Chlamy v5.3 gene\n",
      "# applying aggregate to each group, converting each col to string taking a uniq, converting that to list and converting list to a string\n",
      "gene_function_df = grouped.agg(lambda x: ','.join( list( x.map(str).unique()) ) )"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 193
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_function_df.to_csv('1.csv',sep='\\t')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 194
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#function to get summary of annotation for chlamyv5.3 genes in phytozome\n",
      "def get_chlamy_v5_3_gene_annotation_summary(gene_annot_df):\n",
      "    total_genes = df.gene_name.unique()\n",
      "    total_genes_with_GO_ID = df.gene_name[pandas.notnull(df.GO_ID)].unique()\n",
      "    total_genes_with_PFAM_ID = df.gene_name[pandas.notnull(df.PFAM_ID)].unique()\n",
      "    total_genes_with_Panther_ID = df.gene_name[pandas.notnull(df.Panther_ID)].unique()\n",
      "    total_genes_with_KOG_ID = df.gene_name[pandas.notnull(df.KOG_ID)].unique()\n",
      "    total_genes_with_EC_ID = df.gene_name[pandas.notnull(df.EC_ID)].unique()\n",
      "    \n",
      "           \n",
      "    return {'#genes':total_genes,\n",
      "            '#GO_ID' : total_genes_with_GO_ID,\n",
      "            '#PFAM_ID' : total_genes_with_PFAM_ID,\n",
      "            '#Panther_ID':total_genes_with_Panther_ID,\n",
      "            '#KOG_ID':total_genes_with_KOG_ID,\n",
      "            '#EC_ID' : total_genes_with_EC_ID}\n",
      "\n",
      "annot_sumry_dict = get_chlamy_v5_3_gene_annotation_summary(df)\n",
      "\n",
      "[ (len(genes),catg) for catg,genes in annot_sumry_dict.items() ]\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df[['gene_name','PFAM_ID','Panther_ID','KOG_ID', 'EC_ID','KO_ID','GO_ID']].apply("
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df.gene_name[pandas.isnull(df.GO_ID)].unique()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#the following code reads the Chlamy v5.3 ref gene models\n",
      "# and creates a dict mapping the chalmy common gene names to standard genenames\n",
      "# this is needed to get the standard gene name from the cufflinks produced merged.gtf file which for some genes include the commonn gene name\n",
      "\n",
      "def get_chlamy_common_genesName_to_standard_geneName_dict():\n",
      "    chlamy_common_geneName_to_std_geneName = {}\n",
      "    chlamy_v5_3_ref = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/Creinhardtii_v5.3_223_gene.gff3'\n",
      "    for count,feature in enumerate(pybedtools.BedTool(chlamy_v5_3_ref)):\n",
      "        if count == 100:\n",
      "            pass\n",
      "        #print feature[2]\n",
      "        if feature[2] == 'mRNA':\n",
      "            common_geneName = feature.attrs.get('geneName','NA')\n",
      "            standard_transcriptName = feature.attrs.get('Name','NA')\n",
      "            #removing the transcript prefix\n",
      "            m = re.search('(^.+?).t.*$',standard_transcriptName)\n",
      "            if m:\n",
      "                standard_geneName = m.group(1)\n",
      "            #chlamy_common_geneName_to_std_geneNa\n",
      "            chlamy_common_geneName_to_std_geneName[common_geneName] = standard_geneName\n",
      "    \n",
      "    return chlamy_common_geneName_to_std_geneName\n",
      "            \n",
      "        \n",
      "    \n",
      "chlamy_common_geneName_to_std_geneName = get_chlamy_common_genesName_to_standard_geneName_dict()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chlamy_common_geneName_to_std_geneName.get('AMX2','NA')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#a function to pass only features of a particular featuretype\n",
      "#this is similar to grep\n",
      "# if inverse_flag is set to True, this is similar to grep -v\n",
      "def featuretype_filter(feature,featuretypes,inverse_flag=False):\n",
      "    for featuretype in featuretypes:\n",
      "        if feature[2] == featuretype:\n",
      "            #match found\n",
      "            if inverse_flag is True:\n",
      "                return False\n",
      "            else:\n",
      "                return True\n",
      "            \n",
      "    #no match found\n",
      "    if inverse_flag is True:\n",
      "        return True\n",
      "    else:\n",
      "        return False \n",
      "\n",
      "\n",
      "#function to subset a feature file selecting the features \n",
      "def subset_featuretypes(bedToolsFile,featuretypes,inverse_flag=False):\n",
      "    result = bedToolsFile.filter(featuretype_filter,featuretypes,inverse_flag=inverse_flag).saveas()\n",
      "    return pybedtools.BedTool(result.fn)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def create_pandas_df_from_gtf_file(merged_gtf_file):\n",
      "    '''\n",
      "    given a merged.gtf file produced by cuffmerge this method can\n",
      "    summarize it by gene or transcript level\n",
      "    **PS: assumption is that for each gene the exons are reported, sorted by position**\n",
      "    variable : summarize_by can take the value of gene,transcript\n",
      "    '''\n",
      "    \n",
      "    features = []\n",
      "    for count,line in enumerate(pybedtools.BedTool(merged_gtf_file)):\n",
      "        #parse following from the attrs\n",
      "        attributes = ['gene_id','transcript_id','exon_number','gene_name','class_code','tss_id','p_id']\n",
      "        attr_value = [ line.attrs.get(attr,'NA').strip() for attr in attributes ]\n",
      "        feature = [ x.strip() for x in line[0:8]] + attr_value\n",
      "        features.append(feature)\n",
      "    \n",
      "    #creating pandas df\n",
      "    colnames = ['chr','source','feature','start','stop','score','strand','frame'] + attributes\n",
      "    out_pandas_df = pandas.DataFrame.from_records(features,columns=colnames)\n",
      "    \n",
      "    return out_pandas_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def create_gtf_entry_from_df(group):\n",
      "    output = []\n",
      "    for index,row in group.T.iteritems():\n",
      "        gtf_line = [row['chr'], row['source'], row['feature'], \n",
      "                    str(row['start']), str(row['stop']), row['score'],\n",
      "                    row['strand'], str(row['frame'])]\n",
      "        \n",
      "        gtf_line_attrs = ['gene_id','transcript_id','exon_number','gene_name','tss_id','p_id']\n",
      "        attr_line = ''\n",
      "        \n",
      "        for attr in gtf_line_attrs:\n",
      "            value=str(row.get(attr,'NA'))\n",
      "            if value.strip() != 'NA':\n",
      "                attr_line += ('%s \"%s\"; ' % (attr,value))\n",
      "        \n",
      "        gtf_line.append(attr_line)\n",
      "        output.append(gtf_line)\n",
      "        \n",
      "    return (output)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "test_file='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/assembly/assembly_CWong_modified_cufflinks/test.gtf'\n",
      "\n",
      "rna_seq_assembly_file = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/assembly/assembly_CWong_modified_cufflinks/merged.gtf'\n",
      "\n",
      "#full data\n",
      "df=create_pandas_df_from_gtf_file(rna_seq_assembly_file)\n",
      "\n",
      "#test file\n",
      "#df=create_pandas_df_from_gtf_file(test_file)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#full data\n",
      "#df=create_pandas_df_from_gtf_file(rna_seq_assembly_file)\n",
      "\n",
      "#convert the start and stop it integers\n",
      "df['start']=df.start.map(lambda x: int(x))\n",
      "df['stop']=df.stop.map(lambda x: int(x))\n",
      "\n",
      "\n",
      "#filter unwanted chr\n",
      "# find all chr where tobe filtered chr is True and then take  NOT of it to get rid of them\n",
      "cleaned_df=df[df.chr.isin(['chr_C','chr_M']).map(lambda x: not x)]\n",
      "\n",
      "\n",
      "\n",
      "#####\n",
      "#grouping the data frame for summary\n",
      "####\n",
      "\n",
      "#by transcripts\n",
      "groupby=['gene_id','transcript_id','chr','feature','strand','score','frame','source']\n",
      "grouped_by_transcripts = cleaned_df.groupby(groupby,sort=False)\n",
      "\n",
      "#by genes\n",
      "groupby=['gene_id','chr','feature','strand','score','frame','source']\n",
      "grouped_by_genes = cleaned_df.groupby(groupby,sort=False)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "#########\n",
      "###Creating transcripts from groups\n",
      "#########\n",
      "#Note: similarly dataframe could be summarized by on any number of features\n",
      "\n",
      "#getting the start and end coordinates of the grouped by feature\n",
      "transcripts = pandas.DataFrame(grouped_by_transcripts.agg({'start': np.min,\n",
      "                                                           'stop' : np.max,\n",
      "                                                           'gene_name' : lambda x: ','.join(list(x.unique())),\n",
      "                                                           'exon_number' : np.size\n",
      "                                                          })).reset_index()\n",
      "\n",
      "\n",
      "#update the feature name\n",
      "transcripts['feature'] = 'transcript'\n",
      "\n",
      "#getting transcripts with exon count > 1\n",
      "transcripts_exon_gt_1 = transcripts[transcripts.exon_number > 1]\n",
      "gtf_lines_transcripts = create_gtf_entry_from_df(transcripts_exon_gt_1)\n",
      "\n",
      "\n",
      "\n",
      "#########\n",
      "###Creating genes from groups\n",
      "#########\n",
      "#Note: similarly dataframe could be summarized by on any number of features\n",
      "\n",
      "#getting the start and end coordinates of the grouped by feature\n",
      "genes = pandas.DataFrame(grouped_by_genes.agg({'start': np.min,\n",
      "                                               'stop' : np.max,\n",
      "                                               'gene_name' : lambda x: ','.join(list(x.unique()))\n",
      "                                              })).reset_index()\n",
      "\n",
      "#update the feature name\n",
      "genes['feature'] = 'gene'\n",
      "gtf_lines_genes = create_gtf_entry_from_df(genes)\n",
      "\n",
      "\n",
      "#create a generator for generating exon gtf lines for grouped transcrpt\n",
      "exons_grp_per_transcript =( create_gtf_entry_from_df(group) for name,group in grouped_by_transcripts )\n",
      "\n",
      "\n",
      "#create a generator for generating exon gtf lines for grouped transcrpts, for groups where #group(#exon) > 1\n",
      "exons_grp_per_transcript_exoncount_gt_1 =( create_gtf_entry_from_df(group) for name,group in grouped_by_transcripts if len(group) > 1 )\n",
      "\n",
      " \n",
      "#a = [[3],[4]]\n",
      "#list(itertools.chain.from_iterable(a))\n",
      "#[3,4]\n",
      "#using the concept from the above example\n",
      "gtf_lines_all_exons = list(itertools.chain.from_iterable(exons_grp_per_transcript))\n",
      "gtf_lines_all_exons_filtered = list(itertools.chain.from_iterable(exons_grp_per_transcript_exoncount_gt_1))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes.gene_name[1:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes['check'] = genes.gene_name.apply(lambda x: [ chlamy_common_geneName_to_std_geneName.get(y,y) for y in x.split(',') ] )"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes.check"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rna_seq_genes_fh = tempfile.NamedTemporaryFile(delete=False)\n",
      "rna_seq_trancripts_fh = tempfile.NamedTemporaryFile(delete=False)\n",
      "rna_seq_exons_fh = tempfile.NamedTemporaryFile(delete=False)\n",
      "\n",
      "#write genes file\n",
      "[ rna_seq_genes_fh.write('\\t'.join(gtf_line) + '\\n') for gtf_line in gtf_lines_genes ]\n",
      "#write transcripts file\n",
      "[ rna_seq_trancripts_fh.write('\\t'.join(gtf_line) + '\\n') for gtf_line in gtf_lines_transcripts ]\n",
      "#write exons file\n",
      "[ rna_seq_exons_fh.write('\\t'.join(gtf_line) + '\\n') for gtf_line in gtf_lines_all_exons ]\n",
      "\n",
      "rna_seq_genes_fh.close()\n",
      "rna_seq_trancripts_fh.close()\n",
      "rna_seq_exons_fh.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#ref chlamy v5 transcriptome\n",
      "chlamy_ref_transcriptome = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/Creinhardtii_v5.3_223_gene.gff3'\n",
      "\n",
      "#Chlamy v5 chr sizes\n",
      "chlamy_chr_sizes = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/reference/Chlamy_V5/C_reinhardtii.110311_Chlroplast_mitochondira/chlamy_v5_chr_sizes.txt'\n",
      "\n",
      "#rna_seq_assembly\n",
      "rna_seq_assembly_file = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/assembly/assembly_CWong_modified_cufflinks/merged.gtf'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#cufflinks completed RNA-Seq assembly\n",
      "rna_seq_assembly = pybedtools.BedTool(rna_seq_assembly_file)\n",
      "\n",
      "#RNA-Seq genes\n",
      "rna_seq_genes = pybedtools.BedTool(rna_seq_genes_fh.name)\n",
      "\n",
      "#RNA-Seq Transcripts\n",
      "rna_seq_transcripts = pybedtools.BedTool(rna_seq_trancripts_fh.name)\n",
      "\n",
      "#rna-seq exons\n",
      "rna_seq_exons = pybedtools.BedTool(rna_seq_exons_fh.name)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#available ref trasncriptome\n",
      "cleaned_ref_transcriptome = pybedtools.BedTool(chlamy_ref_transcriptome).remove_invalid().saveas()\n",
      "\n",
      "#ref genes\n",
      "ref_genes = subset_featuretypes(cleaned_ref_transcriptome,['gene']).saveas()\n",
      "\n",
      "#ref transcripts\n",
      "ref_mRNA = subset_featuretypes(cleaned_ref_transcriptome,['mRNA']).saveas()\n",
      "\n",
      "#everything except mRNA and genes feature types\n",
      "ref_exons = subset_featuretypes(cleaned_ref_transcriptome,['gene','mRNA'],inverse_flag=True).saveas()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#get the size of the rna-Seq transcriptome : uniq regions\n",
      "rna_seq_transcriptome_uniq_regions = rna_seq_transcripts.merge(s=True).saveas()\n",
      "rna_seq_transcriptome_uniq_region_size = np.sum([ int(region.length) for region in rna_seq_transcriptome_uniq_regions ])\n",
      "\n",
      "#get the size of the ref mRNA : uniq regions\n",
      "ref_transcriptome_uniq_regions = ref_mRNA.merge(s=True).saveas()\n",
      "ref_transcriptome_uniq_regions_size = np.sum([ int(region.length) for region in ref_transcriptome_uniq_regions ])\n",
      "\n",
      "print rna_seq_transcriptome_uniq_region_size\n",
      "print ref_transcriptome_uniq_regions_size"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#number of novel genes predicted by RNA-Seq assembly\n",
      "rna_seq_novel_genes = rna_seq_genes.intersect(ref_genes, v=True).saveas()\n",
      "\n",
      "#number of novel transcripts predicted by RNA-Seq assembly\n",
      "rna_seq_novel_transcripts = rna_seq_transcripts.intersect(ref_mRNA, v=True).saveas()\n",
      "\n",
      "#number of novel antisense genes\n",
      "rna_seq_antisense_novel_genes = rna_seq_genes.intersect(ref_genes, S=True).saveas()\n",
      "\n",
      "#number of novel antisense transcripts\n",
      "rna_seq_antisense_novel_transcripts = rna_seq_transcripts.intersect(ref_mRNA, S=True).saveas()\n",
      "\n",
      "#RNA-Seq novel exons\n",
      "rna_seq_novel_exons = rna_seq_exons.intersect(ref_exons,v=True).saveas()\n",
      "\n",
      "#RNA-Seq novel exons inside ref genes\n",
      "rna_seq_novel_exons_inside_ref_genes = rna_seq_novel_exons.intersect(ref_genes).saveas()\n",
      "\n",
      "#RNA-Seq novel exons outside ref genes\n",
      "rna_seq_novel_exons_outside_ref_genes = rna_seq_novel_exons.intersect(ref_genes,v=True).saveas()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#finding the mergers and extensions\n",
      "\n",
      "## find the complement of ref genes\n",
      "ref_genes_complement=ref_genes.complement(g=chlamy_chr_sizes).saveas()\n",
      "##see how many RNA_Seq genes interesect with ref_gene complement \n",
      "temp=rna_seq_genes.intersect(ref_genes_complement,u=True).saveas()\n",
      "## do the coverage of rna_seq genes that intersect with ref_genes\n",
      "## count = 1 should be extensions\n",
      "## count > 1 should be mergers\n",
      "count_rna_seq_genes_with_extension_N_mergers = ref_genes.coverage(temp,s=True,counts=True).saveas()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "counts_mergers_extensions = [ int(line[9]) for line in count_rna_seq_genes_with_extension_N_mergers ]\n",
      "print len(counts_mergers_extensions)\n",
      "plt.hist(counts_mergers_extensions)\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import Counter\n",
      "Counter(counts_mergers_extensions)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "######\n",
      "##Comparing the #exons per ref_mRNA and RNA-Seq transcripts\n",
      "######\n",
      "\n",
      "\n",
      "#number of exons(CDS and UTR) in reference mRNA\n",
      "coverage_ref_exons_per_ref_mRNA= ref_exons.coverage(ref_mRNA,s=True,counts=True).saveas()\n",
      "counts_num_exons_per_ref_mRNA = [ math.log(int(line[9]),2) for line in coverage_ref_exons_per_ref_mRNA ]\n",
      "\n",
      "#number of RNA-Seq exons in a RNA-Seq transcripts\n",
      "counts_num_exons_per_rna_seq_transcript = [ math.log(int(line.attrs['exon_number']),2) for line in rna_seq_transcripts]\n",
      "\n",
      "plt.hist(counts_num_exons_per_ref_mRNA,bins=100,label='#exons/ref mRNA')\n",
      "plt.hist(counts_num_exons_per_rna_seq_transcript,bins=100,label='#exons/rna-seq model')\n",
      "plt.legend()\n",
      "plt.title('Comparing #exons per ref-model v/s RNA-Seq Model')\n",
      "plt.xlabel('#exons/RNA-Seq transcript(log2)')\n",
      "plt.ylabel('count')\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def apply_log_on_list(input,log=10):\n",
      "    return [ math.log(x,log) for x in input ]\n",
      "\n",
      "\n",
      "\n",
      "########\n",
      "## Compare the lengths of transcripts\n",
      "########\n",
      "\n",
      "len_ref_genes = [ int(gene.length)  for gene in ref_genes ]\n",
      "len_ref_mRNA = [ int(mRNA.length)  for mRNA in ref_mRNA ]\n",
      "len_rna_seq_genes = [ int(gene.length)  for gene in rna_seq_genes ]\n",
      "len_rna_seq_transcripts = [ int(mRNA.length)  for mRNA in rna_seq_transcripts ]\n",
      "len_rna_seq_transcripts_novel = [ int(mRNA.length)  for mRNA in rna_seq_novel_transcripts ]\n",
      "len_rna_seq_transcripts_novel_antisense = [ int(mRNA.length) for mRNA in rna_seq_antisense_novel_transcripts ]\n",
      "len_rna_seq_transcripts_exon_gt_1 = [ int(mRNA.length)  for mRNA in rna_seq_transcripts if int(mRNA.attrs['exon_number']) > 1]\n",
      "len_rna_seq_transcripts_exon_gt_2 = [ int(mRNA.length)  for mRNA in rna_seq_transcripts if int(mRNA.attrs['exon_number']) > 2]\n",
      "\n",
      "\n",
      "plt.boxplot([   apply_log_on_list(len_ref_genes), \n",
      "                apply_log_on_list(len_ref_mRNA), \n",
      "                apply_log_on_list(len_rna_seq_transcripts),\n",
      "                apply_log_on_list(len_rna_seq_transcripts_novel),\n",
      "                apply_log_on_list(len_rna_seq_transcripts_novel_antisense),\n",
      "                apply_log_on_list(len_rna_seq_transcripts_exon_gt_1),\n",
      "                apply_log_on_list(len_rna_seq_transcripts_exon_gt_2)\n",
      "            ])\n",
      "plt.xticks([1,2,3,4,5,6,7],['ref_genes','ref_mRNA','rna-seq-transcripts(T)','novel T','novel-antisense T','T(exon>1)','T(exon>2)'])\n",
      "plt.ylabel('Transcript Length(log-10)')\n",
      "plt.title('Transcript Length / Types')\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#################\n",
      "##STATS Reporting\n",
      "#################\n",
      "\n",
      "\n",
      "\n",
      "#ref genome\n",
      "#genes\n",
      "\n",
      "print '================'\n",
      "print 'Ref Model'\n",
      "print '================'\n",
      "print 'genes: %d' % len(ref_genes)\n",
      "print 'Length: %d +/- %d Median: %d ' % ( np.mean(len_ref_genes), np.std(len_ref_genes), np.median(len_ref_genes) ) \n",
      "\n",
      "print 'Transcripts %d' % len(ref_mRNA)\n",
      "print 'Length: %d +/- %d  Median: %d' % ( np.mean(len_ref_mRNA), np.std(len_ref_mRNA), np.median(len_ref_mRNA)) \n",
      "\n",
      "print 'Uniq Ref Transcriptome Size %d' % ( ref_transcriptome_uniq_regions_size)\n",
      "\n",
      "print '\\n\\n\\n'\n",
      "\n",
      "print '================'\n",
      "print 'RNA-Seq Model'\n",
      "print '================'\n",
      "print 'genes: %d' % len(rna_seq_genes)\n",
      "print 'Length: %d +/- %d  Median: %d' % ( np.mean(len_rna_seq_genes), np.std(len_rna_seq_genes), np.median(len_rna_seq_genes)) \n",
      "\n",
      "print 'Transcripts: %d' % len(rna_seq_transcripts)\n",
      "print 'Length: %d +/- %d  Median: %d' % ( np.mean(len_rna_seq_transcripts), np.std(len_rna_seq_transcripts), np.median(len_rna_seq_transcripts))\n",
      "print 'Uniq RNA-Seq Transcriptome Size %d' % ( rna_seq_transcriptome_uniq_region_size)\n",
      "#number of novel genes predicted by RNA-Seq assembly\n",
      "print 'Novel genes: %d' % len(rna_seq_novel_genes)\n",
      "#number of novel antisense genes\n",
      "print 'Novel Antisense: %d' % len(rna_seq_antisense_novel_genes)\n",
      "#RNA-Seq novel exons\n",
      "print 'Novel exons: %d' % len(rna_seq_novel_exons)\n",
      "#RNA-Seq novel exons inside ref genes\n",
      "print 'Novel exons inside ref genes: %d' % len(rna_seq_novel_exons_inside_ref_genes)\n",
      "#RNA-Seq novel exons outside ref genes\n",
      "print 'Novel exons inside ref genes: %d' % len(rna_seq_novel_exons_outside_ref_genes)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#########################\n",
      "##Filtering transcripts that are only present in one condition\n",
      "## compute costly step\n",
      "#########################\n",
      "\n",
      "\n",
      "#path to individual assemblies cufflinks assemblies for each condition\n",
      "path_to_individual_assemblies = '/projectb/scratch/chwong/RNA-Seq/Chlamy/NitrogenDeprivation'\n",
      "pattern = 'transcripts.gtf'\n",
      "files = utils.get_FilesList(path_to_individual_assemblies,pattern)\n",
      "\n",
      "#keep only the needed assemblies\n",
      "individual_assem_files = [ file  for file in files if file.find('cuff') != -1 ]\n",
      "\n",
      "\n",
      "\n",
      "conditions = ['0m','10m','30m','1h','2h','6h','8h','24h','48h']\n",
      "\n",
      "data = []\n",
      "for condition in conditions:\n",
      "    temp_holder = [ ]\n",
      "    for assm_file in individual_assem_files:\n",
      "        if assm_file.find('cuff'+condition) != -1:\n",
      "            print 'cuff'+condition,assm_file\n",
      "            temp_bedtool = pybedtools.BedTool(assm_file)\n",
      "            #find the number of transcripts from each condition that overalp with the merged gtf\n",
      "            temp = subset_featuretypes(temp_bedtool,['transcript']).coverage(rna_seq_transcripts,s=True,counts=True)\n",
      "            temp_holder.append(temp)\n",
      "    #finding the minimum count across two replicates of a sample condition\n",
      "    data.append([ min(int(a.fields[9]),int(b.fields[9])) for a,b in itertools.izip(temp_holder[0],temp_holder[1]) ])\n",
      "#print data\n",
      "\n",
      "#coverting to np array\n",
      "np_data = np.array(data).transpose()\n",
      "            \n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#iterate each row of np.array and extracts rows in which count of 0 == 8\n",
      "print len([ row_index for row_index,row in enumerate(np_data) if collections.Counter(row).get(0,0) == 8 ])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "index = [ feature.attrs['transcript_id'] for feature in rna_seq_transcripts ]\n",
      "temp_pandas_df = pandas.DataFrame.from_records(list(np.array(data).transpose()))\n",
      "#?temp_pandas_df.T.set_index()\n",
      "temp_pandas_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "\n"
     ]
    }
   ],
   "metadata": {}
  }
 ]
}