{
 "metadata": {
  "name": "Chlamy_Nitrogren_deprivation_data_analysis"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas\n",
      "import matplotlib.pyplot as plt\n",
      "sys.path.append('/global/homes/a/apratap/dev/eclipse_workspace/python_scripts/lib/');\n",
      "import utils\n",
      "import os\n",
      "import re"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "def create_data_array_from_mapped_files(file_list):\n",
      "    \n",
      "    SUB = 'create_data_array_from_files'\n",
      "    data = []\n",
      "    \n",
      "    for file_name in file_list:\n",
      "        with open(file_name, 'r') as fh:\n",
      "            file_basename = os.path.basename(file_name)\n",
      "            pattern='^.+?experiment\\/(N_.+?)_([AB]).+?accepted_hits_(.+?)\\.bam\\.alignstats$'\n",
      "            matches = re.search(pattern,file_name)\n",
      "            #print '[%s]: Reading data from file file %s' % (SUB,file_basename)\n",
      "            if matches:\n",
      "                #print matches.group(1)\n",
      "                #print matches.group(2)\n",
      "                #print matches.group(3)\n",
      "                #print '----------------------'\n",
      "                condition=matches.group(1)\n",
      "                replicate=matches.group(2)\n",
      "                strand = matches.group(3)\n",
      "            else:\n",
      "                print '[%s]:Warn: Could not find condition and replicate info'\n",
      "                condition = 'NA'\n",
      "                replicate = 'NA' \n",
      "            \n",
      "            \n",
      "            \n",
      "            for (line_num,line) in enumerate(fh):\n",
      "                \n",
      "                #ignore first 3 lines\n",
      "                if line_num <=2:\n",
      "                    continue\n",
      "                    \n",
      "                (variable,value) = line.strip().split()\n",
      "                if variable == '#primary':\n",
      "                    #print variable,value\n",
      "                    \n",
      "                    data.append([condition,replicate,strand,'primary',value])\n",
      "                    #pass\n",
      "        \n",
      "                if variable == '#secondary':\n",
      "                    #print variable,value\n",
      "                    data.append([condition,replicate,strand,'secondary',value])\n",
      "                    \n",
      "                    \n",
      "            #print data\n",
      "    return data\n",
      "    \n",
      "\n",
      "    \n",
      "def create_data_array_from_unmapped_files(file_list):\n",
      "    \n",
      "    SUB = 'create_data_array_from_unmapped_files'\n",
      "    data = []\n",
      "    \n",
      "    for file_name in file_list:\n",
      "        with open(file_name, 'r') as fh:\n",
      "            file_basename = os.path.basename(file_name)\n",
      "            pattern='^.+?experiment\\/(N_.+?)_([AB]).*$'\n",
      "            matches = re.search(pattern,file_name)\n",
      "            #print '[%s]: Reading data from file file %s' % (SUB,file_name)\n",
      "            if matches:\n",
      "                #print matches.group(1)\n",
      "                #print matches.group(2)\n",
      "                #print matches.group(3)\n",
      "                #print '----------------------'\n",
      "                condition=matches.group(1)\n",
      "                replicate=matches.group(2)\n",
      "                strand = 'NA'\n",
      "            else:\n",
      "                print '[%s]:Warn: Could not find condition and replicate info'\n",
      "                condition = 'NA'\n",
      "                replicate = 'NA' \n",
      "            \n",
      "            \n",
      "            \n",
      "            for (line_num,line) in enumerate(fh):\n",
      "                \n",
      "                #ignore first 3 lines\n",
      "                if line_num <=2:\n",
      "                    continue\n",
      "                    \n",
      "                (variable,value) = line.strip().split()\n",
      "         #       print line\n",
      "                if variable == '#unmapped_reads':\n",
      "          #          print variable,value\n",
      "                    data.append([condition,replicate,strand,'unmapped',value])\n",
      "        #print data_per_file\n",
      "    return data\n",
      "\n",
      "        \n",
      "        \n",
      "        \n",
      "#get the list of all the files to aggregate\n",
      "pattern_mapped_reads = 'accepted*sense*alignstats'\n",
      "bam_files_root_path = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment'\n",
      "\n",
      "\n",
      "files = utils.get_FilesList(bam_files_root_path,pattern=pattern_mapped_reads) \n",
      "data_mapped_reads = create_data_array_from_mapped_files(files)\n",
      "print 'Found %d items back' % (len(data_mapped_reads))\n",
      "\n",
      "\n",
      "\n",
      "pattern_unmapped_reads = 'unmapped*alignstats'\n",
      "bam_files_root_path = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment'\n",
      "\n",
      "\n",
      "files = utils.get_FilesList(bam_files_root_path,pattern=pattern_unmapped_reads) \n",
      "data_unmapped_reads = create_data_array_from_unmapped_files(files)\n",
      "print 'Found %d items back' % (len(data_unmapped_reads))\n",
      "\n",
      "\n",
      "\n",
      "#getting clean reads per sample\n",
      "\n",
      "\n",
      "raw_reads ='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/ERCC_mapping/raw_reads_per_sample.txt'\n",
      "\n",
      "data_raw_reads =[ [x.split()[0],x.split()[1],'NA','raw',x.split()[-1]] for (num,x) in enumerate(open(raw_reads,'r')) if num != 0]\n",
      "\n",
      "\n",
      "data = data_mapped_reads + data_unmapped_reads + data_raw_reads\n",
      "\n",
      "df = pandas.DataFrame(data,columns=['condition','replicate','strand','type','count'])\n",
      "\n",
      "\n",
      "print df\n",
      "mapping_summary ='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/ERCC_mapping/mapping_summary.tsv'\n",
      "df.to_csv(mapping_summary,sep='\\t',index=False)\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "[get_FilesList]: Found 36 files at /global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment\n",
        "Found 72 items back\n",
        "[get_FilesList]: Found 18 files at /global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Found 18 items back\n",
        "    condition replicate     strand       type      count\n",
        "0        N_0m         A  antisense    primary   62204046\n",
        "1        N_0m         A  antisense  secondary    7975140\n",
        "2        N_0m         A      sense    primary   61019168\n",
        "3        N_0m         A      sense  secondary   11333047\n",
        "4        N_0m         B      sense    primary   95292456\n",
        "5        N_0m         B      sense  secondary   13373556\n",
        "6        N_0m         B  antisense    primary   98772705\n",
        "7        N_0m         B  antisense  secondary   11621667\n",
        "8       N_10m         A      sense    primary   49608856\n",
        "9       N_10m         A      sense  secondary    9202513\n",
        "10      N_10m         A  antisense    primary   51154038\n",
        "11      N_10m         A  antisense  secondary    6774957\n",
        "12      N_10m         B  antisense    primary   46007007\n",
        "13      N_10m         B  antisense  secondary    5551739\n",
        "14      N_10m         B      sense    primary   44224315\n",
        "15      N_10m         B      sense  secondary    6446127\n",
        "16       N_1h         A      sense    primary   54616787\n",
        "17       N_1h         A      sense  secondary    9159943\n",
        "18       N_1h         A  antisense    primary   56714898\n",
        "19       N_1h         A  antisense  secondary    8471861\n",
        "20       N_1h         B  antisense    primary  138079868\n",
        "21       N_1h         B  antisense  secondary   23681824\n",
        "22       N_1h         B      sense    primary  134258833\n",
        "23       N_1h         B      sense  secondary   28062527\n",
        "24      N_24h         A  antisense    primary   56892492\n",
        "25      N_24h         A  antisense  secondary    7117464\n",
        "26      N_24h         A      sense    primary   52965023\n",
        "27      N_24h         A      sense  secondary    7817425\n",
        "28      N_24h         B  antisense    primary   48089160\n",
        "29      N_24h         B  antisense  secondary    5903421\n",
        "30      N_24h         B      sense    primary   45266617\n",
        "31      N_24h         B      sense  secondary    6386022\n",
        "32       N_2h         A      sense    primary   58712421\n",
        "33       N_2h         A      sense  secondary    9581136\n",
        "34       N_2h         A  antisense    primary   61705347\n",
        "35       N_2h         A  antisense  secondary    8660026\n",
        "36       N_2h         B  antisense    primary   70982335\n",
        "37       N_2h         B  antisense  secondary   11518735\n",
        "38       N_2h         B      sense    primary   67946934\n",
        "39       N_2h         B      sense  secondary   12474355\n",
        "40      N_30m         A      sense    primary   63085088\n",
        "41      N_30m         A      sense  secondary   10430735\n",
        "42      N_30m         A  antisense    primary   65225712\n",
        "43      N_30m         A  antisense  secondary    8865565\n",
        "44      N_30m         B      sense    primary   87083911\n",
        "45      N_30m         B      sense  secondary   12149313\n",
        "46      N_30m         B  antisense    primary   90778614\n",
        "47      N_30m         B  antisense  secondary   11351237\n",
        "48      N_48h         A      sense    primary   45694312\n",
        "49      N_48h         A      sense  secondary    6218094\n",
        "50      N_48h         A  antisense    primary   48676793\n",
        "51      N_48h         A  antisense  secondary    5479428\n",
        "52      N_48h         B      sense    primary   40783220\n",
        "53      N_48h         B      sense  secondary    5805254\n",
        "54      N_48h         B  antisense    primary   42976051\n",
        "55      N_48h         B  antisense  secondary    5178664\n",
        "56       N_6h         A  antisense    primary   62318002\n",
        "57       N_6h         A  antisense  secondary    8397774\n",
        "58       N_6h         A      sense    primary   57754671\n",
        "59       N_6h         A      sense  secondary   10234370\n",
        "60       N_6h         B      sense    primary   67813719\n",
        "61       N_6h         B      sense  secondary   13792577\n",
        "62       N_6h         B  antisense    primary   72929354\n",
        "63       N_6h         B  antisense  secondary   10974171\n",
        "64       N_8h         A      sense    primary   41844914\n",
        "65       N_8h         A      sense  secondary    6693241\n",
        "66       N_8h         A  antisense    primary   45359733\n",
        "67       N_8h         A  antisense  secondary    5371446\n",
        "68       N_8h         B  antisense    primary   48345895\n",
        "69       N_8h         B  antisense  secondary    6085482\n",
        "70       N_8h         B      sense    primary   44445628\n",
        "71       N_8h         B      sense  secondary    7334313\n",
        "72       N_0m         A         NA   unmapped    7291355\n",
        "73       N_0m         B         NA   unmapped    9564216\n",
        "74      N_10m         A         NA   unmapped    6552129\n",
        "75      N_10m         B         NA   unmapped    3876929\n",
        "76       N_1h         A         NA   unmapped    8108125\n",
        "77       N_1h         B         NA   unmapped   30212234\n",
        "78      N_24h         A         NA   unmapped    6310537\n",
        "79      N_24h         B         NA   unmapped    5343857\n",
        "80       N_2h         A         NA   unmapped    8125186\n",
        "81       N_2h         B         NA   unmapped   11182571\n",
        "82      N_30m         A         NA   unmapped    8737593\n",
        "83      N_30m         B         NA   unmapped    8928584\n",
        "84      N_48h         A         NA   unmapped    3715938\n",
        "85      N_48h         B         NA   unmapped    3785309\n",
        "86       N_6h         A         NA   unmapped    7726185\n",
        "87       N_6h         B         NA   unmapped    9148915\n",
        "88       N_8h         A         NA   unmapped    3878563\n",
        "89       N_8h         B         NA   unmapped    4087768\n",
        "90       N_0m         A         NA        raw  142845286\n",
        "91       N_0m         B         NA        raw  220941582\n",
        "92      N_10m         A         NA        raw  122929186\n",
        "93      N_10m         B         NA        raw   99819438\n",
        "94      N_30m         A         NA        raw  151539030\n",
        "95      N_30m         B         NA        raw  194029460\n",
        "96       N_1h         A         NA        raw  129109020\n",
        "97       N_1h         B         NA        raw  334231948\n",
        "98       N_2h         A         NA        raw  156011904\n",
        "99       N_2h         B         NA        raw  163852246\n",
        "100      N_6h         A         NA        raw  138511326\n",
        "101      N_6h         B         NA        raw  162069704\n",
        "102      N_8h         A         NA        raw   95650610\n",
        "103      N_8h         B         NA        raw  100741726\n",
        "104     N_24h         A         NA        raw  124011462\n",
        "105     N_24h         B         NA        raw  109200410\n",
        "106     N_48h         A         NA        raw  104185460\n",
        "107     N_48h         B         NA        raw   93643170\n"
       ]
      }
     ],
     "prompt_number": 128
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rc = '#raw_reads        72352215'\n",
      "print c.strip().split(q)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "plantTFDB_PFAM_hits_file='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/chlamy_PlnTFDB/PlnTFDB_PFAM_hits_592035950.txt'\n",
      "plantTFDB_tf_file='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/chlamy_PlnTFDB/PlnTFDB_tf_1680431410.txt'\n",
      "\n",
      "def read_chlamy_plantTFdb_TF_info(plantTFDB_PFAM_hits_file,plantTFDB_tf_file):\n",
      "    \n",
      "    protein_id_to_TF_name = {};\n",
      "    \n",
      "    #reading the TF domains\n",
      "    for line in open(plantTFDB_PFAM_hits_file,'r'):\n",
      "        line_split = line.strip().split('\\t')\n",
      "        protein_id = line_split[1]\n",
      "        tf_domain  = line_split[2]\n",
      "        \n",
      "        if protein_id_to_TF_name.get(protein_id) is None:\n",
      "            protein_id_to_TF_name[protein_id] = set([tf_domain])\n",
      "        else:\n",
      "            #add to set\n",
      "            protein_id_to_TF_name[protein_id].add(tf_domain)\n",
      "    \n",
      "    #reading the TF family\n",
      "    for line in open(plantTFDB_tf_file,'r'):\n",
      "        line_split = line.strip().split('\\t')\n",
      "        protein_id = line_split[1]\n",
      "        tf_domain  = str(line_split[2])\n",
      "        \n",
      "        if protein_id_to_TF_name.get(protein_id) is None:\n",
      "            protein_id_to_TF_name[protein_id] = set([tf_domain])  #like this logic\n",
      "        else:\n",
      "            #add to set\n",
      "            protein_id_to_TF_name[protein_id].add(tf_domain)\n",
      "    \n",
      "    print \"retreived annotation for %d proteins id's \" % (len(protein_id_to_TF_name))\n",
      "    return protein_id_to_TF_name\n",
      "    \n",
      "protein_id_to_TF_name = read_chlamy_plantTFdb_TF_info(plantTFDB_PFAM_hits_file,plantTFDB_tf_file)\n",
      "\n",
      "\n",
      "\n",
      "#sample output\n",
      "'''\n",
      "protein-id : PFAM-domain, TF-family\n",
      "414959 set(['TRAF', 'BTB'])\n",
      "400305 set(['GNAT', 'Acetyltransf_1', 'zf-MYND'])\n",
      "426843 set(['PLATZ'])\n",
      "142190 set(['C2C2-GATA', 'GATA'])\n",
      "145759 set(['TRAF', 'BTB'])\n",
      "20566 set(['VARL', 'Involucrin'])\n",
      "177225 set(['ACT', 'bHLH', 'HLH'])\n",
      "410707 set(['Response_reg', 'ARR-B', 'G2-like'])\n",
      "193299 set(['SAM_1', 'HisKA', 'Orphans', 'HATPase_c', 'Response_reg', 'Bac_rhodopsin', 'Guanylate_cyc'])\n",
      "187248 set(['RB_A', 'RB', 'RB_B'])\n",
      "377858 set(['Myb_DNA-binding', 'MYB'])\n",
      "'''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "retreived annotation for 368 proteins id's \n"
       ]
      },
      {
       "output_type": "pyout",
       "prompt_number": 46,
       "text": [
        "\"\\nprotein-id : PFAM-domain, TF-family\\n414959 set(['TRAF', 'BTB'])\\n400305 set(['GNAT', 'Acetyltransf_1', 'zf-MYND'])\\n426843 set(['PLATZ'])\\n142190 set(['C2C2-GATA', 'GATA'])\\n145759 set(['TRAF', 'BTB'])\\n20566 set(['VARL', 'Involucrin'])\\n177225 set(['ACT', 'bHLH', 'HLH'])\\n410707 set(['Response_reg', 'ARR-B', 'G2-like'])\\n193299 set(['SAM_1', 'HisKA', 'Orphans', 'HATPase_c', 'Response_reg', 'Bac_rhodopsin', 'Guanylate_cyc'])\\n187248 set(['RB_A', 'RB', 'RB_B'])\\n377858 set(['Myb_DNA-binding', 'MYB'])\\n\""
       ]
      }
     ],
     "prompt_number": 46
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "####\n",
      "#Chlamy Differential expression analysis\n",
      "####\n",
      "\n",
      "\n",
      "import sys\n",
      "import re\n",
      "from pysam import Fastafile\n",
      "import fileinput\n",
      "\n",
      "def dna_RevCompliment(seq):\n",
      "    '''given a dna sequence just returns the reverser compliment on the same strand\n",
      "    mainly done to find the sequence of the other strand in the 5' -> 3' direction\n",
      "    '''\n",
      "    rev_seq = seq[::-1]\n",
      "    rev_table = string.maketrans('ACGTacgt','TGCAtgca')\n",
      "    return rev_seq.translate(rev_table)\n",
      "\n",
      "def create_cuff_gene_ref_gene_loopup_table():\n",
      "    '''\n",
      "    create a dict to lookup reference gene names \n",
      "    given a cuff gene name\n",
      "    '''\n",
      "    chlamy_lookup_table_cuff_gene_to_ref_gene = {}\n",
      "    for num,line in enumerate(open('/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/diff_expression/2_with_merged_gtf/cuff_genes_to_ref_gene_names.txt','r')):\n",
      "        line_split=line.strip().split('\\t')\n",
      "        chlamy_lookup_table_cuff_gene_to_ref_gene[line_split[0]] = line_split[1]\n",
      "    return (chlamy_lookup_table_cuff_gene_to_ref_gene)\n",
      "\n",
      "\n",
      "\n",
      "def get_chlamy_gene_function_info():\n",
      "    #reading the function annotation\n",
      "    chlamy_func_annotation_file='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/Creinhardtii_v5.3_223_phytozome_functional_annotation.csv'\n",
      "    test_data='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/test.data'\n",
      "\n",
      "    chlamy_function_info = {}\n",
      "    for num,line in enumerate(open(chlamy_func_annotation_file,'r')):\n",
      "        if num == 0 :\n",
      "            continue\n",
      "        line_split = line.replace('\"','').strip().split('\\t')\n",
      "        \n",
      "        #skipping lines which only have gene name\n",
      "        if len(line_split) < 2:\n",
      "            #print line_split\n",
      "            continue\n",
      "        \n",
      "       #gene name \n",
      "        if line_split[0]:\n",
      "            gene_name = re.sub('\\.t.*','',line_split[0])\n",
      "       #gene common name\n",
      "        if len(line_split) >= 4 and line_split[3]:\n",
      "            gene_common_name=line_split[3]\n",
      "        else:\n",
      "            gene_common_name='NA' \n",
      "        #defline : col2\n",
      "        if line_split[1]:\n",
      "            gene_defline = line_split[1]\n",
      "        else:\n",
      "            gene_defline = 'NA'\n",
      "        #function : col14\n",
      "        \n",
      "        if len(line_split) >= 14 and line_split[13]:\n",
      "            gene_func = line_split[13]\n",
      "        else:\n",
      "            gene_func = 'NA'\n",
      "        \n",
      "        \n",
      "        #print \"%s \\t %s \\t %s \\t %s\" % (gene_name,gene_common_name,gene_defline,gene_func)\n",
      "        \n",
      "        chlamy_function_info[gene_name] = (gene_defline,gene_func)\n",
      "        chlamy_function_info[gene_common_name] = (gene_defline,gene_func)\n",
      "        \n",
      "    return chlamy_function_info\n",
      "\n",
      "\n",
      "\n",
      "def read_cuff_diff_gene_exp_data(cuff_diff_gene_file,base_condition,test_condition):\n",
      "    cuff_gene_fold_change = {}\n",
      "    for line in open(cuff_diff_gene_file,'r'):\n",
      "        line_split=line.strip().split('\\t')\n",
      "        \n",
      "        if line_split[4].strip() == base_condition and line_split[5].strip() == test_condition:\n",
      "            gene_name = line_split[0].strip()\n",
      "            gene_fold_change = line_split[9].strip()\n",
      "            cuff_gene_fold_change[gene_name]=gene_fold_change\n",
      "    return cuff_gene_fold_change\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def parse_gtf_file(gtf_file):\n",
      "    \"\"\"\n",
      "    given a gtf file as input this function will parse and return a single line\n",
      "    is a generator : yield single parsed gtf entry one at a time\n",
      "    \"\"\"\n",
      "    \n",
      "    #sample_gtf_line (split in 3 lines for display purposes)\n",
      "    #chromosome_1    Cufflinks       exon    20356   20687   .       +       .       \n",
      "    #gene_id \"XLOC_000036\"; transcript_id \"TCONS_00000050\"; exon_number \"1\"; gene_name \"g3\"; oId \n",
      "    #\"PAC:26903463\"; contained_in \"TCONS_00000051\"; nearest_ref \"PAC:26903463\"; class_code \"=\"; tss_id \"TSS41\"; p_id \"P3\";\n",
      "\n",
      "    ##just creating an abtract class/container to hold the values for features of a gtf line\n",
      "    class gtf_line_container(object):\n",
      "        \n",
      "        def __init__(self):\n",
      "            self.chr=None\n",
      "            self.start=None\n",
      "            self.end=None\n",
      "            self.transcript_id=None\n",
      "            self.gene_id=None\n",
      "            self.feature_type=None\n",
      "            self.attributes=None\n",
      "    \n",
      "    for num,line in enumerate(open(gtf_file,'r')):\n",
      "        \n",
      "        line_split = line.strip().split('\\t')\n",
      "        \n",
      "        #instantiate the container\n",
      "        gtf_line = gtf_line_container()\n",
      "        \n",
      "        attributes_split=line_split[8].split(';')\n",
      "        #sample attribute split line\n",
      "        #['gene_id \"XLOC_000001\"', ' transcript_id \"TCONS_00000001\"', ' exon_number \"1\"', ' oId \"CUFF.31.1\"', ' class_code \"u\"', ' tss_id \"TSS1\"', '']\n",
      "        \n",
      "        #gene_id\n",
      "        gene_id=attributes_split[0].split(' ')[1].replace('\"','')\n",
      "        \n",
      "        #transcrpit_id\n",
      "        transcript_id=attributes_split[1].split(' ')[2].replace('\"','')\n",
      "        \n",
      "        \n",
      "        gtf_line.chr=line_split[0].strip()\n",
      "        gtf_line.start=line_split[3].strip()\n",
      "        gtf_line.end=line_split[4].strip()\n",
      "        gtf_line.strand=line_split[6].strip()\n",
      "        gtf_line.feature_type=line_split[2].strip()\n",
      "        gtf_line.gene_id = gene_id\n",
      "        gtf_line.transcript_id = transcript_id\n",
      "        gtf_line.attributes = attributes_split\n",
      "        yield gtf_line\n",
      "    \n",
      "\n",
      "\n",
      "def summarize_cuffmerge_output(merged_gtf_file,summarize_by=None):\n",
      "    '''\n",
      "    given a merged.gtf file produced by cuffmerge this method can\n",
      "    summarize it by gene or transcript level\n",
      "    **PS: assumption is that for each gene the exons are reported, sorted by position**\n",
      "    variable : summarize_by can take the value of gene,transcript\n",
      "    '''\n",
      "    \n",
      "    \n",
      "    features = {}\n",
      "    \n",
      "    class feature(object):\n",
      "        \n",
      "        def __init__(self):\n",
      "            self.gene_id = None\n",
      "            self.transcript_id = None\n",
      "            self.start = None\n",
      "            self.end = None\n",
      "            self.strand = None\n",
      "            self.attributes = None\n",
      "            \n",
      "    #default for each feature\n",
      "    last_feature_id = None\n",
      "    feature_start = 0;\n",
      "    feature_chromosome = 'NA';\n",
      "    feature_end = 0;\n",
      "    feature_strand = 0;\n",
      "    \n",
      "    \n",
      "    #setting the right feture name\n",
      "    if summarize_by == 'gene':\n",
      "        feature_name = 'gene_id'\n",
      "    elif summarize_by == 'transcript':\n",
      "        feature_name = 'transcript_id'\n",
      "    else: \n",
      "        sys.stderr('Error : not a recognized summarize_by id %s ' % summarize_by)\n",
      "        sys.exit(2)\n",
      "    \n",
      "    \n",
      "    \n",
      "    for gtf_line in parse_gtf_file(merged_gtf_file):\n",
      "            \n",
      "            #interesting peice of logic\n",
      "            #depending on the user entered summarize_by varibale either the transcript_id or gene_id is used\n",
      "            this_feature_id=eval('gtf_line.%s' % feature_name)\n",
      "        \n",
      "            \n",
      "            #initialization\n",
      "            if ( last_feature_id is None):\n",
      "                last_feature_id = this_feature_id\n",
      "                feature_start = gtf_line.start\n",
      "                feature_end   = gtf_line.end\n",
      "                feature_strand = gtf_line.strand\n",
      "                feature_chromosome = gtf_line.chr\n",
      "                continue\n",
      "            \n",
      "            #if we see the same feature, just update the end\n",
      "            if ( last_feature_id == this_feature_id):\n",
      "                feature_end   = gtf_line.end\n",
      "                \n",
      "            else:\n",
      "                #create the new feature summary object before updating to the new feature\n",
      "                feature_summary = feature()\n",
      "                feature_summary.feature_id = last_feature_id\n",
      "                feature_summary.start = feature_start\n",
      "                feature_summary.end = feature_end\n",
      "                feature_summary.strand = feature_strand\n",
      "                feature_summary.chr = feature_chromosome\n",
      "                \n",
      "                #store the feature summary object in a dict\n",
      "                features[last_feature_id] = feature_summary\n",
      "                \n",
      "                ##now update and store the new feature\n",
      "                last_feature_id = this_feature_id\n",
      "                feature_start = gtf_line.start\n",
      "                feature_end   = gtf_line.end\n",
      "                feature_strand = gtf_line.strand\n",
      "                feature_chromosome = gtf_line.chr\n",
      "    \n",
      "    \n",
      "    feature_summary = feature()\n",
      "    feature_summary.feature_id = last_feature_id\n",
      "    feature_summary.start = feature_start\n",
      "    feature_summary.end = feature_end\n",
      "    feature_summary.strand = feature_strand\n",
      "    feature_summary.chr = feature_chromosome\n",
      "                \n",
      "    #store the feature summary object in a dict\n",
      "    features[last_feature_id] = feature_summary\n",
      "    \n",
      "    print 'Created features dict, summarized by %s of len %d' % (summarize_by,len(features))\n",
      "    return (features)\n",
      " \n",
      "  \n",
      "    \n",
      "        \n",
      "cuffmerge_gtf_file = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/assembly/cuffmerge_with_reference_annotation/merged_asm/merged.gtf'\n",
      "cuffmerge_gtf_file_test = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/assembly/cuffmerge_with_reference_annotation/merged_asm/test.gtf'\n",
      "\n",
      "#dict of cuff gene names poiting to class object which stores start,end,strand,chr\n",
      "cuff_genes_coords = summarize_cuffmerge_output(cuffmerge_gtf_file,summarize_by='gene')\n",
      "\n",
      "#dict of cuff gene to ref_gene name\n",
      "cuff_gene_name_to_ref_gene_lookUP = create_cuff_gene_ref_gene_loopup_table()\n",
      "\n",
      "#dict of ref_gene to annotated function\n",
      "chlamy_gene_function_info = get_chlamy_gene_function_info()\n",
      "\n",
      "#chlamy diff genes file\n",
      "chlamy_diff_exp_gene_file='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/diff_expression/2_with_merged_gtf/gene_exp.diff'\n",
      "cuff_gene_fold_change = read_cuff_diff_gene_exp_data(chlamy_diff_exp_gene_file,'N_0m','N_10m')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Created features dict, summarized by gene of len 14418\n"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "\n",
      "DE_gene_list_0_10m = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/diff_expression/2_with_merged_gtf/analysis_DE_0m_10m_genes/genes_DE_0_10m.txt'\n",
      "DE_gene_list_0_30m = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/diff_expression/2_with_merged_gtf/analysis_DE_0m_30m_genes/genes_DE_0_30m.txt'\n",
      "\n",
      "\n",
      "chlamy_v5_genome='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/reference/Chlamy_V5/C_reinhardtii.110311_Chlroplast_mitochondira/Chlre5_genomic_scaffold_plastids.fasta'\n",
      "chlamy_v5_genome_fasta = Fastafile(chlamy_v5_genome)\n",
      "\n",
      "de_genes_list_file = DE_gene_list_0_10m\n",
      "fout1 = open(de_genes_list_file+'.function','w')\n",
      "#fout2 = open(de_genes_list_file+'.fasta','w')\n",
      "fout1.write(('%s \\t %s \\t %s \\t %s \\t %s \\t %s \\n') % ('cuff_gene_name','ref_gene','gene_regulation','fold_change','phytozome_gene_function','plant_TFDB_annotation'))\n",
      "        \n",
      "\n",
      "total_cuff_DE_genes_up_regulated =0\n",
      "total_cuff_DE_genes_down_regulated =0\n",
      "count_novel_DE_genes = 0\n",
      "count_genes_with_plant_TFDB_annotation = 0\n",
      "count_genes_with_plant_TFDB_annotation_up_reg = 0\n",
      "count_genes_with_plant_TFDB_annotation_down_reg = 0\n",
      "count_genes_with_NO_phytozome_annotation = 0\n",
      "#count_genes_with_plant_TFDB_annot_and_NO_phytozome_annot = 0\n",
      "\n",
      "\n",
      "#if the manual blast is done read the blast results\n",
      "blast_result_file = de_genes_list_file + '.blastresults'\n",
      "cuff_gene_id_to_protein_id = {}\n",
      "#cuff_gene_id_to_protein_id maps the names of the cuff genes for which a blastx hit was found against TF proteins in Plant TFDB\n",
      "if os.path.exists(blast_result_file):\n",
      "    for line in open(blast_result_file,'r'):\n",
      "        line_split=line.strip().split('\\t')\n",
      "        \n",
      "        #sample blast line\n",
      "        #XLOC_000635_chr:chromosome_1_start:1877171_end:1895794_strand:- Chlamydomonas_reinhardtii_187405        70.73   41      10      1       11581   11465   1559    1599    1e-14 58.2\n",
      "        gene_id=line_split[0].split(':')[0].replace('_chr','')\n",
      "        protein_id = line_split[1].split('_')[2]\n",
      "        \n",
      "        #print gene_id,protein_id\n",
      "        cuff_gene_id_to_protein_id[gene_id]=protein_id\n",
      "        \n",
      "        '''\n",
      "        #was trying to store multiple(if any) hits for a each gene\n",
      "        if cuff_gene_id_to_protein_id.get(gene_id) is None:\n",
      "            cuff_gene_id_to_protein_id[gene_id]=set(str(protein_id))\n",
      "        else:\n",
      "            cuff_gene_id_to_protein_id[gene_id].add(str(protein_id))\n",
      "        '''\n",
      "\n",
      "#read the DE genes \n",
      "for num,line in enumerate(open(de_genes_list_file,'r')):\n",
      "    cuff_gene_name = line.strip();\n",
      "    \n",
      "    #print cuff_gene_name\n",
      "    \n",
      "    #following blob to create a fasta which could be blasted against the PltTFD TF protein sequences to find possible matches\n",
      "    '''\n",
      "    if cuff_genes_coords[cuff_gene_name]:\n",
      "        #access the gene feature object\n",
      "        gf = cuff_genes_coords[cuff_gene_name]\n",
      "        #fetching the fastq from the Chlamy_v5_genome\n",
      "        gene_seq = chlamy_v5_genome_fasta.fetch(gf.chr,int(gf.start),int(gf.end)).strip()\n",
      "        fout2.write('>%s_chr:%s_start:%s_end:%s_strand:%s\\n%s\\n' %(cuff_gene_name,gf.chr,gf.start,gf.end,gf.strand,gene_seq))\n",
      "    else:\n",
      "        print 'Warn not found the coords for %s gene' % cuff_gene_name\n",
      "    '''\n",
      "    #the next logical step is to blast the reads against the protein seqeunce\n",
      "    #of TF of Chlamy downloaded from Plant TFDB\n",
      "    #did this step manually\n",
      "    '''\n",
      "    blastx -db ../../../../../annotation/chlamy_PlnTFDB/PlnTFDB_PEP_343042821.fasta -query genes_DE_0_10m.txt.fasta \n",
      "    -outfmt 6 -evalue 1e-10 -show_gis -best_hit_overhang .05 -num_threads 8 -best_hit_score_edge .10 > genes_DE_0_10m.txt.fasta.blastresults\n",
      "    '''\n",
      "    \n",
      "    \n",
      "    #get the fold enrichment\n",
      "    gene_fold_change = cuff_gene_fold_change.get(cuff_gene_name)\n",
      "    if gene_fold_change is None:\n",
      "        gene_fold_change = 0.0\n",
      "    elif float(gene_fold_change) > 0:\n",
      "        gene_regulation = 'up'\n",
      "        total_cuff_DE_genes_up_regulated += 1\n",
      "    else:\n",
      "        gene_regulation='down'\n",
      "        total_cuff_DE_genes_down_regulated += 1\n",
      "    \n",
      "    \n",
      "    \n",
      "    #get the annotation of the cuff gene based on blast results from PlantTFDB\n",
      "    #first get the protein id match for gene and for each protein id get the annotation\n",
      "    #print [  protein_id_to_TF_name.get(each_protein)  for each_protein_id in cuff_gene_id_to_protein_id[cuff_gene_name] ]\n",
      "    gene_proteid_id = cuff_gene_id_to_protein_id.get(cuff_gene_name)\n",
      "    #getting the annotation\n",
      "    plant_TFDB_annotation =  protein_id_to_TF_name.get(gene_proteid_id)\n",
      "    \n",
      "    \n",
      "    if plant_TFDB_annotation is None:\n",
      "        plant_TFDB_annotation_line = 'NA'\n",
      "    else:\n",
      "        plant_TFDB_annotation_line = ','.join(list(plant_TFDB_annotation))\n",
      "        count_genes_with_plant_TFDB_annotation += 1\n",
      "        \n",
      "        if gene_regulation == 'up':\n",
      "            count_genes_with_plant_TFDB_annotation_up_reg +=1\n",
      "        else:\n",
      "            count_genes_with_plant_TFDB_annotation_down_reg +=1\n",
      "        \n",
      "    #print cuff_gene_name,plant_TFDB_annotation_line\n",
      "    \n",
      "    \n",
      "    \n",
      "    \n",
      "    \n",
      "    #get the chlamy ref gene name\n",
      "    ref_gene_name=cuff_gene_name_to_ref_gene_lookUP.get(cuff_gene_name)\n",
      "    \n",
      "    #getting the function for ref_gene based on \n",
      "    for each_ref_gene in ref_gene_name.split(','):\n",
      "        phytozome_gene_function = chlamy_gene_function_info.get(each_ref_gene)\n",
      "        #print cuff_gene_name,ref_gene_name,phytozome_gene_function\n",
      "        if phytozome_gene_function is not None:\n",
      "           phytozome_gene_function_line=''.join([x for x in phytozome_gene_function if x != 'NA'])\n",
      "           if phytozome_gene_function_line == '':  #for the case of NA,NA in phytozome function\n",
      "                phytozome_gene_function_line = 'NA'\n",
      "        else:\n",
      "            phytozome_gene_function_line='NA'\n",
      "            count_genes_with_NO_phytozome_annotation += 1\n",
      "        \n",
      "        #novel gene that is differentially expressed\n",
      "        if ref_gene_name == '-':\n",
      "            count_novel_DE_genes += 1\n",
      "            phytozome_gene_function_line = 'novel-gene'\n",
      "        fout1.write('%s \\t %s \\t %s \\t %s \\t %s \\t %s \\n' % (cuff_gene_name, each_ref_gene,gene_regulation,str(gene_fold_change),phytozome_gene_function_line,plant_TFDB_annotation_line))\n",
      "        \n",
      "    \n",
      "fout1.close()\n",
      "fout2.close()\n",
      "print '#cuff DE genes: %d' % (num+1)\n",
      "print '#up-regulated: %d' % (total_cuff_DE_genes_up_regulated) \n",
      "print '#down-regulated: %d' % (total_cuff_DE_genes_down_regulated)\n",
      "print '#novel DE genes: %d' % (count_novel_DE_genes)\n",
      "print '#DE genes with match to Chlamy Plant TFDB: %d' % (count_genes_with_plant_TFDB_annotation)\n",
      "print '#DE genes with match to Chlamy Plant TFDB and up reg: %d' % (count_genes_with_plant_TFDB_annotation_up_reg)\n",
      "print '#DE genes with match to Chlamy Plant TFDB and down reg: %d' % (count_genes_with_plant_TFDB_annotation_down_reg)\n",
      "print '#DE genes with NO phytozome annotation : %d' % (count_genes_with_NO_phytozome_annotation)\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "XLOC_000001 NA\n",
        "XLOC_000002 NA\n",
        "XLOC_000003 NA\n",
        "XLOC_000004 NA\n",
        "XLOC_000005 NA\n",
        "XLOC_000006 NA\n",
        "XLOC_000007 NA\n",
        "XLOC_000008 NA\n",
        "XLOC_000011 NA\n",
        "XLOC_000012 NA\n",
        "XLOC_000015 NA\n",
        "XLOC_000016 NA\n",
        "XLOC_000018 NA\n",
        "XLOC_000022 NA\n",
        "XLOC_000023 NA\n",
        "XLOC_000025 NA\n",
        "XLOC_000026 NA\n",
        "XLOC_000027 NA\n",
        "XLOC_000028 NA\n",
        "XLOC_000029 NA\n",
        "XLOC_000037 NA\n",
        "XLOC_000046 NA\n",
        "XLOC_000047 NA\n",
        "XLOC_000061 NA\n",
        "XLOC_000065 NA\n",
        "XLOC_000075 NA\n",
        "XLOC_000076 NA\n",
        "XLOC_000079 NA\n",
        "XLOC_000088 NA\n",
        "XLOC_000095 NA\n",
        "XLOC_000102 NA\n",
        "XLOC_000113 NA\n",
        "XLOC_000114 NA\n",
        "XLOC_000120 NA\n",
        "XLOC_000124 NA\n",
        "XLOC_000127 NA\n",
        "XLOC_000145 NA\n",
        "XLOC_000154 NA\n",
        "XLOC_000155 NA\n",
        "XLOC_000156 NA\n",
        "XLOC_000158 SBP\n",
        "XLOC_000162 NA\n",
        "XLOC_000163 NA\n",
        "XLOC_000172 PHD\n",
        "XLOC_000175 NA\n",
        "XLOC_000183 NA\n",
        "XLOC_000184 NA\n",
        "XLOC_000201 NA\n",
        "XLOC_000207 NA\n",
        "XLOC_000220 NA\n",
        "XLOC_000230 NA\n",
        "XLOC_000244 NA\n",
        "XLOC_000259 NA\n",
        "XLOC_000280 NA\n",
        "XLOC_000292 NA\n",
        "XLOC_000294 NA\n",
        "XLOC_000300 NA\n",
        "XLOC_000317 NA\n",
        "XLOC_000329 NA\n",
        "XLOC_000341 NA\n",
        "XLOC_000355 Response_reg,Orphans,HisKA,Bac_rhodopsin,HATPase_c\n",
        "XLOC_000358 NA\n",
        "XLOC_000366 NA\n",
        "XLOC_000372 NA\n",
        "XLOC_000373 NA\n",
        "XLOC_000385 NA\n",
        "XLOC_000387 NA\n",
        "XLOC_000390 NA\n",
        "XLOC_000397 NA\n",
        "XLOC_000398 NA\n",
        "XLOC_000404 NA\n",
        "XLOC_000409 NA\n",
        "XLOC_000425 NA\n",
        "XLOC_000443 NA\n",
        "XLOC_000445 NA\n",
        "XLOC_000451 NA\n",
        "XLOC_000463 NA\n",
        "XLOC_000465 NA\n",
        "XLOC_000471 NA\n",
        "XLOC_000474 NA\n",
        "XLOC_000477 NA\n",
        "XLOC_000479 NA\n",
        "XLOC_000488 NA\n",
        "XLOC_000499 NA\n",
        "XLOC_000500 NA\n",
        "XLOC_000506 NA\n",
        "XLOC_000512 NA\n",
        "XLOC_000517 NA\n",
        "XLOC_000532 NA\n",
        "XLOC_000539 NA\n",
        "XLOC_000545 NA\n",
        "XLOC_000562 PHD,Bromodomain,DDT\n",
        "XLOC_000566 NA\n",
        "XLOC_000583 NA\n",
        "XLOC_000584 NA\n",
        "XLOC_000592 NA\n",
        "XLOC_000605 NA\n",
        "XLOC_000606 NA\n",
        "XLOC_000608 NA\n",
        "XLOC_000610 NA\n",
        "XLOC_000618 PP2C,zf-CCCH,C3H\n",
        "XLOC_000620 NA\n",
        "XLOC_000625 NA\n",
        "XLOC_000632 NA\n",
        "XLOC_000635 PHD,Bromodomain,DDT\n",
        "XLOC_000637 NA\n",
        "XLOC_000638 NA\n",
        "XLOC_000640 NA\n",
        "XLOC_000641 NA\n",
        "XLOC_000644 NA\n",
        "XLOC_000663 NA\n",
        "XLOC_000696 NA\n",
        "XLOC_000719 NA\n",
        "XLOC_000724 NA\n",
        "XLOC_000727 NA\n",
        "XLOC_000741 NA\n",
        "XLOC_000742 NA\n",
        "XLOC_000757 NA\n",
        "XLOC_000759 NA\n",
        "XLOC_000776 NA\n",
        "XLOC_000799 NA\n",
        "XLOC_000809 NA\n",
        "XLOC_000825 NA\n",
        "XLOC_000832 NA\n",
        "XLOC_000840 NA\n",
        "XLOC_000893 NA\n",
        "XLOC_000896 NA\n",
        "XLOC_000898 NA\n",
        "XLOC_000909 NA\n",
        "XLOC_000922 NA\n",
        "XLOC_000934 NA\n",
        "XLOC_000936 NA\n",
        "XLOC_000944 NA\n",
        "XLOC_000953 NA\n",
        "XLOC_000957 NA\n",
        "XLOC_000962 NA\n",
        "XLOC_000968 NA\n",
        "XLOC_000969 NA\n",
        "XLOC_000974 NA\n",
        "XLOC_000989 NA\n",
        "XLOC_000994 WD40,Response_reg,Orphans\n",
        "XLOC_001024 NA\n",
        "XLOC_001037 NA\n",
        "XLOC_001044 NA\n",
        "XLOC_001061 NA\n",
        "XLOC_001081 NA\n",
        "XLOC_001083 NA\n",
        "XLOC_001092 NA\n",
        "XLOC_001102 NA\n",
        "XLOC_001103 NA\n",
        "XLOC_001106 NA\n",
        "XLOC_001116 NA\n",
        "XLOC_001123 NA\n",
        "XLOC_001128 NA\n",
        "XLOC_001129 NA\n",
        "XLOC_001130 NA\n",
        "XLOC_001145 NA\n",
        "XLOC_001155 NA\n",
        "XLOC_001174 NA\n",
        "XLOC_001184 NA\n",
        "XLOC_001191 NA\n",
        "XLOC_001195 NA\n",
        "XLOC_001197 NA\n",
        "XLOC_001219 NA\n",
        "XLOC_001234 NA\n",
        "XLOC_001248 NA\n",
        "XLOC_001254 NA\n",
        "XLOC_001274 NA\n",
        "XLOC_001278 NA\n",
        "XLOC_001287 NA\n",
        "XLOC_001291 NA\n",
        "XLOC_001298 NA\n",
        "XLOC_001303 NA\n",
        "XLOC_001312 NA\n",
        "XLOC_001318 NA\n",
        "XLOC_001321 NA\n",
        "XLOC_001325 NA\n",
        "XLOC_001329 NA\n",
        "XLOC_001332 NA\n",
        "XLOC_001337 NA\n",
        "XLOC_001339 NA\n",
        "XLOC_001340 NA\n",
        "XLOC_001342 NA\n",
        "XLOC_001343 NA\n",
        "XLOC_001347 NA\n",
        "XLOC_001352 zf-C3HC4,Helicase_C,SNF2_N,SNF2\n",
        "XLOC_001359 NA\n",
        "XLOC_001361 NA\n",
        "XLOC_001362 NA\n",
        "XLOC_001368 NA\n",
        "XLOC_001401 NA\n",
        "XLOC_001433 NA\n",
        "XLOC_001439 NA\n",
        "XLOC_001443 NA\n",
        "XLOC_001444 NA\n",
        "XLOC_001448 NA\n",
        "XLOC_001450 NA\n",
        "XLOC_001456 NA\n",
        "XLOC_001494 NA\n",
        "XLOC_001497 NA\n",
        "XLOC_001500 NA\n",
        "XLOC_001504 NA\n",
        "XLOC_001508 NA\n",
        "XLOC_001519 NA\n",
        "XLOC_001523 NA\n",
        "XLOC_001527 NA\n",
        "XLOC_001530 NA\n",
        "XLOC_001535 NA\n",
        "XLOC_001554 NA\n",
        "XLOC_001572 NA\n",
        "XLOC_001575 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin,Guanylate_cyc\n",
        "XLOC_001580 NA\n",
        "XLOC_001603 NA\n",
        "XLOC_001604 NA\n",
        "XLOC_001617 NA\n",
        "XLOC_001634 NA\n",
        "XLOC_001635 NA\n",
        "XLOC_001638 NA\n",
        "XLOC_001641 NA\n",
        "XLOC_001642 NA\n",
        "XLOC_001643 NA\n",
        "XLOC_001648 NA\n",
        "XLOC_001655 NA\n",
        "XLOC_001665 NA\n",
        "XLOC_001675 NA\n",
        "XLOC_001684 NA\n",
        "XLOC_001685 NA\n",
        "XLOC_001697 NA\n",
        "XLOC_001698 NA\n",
        "XLOC_001702 NA\n",
        "XLOC_001718 NA\n",
        "XLOC_001727 NA\n",
        "XLOC_001730 NA\n",
        "XLOC_001733 NA\n",
        "XLOC_001738 NA\n",
        "XLOC_001744 NA\n",
        "XLOC_001745 NA\n",
        "XLOC_001760 NA\n",
        "XLOC_001767 NA\n",
        "XLOC_001770 NA\n",
        "XLOC_001775 NA\n",
        "XLOC_001777 NA\n",
        "XLOC_001787 NA\n",
        "XLOC_001788 NA\n",
        "XLOC_001794 NA\n",
        "XLOC_001806 zf-CCCH,C3H\n",
        "XLOC_001816 NA\n",
        "XLOC_001820 NA\n",
        "XLOC_001824 NA\n",
        "XLOC_001826 NA\n",
        "XLOC_001832 NA\n",
        "XLOC_001837 NA\n",
        "XLOC_001842 NA\n",
        "XLOC_001851 NA\n",
        "XLOC_001859 NA\n",
        "XLOC_001875 AP2,AP2-EREBP\n",
        "XLOC_001885 NA\n",
        "XLOC_001888 NA\n",
        "XLOC_001904 NA\n",
        "XLOC_001906 NA\n",
        "XLOC_001912 NA\n",
        "XLOC_001921 NA\n",
        "XLOC_001922 NA\n",
        "XLOC_001924 NA\n",
        "XLOC_001926 NA\n",
        "XLOC_001928 NA\n",
        "XLOC_001939 NA\n",
        "XLOC_001954 NA\n",
        "XLOC_001964 NA\n",
        "XLOC_001978 NA\n",
        "XLOC_001982 NA\n",
        "XLOC_001988 NA\n",
        "XLOC_002009 NA\n",
        "XLOC_002016 NA\n",
        "XLOC_002027 PHD,Bromodomain,DDT\n",
        "XLOC_002032 NA\n",
        "XLOC_002033 NA\n",
        "XLOC_002034 NA\n",
        "XLOC_002035 NA\n",
        "XLOC_002043 NA\n",
        "XLOC_002046 NA\n",
        "XLOC_002048 NA\n",
        "XLOC_002056 PHD,Bromodomain,DDT\n",
        "XLOC_002102 NA\n",
        "XLOC_002113 NA\n",
        "XLOC_002114 NA\n",
        "XLOC_002118 NA\n",
        "XLOC_002127 NA\n",
        "XLOC_002141 NA\n",
        "XLOC_002149 NA\n",
        "XLOC_002159 NA\n",
        "XLOC_002170 NA\n",
        "XLOC_002173 NA\n",
        "XLOC_002188 NA\n",
        "XLOC_002196 NA\n",
        "XLOC_002200 NA\n",
        "XLOC_002202 NA\n",
        "XLOC_002209 NA\n",
        "XLOC_002213 NA\n",
        "XLOC_002216 NA\n",
        "XLOC_002218 NA\n",
        "XLOC_002237 NA\n",
        "XLOC_002239 zf-CCCH,C3H\n",
        "XLOC_002255 NA\n",
        "XLOC_002263 NA\n",
        "XLOC_002269 NA\n",
        "XLOC_002271 NA\n",
        "XLOC_002272 NA\n",
        "XLOC_002274 NA\n",
        "XLOC_002276 NA\n",
        "XLOC_002277 NA\n",
        "XLOC_002279 NA\n",
        "XLOC_002291 NA\n",
        "XLOC_002320 NA\n",
        "XLOC_002325 NA\n",
        "XLOC_002331 NA\n",
        "XLOC_002339 NA\n",
        "XLOC_002342 NA\n",
        "XLOC_002352 NA\n",
        "XLOC_002365 NA\n",
        "XLOC_002368 NA\n",
        "XLOC_002369 NA\n",
        "XLOC_002380 NA\n",
        "XLOC_002394 NA\n",
        "XLOC_002398 NA\n",
        "XLOC_002406 NA\n",
        "XLOC_002419 NA\n",
        "XLOC_002433 NA\n",
        "XLOC_002439 NA\n",
        "XLOC_002442 NA\n",
        "XLOC_002443 NA\n",
        "XLOC_002452 NA\n",
        "XLOC_002469 NA\n",
        "XLOC_002479 NA\n",
        "XLOC_002488 NA\n",
        "XLOC_002505 NA\n",
        "XLOC_002528 NA\n",
        "XLOC_002550 NA\n",
        "XLOC_002566 NA\n",
        "XLOC_002580 NA\n",
        "XLOC_002599 NA\n",
        "XLOC_002633 NA\n",
        "XLOC_002636 NA\n",
        "XLOC_002638 NA\n",
        "XLOC_002710 NA\n",
        "XLOC_002715 NA\n",
        "XLOC_002727 NA\n",
        "XLOC_002734 TRAF,BTB\n",
        "XLOC_002754 NA\n",
        "XLOC_002766 NA\n",
        "XLOC_002783 C2C2-GATA,Involucrin,GATA\n",
        "XLOC_002800 NA\n",
        "XLOC_002801 NA\n",
        "XLOC_002802 NA\n",
        "XLOC_002814 NA\n",
        "XLOC_002821 NA\n",
        "XLOC_002824 NA\n",
        "XLOC_002831 NA\n",
        "XLOC_002844 NA\n",
        "XLOC_002860 NA\n",
        "XLOC_002861 NA\n",
        "XLOC_002864 NA\n",
        "XLOC_002878 NA\n",
        "XLOC_002879 NA\n",
        "XLOC_002885 NA\n",
        "XLOC_002888 NA\n",
        "XLOC_002891 NA\n",
        "XLOC_002895 NA\n",
        "XLOC_002899 NA\n",
        "XLOC_002909 NA\n",
        "XLOC_002920 NA\n",
        "XLOC_002946 NA\n",
        "XLOC_002949 NA\n",
        "XLOC_002960 NA\n",
        "XLOC_002963 NA\n",
        "XLOC_002986 NA\n",
        "XLOC_003000 NA\n",
        "XLOC_003001 NA\n",
        "XLOC_003007 NA\n",
        "XLOC_003043 NA\n",
        "XLOC_003055 NA\n",
        "XLOC_003060 NA\n",
        "XLOC_003062 NA\n",
        "XLOC_003074 NA\n",
        "XLOC_003075 NA\n",
        "XLOC_003093 NA\n",
        "XLOC_003109 NA\n",
        "XLOC_003110 NA\n",
        "XLOC_003111 NA\n",
        "XLOC_003137 NA\n",
        "XLOC_003138 NA\n",
        "XLOC_003155 NA\n",
        "XLOC_003163 NA\n",
        "XLOC_003189 bZIP_2,bZIP\n",
        "XLOC_003202 NA\n",
        "XLOC_003220 NA\n",
        "XLOC_003231 NA\n",
        "XLOC_003236 NA\n",
        "XLOC_003247 HMG_box,SET,PHD\n",
        "XLOC_003254 NA\n",
        "XLOC_003267 NA\n",
        "XLOC_003272 NA\n",
        "XLOC_003296 NA\n",
        "XLOC_003319 NA\n",
        "XLOC_003320 PHD,Bromodomain,DDT\n",
        "XLOC_003325 NA\n",
        "XLOC_003328 NA\n",
        "XLOC_003335 NA\n",
        "XLOC_003341 NA\n",
        "XLOC_003345 NA\n",
        "XLOC_003365 NA\n",
        "XLOC_003369 NA\n",
        "XLOC_003372 NA\n",
        "XLOC_003383 NA\n",
        "XLOC_003393 NA\n",
        "XLOC_003400 NA\n",
        "XLOC_003401 NA\n",
        "XLOC_003422 NA\n",
        "XLOC_003428 NA\n",
        "XLOC_003429 NA\n",
        "XLOC_003432 NA\n",
        "XLOC_003447 NA\n",
        "XLOC_003456 NA\n",
        "XLOC_003457 NA\n",
        "XLOC_003509 NA\n",
        "XLOC_003526 NA\n",
        "XLOC_003536 NA\n",
        "XLOC_003551 NA\n",
        "XLOC_003558 NA\n",
        "XLOC_003562 NA\n",
        "XLOC_003564 NA\n",
        "XLOC_003575 NA\n",
        "XLOC_003582 NA\n",
        "XLOC_003614 NA\n",
        "XLOC_003616 NA\n",
        "XLOC_003622 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin,Guanylate_cyc\n",
        "XLOC_003624 NA\n",
        "XLOC_003632 NA\n",
        "XLOC_003633 NA\n",
        "XLOC_003636 NA\n",
        "XLOC_003642 NA\n",
        "XLOC_003698 NA\n",
        "XLOC_003700 NA\n",
        "XLOC_003702 NA\n",
        "XLOC_003720 NA\n",
        "XLOC_003721 NA\n",
        "XLOC_003732 NA\n",
        "XLOC_003737 NA\n",
        "XLOC_003741 NA\n",
        "XLOC_003747 NA\n",
        "XLOC_003750 NA\n",
        "XLOC_003754 NA\n",
        "XLOC_003762 NA\n",
        "XLOC_003771 NA\n",
        "XLOC_003773 NA\n",
        "XLOC_003779 NA\n",
        "XLOC_003784 NA\n",
        "XLOC_003822 NA\n",
        "XLOC_003823 NA\n",
        "XLOC_003840 NA\n",
        "XLOC_003853 NA\n",
        "XLOC_003866 NA\n",
        "XLOC_003871 NA\n",
        "XLOC_003873 NA\n",
        "XLOC_003884 NA\n",
        "XLOC_003889 NA\n",
        "XLOC_003891 NA\n",
        "XLOC_003896 NA\n",
        "XLOC_003913 NA\n",
        "XLOC_003915 PHD\n",
        "XLOC_003917 NA\n",
        "XLOC_003924 NA\n",
        "XLOC_003945 GNAT,Acetyltransf_1\n",
        "XLOC_003952 NA\n",
        "XLOC_003956 PHD,Bromodomain,DDT\n",
        "XLOC_003957 NA\n",
        "XLOC_003965 NA\n",
        "XLOC_003967 NA\n",
        "XLOC_003970 NA\n",
        "XLOC_003974 NA\n",
        "XLOC_003990 NA\n",
        "XLOC_003994 NA\n",
        "XLOC_004008 NA\n",
        "XLOC_004016 NA\n",
        "XLOC_004024 NA\n",
        "XLOC_004032 NA\n",
        "XLOC_004044 NA\n",
        "XLOC_004052 NA\n",
        "XLOC_004053 NA\n",
        "XLOC_004058 NA\n",
        "XLOC_004089 NA\n",
        "XLOC_004092 NA\n",
        "XLOC_004094 NA\n",
        "XLOC_004097 NA\n",
        "XLOC_004100 NA\n",
        "XLOC_004109 NA\n",
        "XLOC_004121 NA\n",
        "XLOC_004141 NA\n",
        "XLOC_004149 NA\n",
        "XLOC_004171 NA\n",
        "XLOC_004196 NA\n",
        "XLOC_004210 TRAF,BTB\n",
        "XLOC_004231 NA\n",
        "XLOC_004235 NA\n",
        "XLOC_004239 NA\n",
        "XLOC_004259 NA\n",
        "XLOC_004278 NA\n",
        "XLOC_004284 NA\n",
        "XLOC_004285 NA\n",
        "XLOC_004287 NA\n",
        "XLOC_004300 NA\n",
        "XLOC_004318 NA\n",
        "XLOC_004344 NA\n",
        "XLOC_004345 NA\n",
        "XLOC_004348 NA\n",
        "XLOC_004352 NA\n",
        "XLOC_004354 NA\n",
        "XLOC_004365 NA\n",
        "XLOC_004366 NA\n",
        "XLOC_004367 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin,Guanylate_cyc\n",
        "XLOC_004388 NA\n",
        "XLOC_004392 NA\n",
        "XLOC_004394 NA\n",
        "XLOC_004408 NA\n",
        "XLOC_004410 NA\n",
        "XLOC_004419 NA\n",
        "XLOC_004422 NA\n",
        "XLOC_004440 NA\n",
        "XLOC_004445 NA\n",
        "XLOC_004454 NA\n",
        "XLOC_004464 NA\n",
        "XLOC_004474 NA\n",
        "XLOC_004491 NA\n",
        "XLOC_004495 AP2,AP2-EREBP\n",
        "XLOC_004500 NA\n",
        "XLOC_004505 NA\n",
        "XLOC_004510 NA\n",
        "XLOC_004524 NA\n",
        "XLOC_004531 NA\n",
        "XLOC_004532 NA\n",
        "XLOC_004544 NA\n",
        "XLOC_004549 NA\n",
        "XLOC_004559 NA\n",
        "XLOC_004568 NA\n",
        "XLOC_004569 NA\n",
        "XLOC_004578 NA\n",
        "XLOC_004580 NA\n",
        "XLOC_004585 NA\n",
        "XLOC_004589 NA\n",
        "XLOC_004599 NA\n",
        "XLOC_004605 NA\n",
        "XLOC_004608 NA\n",
        "XLOC_004617 NA\n",
        "XLOC_004619 NA\n",
        "XLOC_004622 NA\n",
        "XLOC_004637 NA\n",
        "XLOC_004640 NA\n",
        "XLOC_004648 NA\n",
        "XLOC_004651 NA\n",
        "XLOC_004661 NA\n",
        "XLOC_004665 NA\n",
        "XLOC_004669 NA\n",
        "XLOC_004675 GNAT,Acetyltransf_1\n",
        "XLOC_004685 NA\n",
        "XLOC_004687 NA\n",
        "XLOC_004691 NA\n",
        "XLOC_004700 NA\n",
        "XLOC_004703 NA\n",
        "XLOC_004713 NA\n",
        "XLOC_004721 NA\n",
        "XLOC_004728 bHLH,HLH\n",
        "XLOC_004747 NA\n",
        "XLOC_004752 NA\n",
        "XLOC_004755 NA\n",
        "XLOC_004764 NA\n",
        "XLOC_004774 NA\n",
        "XLOC_004777 NA\n",
        "XLOC_004795 NA\n",
        "XLOC_004796 NA\n",
        "XLOC_004804 NA\n",
        "XLOC_004818 NA\n",
        "XLOC_004823 NA\n",
        "XLOC_004824 NA\n",
        "XLOC_004829 NA\n",
        "XLOC_004830 NA\n",
        "XLOC_004832 NA\n",
        "XLOC_004838 NA\n",
        "XLOC_004845 NA\n",
        "XLOC_004848 NA\n",
        "XLOC_004856 NA\n",
        "XLOC_004872 NA\n",
        "XLOC_004876 NA\n",
        "XLOC_004880 NA\n",
        "XLOC_004889 NA\n",
        "XLOC_004892 NA\n",
        "XLOC_004893 NA\n",
        "XLOC_004901 NA\n",
        "XLOC_004906 NA\n",
        "XLOC_004909 NA\n",
        "XLOC_004943 NA\n",
        "XLOC_004945 NA\n",
        "XLOC_004946 NA\n",
        "XLOC_004952 NA\n",
        "XLOC_004960 NA\n",
        "XLOC_004967 NA\n",
        "XLOC_004983 NA\n",
        "XLOC_004994 NA\n",
        "XLOC_004997 NA\n",
        "XLOC_005003 NA\n",
        "XLOC_005012 NA\n",
        "XLOC_005013 NA\n",
        "XLOC_005033 VARL\n",
        "XLOC_005057 NA\n",
        "XLOC_005061 PHD\n",
        "XLOC_005067 NA\n",
        "XLOC_005080 NA\n",
        "XLOC_005104 NA\n",
        "XLOC_005113 NA\n",
        "XLOC_005115 NA\n",
        "XLOC_005145 NA\n",
        "XLOC_005148 NA\n",
        "XLOC_005151 NA\n",
        "XLOC_005168 NA\n",
        "XLOC_005176 NA\n",
        "XLOC_005187 NA\n",
        "XLOC_005217 NA\n",
        "XLOC_005220 NA\n",
        "XLOC_005225 NA\n",
        "XLOC_005229 NA\n",
        "XLOC_005245 NA\n",
        "XLOC_005269 NA\n",
        "XLOC_005277 NA\n",
        "XLOC_005291 NA\n",
        "XLOC_005301 NA\n",
        "XLOC_005313 NA\n",
        "XLOC_005323 NA\n",
        "XLOC_005328 NA\n",
        "XLOC_005331 NA\n",
        "XLOC_005346 NA\n",
        "XLOC_005357 NA\n",
        "XLOC_005372 NA\n",
        "XLOC_005382 NA\n",
        "XLOC_005386 NA\n",
        "XLOC_005389 NA\n",
        "XLOC_005398 NA\n",
        "XLOC_005405 NA\n",
        "XLOC_005409 NA\n",
        "XLOC_005411 NA\n",
        "XLOC_005416 NA\n",
        "XLOC_005434 Response_reg,Orphans\n",
        "XLOC_005437 NA\n",
        "XLOC_005439 NA\n",
        "XLOC_005443 NA\n",
        "XLOC_005447 NA\n",
        "XLOC_005451 NA\n",
        "XLOC_005452 NA\n",
        "XLOC_005455 NA\n",
        "XLOC_005479 NA\n",
        "XLOC_005482 NA\n",
        "XLOC_005498 NA\n",
        "XLOC_005515 NA\n",
        "XLOC_005524 NA\n",
        "XLOC_005532 NA\n",
        "XLOC_005536 NA\n",
        "XLOC_005541 NA\n",
        "XLOC_005553 NA\n",
        "XLOC_005562 NA\n",
        "XLOC_005567 NA\n",
        "XLOC_005579 NA\n",
        "XLOC_005597 NA\n",
        "XLOC_005599 SBP\n",
        "XLOC_005620 NA\n",
        "XLOC_005625 NA\n",
        "XLOC_005628 NA\n",
        "XLOC_005632 NA\n",
        "XLOC_005633 NA\n",
        "XLOC_005635 NA\n",
        "XLOC_005650 NA\n",
        "XLOC_005660 NA\n",
        "XLOC_005668 NA\n",
        "XLOC_005669 NA\n",
        "XLOC_005672 NA\n",
        "XLOC_005678 NA\n",
        "XLOC_005684 NA\n",
        "XLOC_005685 NA\n",
        "XLOC_005695 NA\n",
        "XLOC_005709 NA\n",
        "XLOC_005729 NA\n",
        "XLOC_005742 GNAT,Acetyltransf_1\n",
        "XLOC_005746 NA\n",
        "XLOC_005751 NA\n",
        "XLOC_005759 NA\n",
        "XLOC_005774 NA\n",
        "XLOC_005793 AP2,AP2-EREBP\n",
        "XLOC_005824 NA\n",
        "XLOC_005851 AP2,AP2-EREBP\n",
        "XLOC_005857 NA\n",
        "XLOC_005881 NA\n",
        "XLOC_005899 NA\n",
        "XLOC_005900 NA\n",
        "XLOC_005901 NA\n",
        "XLOC_005906 NA\n",
        "XLOC_005952 NA\n",
        "XLOC_005957 NA\n",
        "XLOC_005968 NA\n",
        "XLOC_005969 NA\n",
        "XLOC_005974 NA\n",
        "XLOC_005986 NA\n",
        "XLOC_006000 NA\n",
        "XLOC_006010 NA\n",
        "XLOC_006024 NA\n",
        "XLOC_006027 NA\n",
        "XLOC_006032 NA\n",
        "XLOC_006048 NA\n",
        "XLOC_006052 NA\n",
        "XLOC_006056 NA\n",
        "XLOC_006060 NA\n",
        "XLOC_006062 NA\n",
        "XLOC_006070 NA\n",
        "XLOC_006072 NA\n",
        "XLOC_006099 NA\n",
        "XLOC_006108 NA\n",
        "XLOC_006112 NA\n",
        "XLOC_006116 NA\n",
        "XLOC_006122 NA\n",
        "XLOC_006131 NA\n",
        "XLOC_006138 NA\n",
        "XLOC_006140 NA\n",
        "XLOC_006153 SET\n",
        "XLOC_006155 NA\n",
        "XLOC_006164 NA\n",
        "XLOC_006168 NA\n",
        "XLOC_006180 NA\n",
        "XLOC_006203 NA\n",
        "XLOC_006204 NA\n",
        "XLOC_006215 NA\n",
        "XLOC_006227 NA\n",
        "XLOC_006233 NA\n",
        "XLOC_006235 NA\n",
        "XLOC_006237 NA\n",
        "XLOC_006240 NA\n",
        "XLOC_006241 NA\n",
        "XLOC_006275 NA\n",
        "XLOC_006282 NA\n",
        "XLOC_006308 NA\n",
        "XLOC_006310 NA\n",
        "XLOC_006318 NA\n",
        "XLOC_006322 NA\n",
        "XLOC_006326 NA\n",
        "XLOC_006329 NA\n",
        "XLOC_006334 NA\n",
        "XLOC_006338 NA\n",
        "XLOC_006362 NA\n",
        "XLOC_006368 NA\n",
        "XLOC_006379 NA\n",
        "XLOC_006388 NA\n",
        "XLOC_006390 NA\n",
        "XLOC_006397 NA\n",
        "XLOC_006404 NA\n",
        "XLOC_006405 NA\n",
        "XLOC_006408 NA\n",
        "XLOC_006444 NA\n",
        "XLOC_006460 NA\n",
        "XLOC_006476 NA\n",
        "XLOC_006480 NA\n",
        "XLOC_006521 NA\n",
        "XLOC_006529 NA\n",
        "XLOC_006537 NA\n",
        "XLOC_006542 NA\n",
        "XLOC_006568 G2-like\n",
        "XLOC_006571 NA\n",
        "XLOC_006572 NA\n",
        "XLOC_006582 Pericardin_rpt,SBP\n",
        "XLOC_006584 NA\n",
        "XLOC_006585 NA\n",
        "XLOC_006586 NA\n",
        "XLOC_006588 NA\n",
        "XLOC_006596 NA\n",
        "XLOC_006606 NA\n",
        "XLOC_006611 NA\n",
        "XLOC_006629 NA\n",
        "XLOC_006642 NA\n",
        "XLOC_006647 NA\n",
        "XLOC_006659 NA\n",
        "XLOC_006669 NA\n",
        "XLOC_006670 NA\n",
        "XLOC_006674 NA\n",
        "XLOC_006685 NA\n",
        "XLOC_006690 VARL\n",
        "XLOC_006724 NA\n",
        "XLOC_006730 Myb_DNA-binding,MYB\n",
        "XLOC_006742 NA\n",
        "XLOC_006747 NA\n",
        "XLOC_006750 PHD,Bromodomain,DDT\n",
        "XLOC_006752 NA\n",
        "XLOC_006753 NA\n",
        "XLOC_006758 NA\n",
        "XLOC_006777 NA\n",
        "XLOC_006780 NA\n",
        "XLOC_006788 NA\n",
        "XLOC_006822 NA\n",
        "XLOC_006833 NA\n",
        "XLOC_006846 NA\n",
        "XLOC_006857 NA\n",
        "XLOC_006867 NA\n",
        "XLOC_006868 NA\n",
        "XLOC_006872 NA\n",
        "XLOC_006876 NA\n",
        "XLOC_006878 NA\n",
        "XLOC_006883 PHD\n",
        "XLOC_006896 NA\n",
        "XLOC_006901 NA\n",
        "XLOC_006904 NA\n",
        "XLOC_006919 NA\n",
        "XLOC_006922 NA\n",
        "XLOC_006935 NA\n",
        "XLOC_006938 NA\n",
        "XLOC_006959 NA\n",
        "XLOC_006970 NA\n",
        "XLOC_006971 NA\n",
        "XLOC_006978 NA\n",
        "XLOC_006987 NA\n",
        "XLOC_006989 NA\n",
        "XLOC_006992 NA\n",
        "XLOC_007000 NA\n",
        "XLOC_007031 GNAT,Acetyltransf_1\n",
        "XLOC_007032 GNAT,Acetyltransf_1\n",
        "XLOC_007034 NA\n",
        "XLOC_007035 NA\n",
        "XLOC_007039 bZIP_2,bZIP\n",
        "XLOC_007041 NA\n",
        "XLOC_007060 NA\n",
        "XLOC_007069 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin,Guanylate_cyc\n",
        "XLOC_007073 NA\n",
        "XLOC_007083 NA\n",
        "XLOC_007091 NA\n",
        "XLOC_007106 NA\n",
        "XLOC_007111 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin\n",
        "XLOC_007129 NA\n",
        "XLOC_007133 FHA,DAO,FAD_binding_3\n",
        "XLOC_007154 SBP\n",
        "XLOC_007155 NA\n",
        "XLOC_007159 NA\n",
        "XLOC_007160 NA\n",
        "XLOC_007165 NA\n",
        "XLOC_007166 NA\n",
        "XLOC_007174 NA\n",
        "XLOC_007184 NA\n",
        "XLOC_007192 NA\n",
        "XLOC_007202 NA\n",
        "XLOC_007213 NA\n",
        "XLOC_007216 NA\n",
        "XLOC_007219 NA\n",
        "XLOC_007258 NA\n",
        "XLOC_007270 NA\n",
        "XLOC_007274 NA\n",
        "XLOC_007275 NA\n",
        "XLOC_007276 NA\n",
        "XLOC_007283 NA\n",
        "XLOC_007290 NA\n",
        "XLOC_007314 NA\n",
        "XLOC_007326 NA\n",
        "XLOC_007337 NA\n",
        "XLOC_007338 NA\n",
        "XLOC_007347 NA\n",
        "XLOC_007361 NA\n",
        "XLOC_007387 NA\n",
        "XLOC_007389 NA\n",
        "XLOC_007394 ACT,bHLH,HLH\n",
        "XLOC_007395 NA\n",
        "XLOC_007418 NA\n",
        "XLOC_007420 NA\n",
        "XLOC_007428 NA\n",
        "XLOC_007429 Myb_DNA-binding,MYB-related\n",
        "XLOC_007465 NA\n",
        "XLOC_007473 NA\n",
        "XLOC_007478 NA\n",
        "XLOC_007498 NA\n",
        "XLOC_007501 NA\n",
        "XLOC_007502 NA\n",
        "XLOC_007507 NA\n",
        "XLOC_007512 NA\n",
        "XLOC_007520 PHD,Bromodomain,DDT\n",
        "XLOC_007523 NA\n",
        "XLOC_007525 NA\n",
        "XLOC_007531 NA\n",
        "XLOC_007552 NA\n",
        "XLOC_007568 NA\n",
        "XLOC_007577 NA\n",
        "XLOC_007578 NA\n",
        "XLOC_007600 NA\n",
        "XLOC_007629 NA\n",
        "XLOC_007644 NA\n",
        "XLOC_007664 NA\n",
        "XLOC_007680 NA\n",
        "XLOC_007695 NA\n",
        "XLOC_007700 NA\n",
        "XLOC_007719 NA\n",
        "XLOC_007733 NA\n",
        "XLOC_007736 NA\n",
        "XLOC_007782 NA\n",
        "XLOC_007785 NA\n",
        "XLOC_007786 NA\n",
        "XLOC_007787 NA\n",
        "XLOC_007796 NA\n",
        "XLOC_007801 NA\n",
        "XLOC_007804 NA\n",
        "XLOC_007808 NA\n",
        "XLOC_007832 NA\n",
        "XLOC_007835 NA\n",
        "XLOC_007851 NA\n",
        "XLOC_007854 NA\n",
        "XLOC_007857 NA\n",
        "XLOC_007881 NA\n",
        "XLOC_007896 NA\n",
        "XLOC_007897 NA\n",
        "XLOC_007899 NA\n",
        "XLOC_007906 NA\n",
        "XLOC_007909 NA\n",
        "XLOC_007913 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin,Guanylate_cyc\n",
        "XLOC_007915 NA\n",
        "XLOC_007945 Myb_DNA-binding,MYB\n",
        "XLOC_007958 NA\n",
        "XLOC_007967 NA\n",
        "XLOC_007985 NA\n",
        "XLOC_007995 NA\n",
        "XLOC_007997 NA\n",
        "XLOC_008003 NA\n",
        "XLOC_008004 NA\n",
        "XLOC_008008 NA\n",
        "XLOC_008015 NA\n",
        "XLOC_008025 NA\n",
        "XLOC_008032 NA\n",
        "XLOC_008043 NA\n",
        "XLOC_008057 SAM_1,HisKA,Orphans,HATPase_c,Response_reg,Bac_rhodopsin,Guanylate_cyc\n",
        "XLOC_008069 NA\n",
        "XLOC_008072 NA\n",
        "XLOC_008082 NA\n",
        "XLOC_008096 NA\n",
        "XLOC_008101 NA\n",
        "XLOC_008112 NA\n",
        "XLOC_008123 NA\n",
        "XLOC_008126 NA\n",
        "XLOC_008134 NA\n",
        "XLOC_008139 NA\n",
        "XLOC_008141 NA\n",
        "XLOC_008146 NA\n",
        "XLOC_008148 NA\n",
        "XLOC_008149 NA\n",
        "XLOC_008150 NA\n",
        "XLOC_008171 NA\n",
        "XLOC_008198 NA\n",
        "XLOC_008214 NA\n",
        "XLOC_008222 NA\n",
        "XLOC_008254 NA\n",
        "XLOC_008257 NA\n",
        "XLOC_008286 NA\n",
        "XLOC_008308 NA\n",
        "XLOC_008313 NA\n",
        "XLOC_008317 NA\n",
        "XLOC_008325 NA\n",
        "XLOC_008327 NA\n",
        "XLOC_008340 NA\n",
        "XLOC_008342 NA\n",
        "XLOC_008373 NA\n",
        "XLOC_008376 NA\n",
        "XLOC_008387 NA\n",
        "XLOC_008389 NA\n",
        "XLOC_008401 NA\n",
        "XLOC_008420 NA\n",
        "XLOC_008423 NA\n",
        "XLOC_008435 NA\n",
        "XLOC_008442 NA\n",
        "XLOC_008465 NA\n",
        "XLOC_008475 NA\n",
        "XLOC_008482 NA\n",
        "XLOC_008495 NA\n",
        "XLOC_008500 NA\n",
        "XLOC_008508 NA\n",
        "XLOC_008516 NA\n",
        "XLOC_008532 NA\n",
        "XLOC_008545 NA\n",
        "XLOC_008552 NA\n",
        "XLOC_008580 NA\n",
        "XLOC_008584 NA\n",
        "XLOC_008593 NA\n",
        "XLOC_008616 NA\n",
        "XLOC_008622 NA\n",
        "XLOC_008623 NA\n",
        "XLOC_008632 NA\n",
        "XLOC_008636 NA\n",
        "XLOC_008652 NA\n",
        "XLOC_008673 NA\n",
        "XLOC_008681 NA\n",
        "XLOC_008682 NA\n",
        "XLOC_008696 Sigma70_r2,Orphans\n",
        "XLOC_008697 NA\n",
        "XLOC_008698 NA\n",
        "XLOC_008710 NA\n",
        "XLOC_008724 NA\n",
        "XLOC_008725 Myb_DNA-binding,MYB\n",
        "XLOC_008733 NA\n",
        "XLOC_008736 NA\n",
        "XLOC_008737 NA\n",
        "XLOC_008738 NA\n",
        "XLOC_008755 NA\n",
        "XLOC_008770 NA\n",
        "XLOC_008772 NA\n",
        "XLOC_008773 NA\n",
        "XLOC_008777 NA\n",
        "XLOC_008795 NA\n",
        "XLOC_008813 NA\n",
        "XLOC_008826 NA\n",
        "XLOC_008830 NA\n",
        "XLOC_008838 NA\n",
        "XLOC_008840 NA\n",
        "XLOC_008853 NA\n",
        "XLOC_008867 NA\n",
        "XLOC_008872 zf-C3HC4,Helicase_C,SNF2_N,SNF2\n",
        "XLOC_008880 NA\n",
        "XLOC_008883 zf-TAZ,TAZ\n",
        "XLOC_008917 NA\n",
        "XLOC_008921 NA\n",
        "XLOC_008945 NA\n",
        "XLOC_008955 NA\n",
        "XLOC_008960 NA\n",
        "XLOC_008965 RWP-RK\n",
        "XLOC_008966 NA\n",
        "XLOC_008996 RWP-RK\n",
        "XLOC_008998 NA\n",
        "XLOC_008999 NA\n",
        "XLOC_009019 NA\n",
        "XLOC_009020 NA\n",
        "XLOC_009026 NA\n",
        "XLOC_009032 NA\n",
        "XLOC_009051 NA\n",
        "XLOC_009072 NA\n",
        "XLOC_009077 NA\n",
        "XLOC_009082 NA\n",
        "XLOC_009083 NA\n",
        "XLOC_009089 NA\n",
        "XLOC_009100 NA\n",
        "XLOC_009102 NA\n",
        "XLOC_009111 NA\n",
        "XLOC_009146 NA\n",
        "XLOC_009163 NA\n",
        "XLOC_009165 NA\n",
        "XLOC_009166 NA\n",
        "XLOC_009193 NA\n",
        "XLOC_009201 NA\n",
        "XLOC_009213 NA\n",
        "XLOC_009221 TRAF,DUF6,BTB\n",
        "XLOC_009231 NA\n",
        "XLOC_009237 NA\n",
        "XLOC_009245 NA\n",
        "XLOC_009252 NA\n",
        "XLOC_009270 NA\n",
        "XLOC_009291 NA\n",
        "XLOC_009305 NA\n",
        "XLOC_009321 NA\n",
        "XLOC_009322 NA\n",
        "XLOC_009325 NA\n",
        "XLOC_009347 NA\n",
        "XLOC_009348 NA\n",
        "XLOC_009364 NA\n",
        "XLOC_009368 NA\n",
        "XLOC_009371 NA\n",
        "XLOC_009386 NA\n",
        "XLOC_009392 NA\n",
        "XLOC_009398 NA\n",
        "XLOC_009401 NA\n",
        "XLOC_009404 NA\n",
        "XLOC_009433 NA\n",
        "XLOC_009438 NA\n",
        "XLOC_009455 NA\n",
        "XLOC_009461 NA\n",
        "XLOC_009463 NA\n",
        "XLOC_009465 NA\n",
        "XLOC_009471 NA\n",
        "XLOC_009494 NA\n",
        "XLOC_009507 NA\n",
        "XLOC_009508 NA\n",
        "XLOC_009509 NA\n",
        "XLOC_009510 NA\n",
        "XLOC_009514 zf-TAZ,TAZ\n",
        "XLOC_009522 NA\n",
        "XLOC_009523 NA\n",
        "XLOC_009572 NA\n",
        "XLOC_009582 NA\n",
        "XLOC_009584 NA\n",
        "XLOC_009593 NA\n",
        "XLOC_009610 NA\n",
        "XLOC_009635 NA\n",
        "XLOC_009640 NA\n",
        "XLOC_009652 NA\n",
        "XLOC_009656 NA\n",
        "XLOC_009668 NA\n",
        "XLOC_009686 NF-YC,CCAAT\n",
        "XLOC_009698 NA\n",
        "XLOC_009704 NA\n",
        "XLOC_009705 NA\n",
        "XLOC_009706 NA\n",
        "XLOC_009708 NA\n",
        "XLOC_009710 NA\n",
        "XLOC_009711 NA\n",
        "XLOC_009720 NA\n",
        "XLOC_009742 NA\n",
        "XLOC_009746 NA\n",
        "XLOC_009753 NA\n",
        "XLOC_009755 NA\n",
        "XLOC_009775 NA\n",
        "XLOC_009795 ACT,bHLH,HLH\n",
        "XLOC_009806 NA\n",
        "XLOC_009808 NA\n",
        "XLOC_009819 NA\n",
        "XLOC_009826 NA\n",
        "XLOC_009827 TIG,G8\n",
        "XLOC_009881 NA\n",
        "XLOC_009888 NA\n",
        "XLOC_009897 NA\n",
        "XLOC_009904 NA\n",
        "XLOC_009907 NA\n",
        "XLOC_009932 NA\n",
        "XLOC_009937 PHD,Bromodomain,DDT\n",
        "XLOC_009959 NA\n",
        "XLOC_009976 NA\n",
        "XLOC_009985 NA\n",
        "XLOC_009987 NA\n",
        "XLOC_009997 NA\n",
        "XLOC_009998 NA\n",
        "XLOC_010000 NA\n",
        "XLOC_010003 NA\n",
        "XLOC_010004 NA\n",
        "XLOC_010007 NA"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "XLOC_010013 NA\n",
        "XLOC_010016 NA\n",
        "XLOC_010031 NA\n",
        "XLOC_010034 NA\n",
        "XLOC_010050 C2C2-GATA,GATA\n",
        "XLOC_010053 NA\n",
        "XLOC_010056 NA\n",
        "XLOC_010057 NA\n",
        "XLOC_010072 NA\n",
        "XLOC_010098 NA\n",
        "XLOC_010104 NA\n",
        "XLOC_010106 NA\n",
        "XLOC_010125 NA\n",
        "XLOC_010135 NA\n",
        "XLOC_010137 NA\n",
        "XLOC_010142 NA\n",
        "XLOC_010150 NA\n",
        "XLOC_010151 NA\n",
        "XLOC_010163 NA\n",
        "XLOC_010174 NA\n",
        "XLOC_010186 NA\n",
        "XLOC_010190 NA\n",
        "XLOC_010193 NA\n",
        "XLOC_010202 NA\n",
        "XLOC_010204 NA\n",
        "XLOC_010205 NA\n",
        "XLOC_010207 NA\n",
        "XLOC_010223 NA\n",
        "XLOC_010226 NA\n",
        "XLOC_010233 NA\n",
        "XLOC_010278 NA\n",
        "XLOC_010282 NA\n",
        "XLOC_010330 NA\n",
        "XLOC_010376 NA\n",
        "XLOC_010419 NA\n",
        "XLOC_010427 NA\n",
        "XLOC_010428 NA\n",
        "XLOC_010459 NA\n",
        "XLOC_010465 NA\n",
        "XLOC_010468 NA\n",
        "XLOC_010470 NA\n",
        "XLOC_010473 NA\n",
        "XLOC_010478 NA\n",
        "XLOC_010488 NA\n",
        "XLOC_010491 NA\n",
        "XLOC_010504 NA\n",
        "XLOC_010519 C2C2-GATA,GATA\n",
        "XLOC_010523 NA\n",
        "XLOC_010528 NA\n",
        "XLOC_010536 NA\n",
        "XLOC_010537 RRM_1,CSD\n",
        "XLOC_010539 NA\n",
        "XLOC_010542 GNAT,Acetyltransf_1\n",
        "XLOC_010551 NA\n",
        "XLOC_010554 NA\n",
        "XLOC_010562 NA\n",
        "XLOC_010604 NA\n",
        "XLOC_010612 NA\n",
        "XLOC_010617 NA\n",
        "XLOC_010621 NA\n",
        "XLOC_010624 NA\n",
        "XLOC_010625 NA\n",
        "XLOC_010630 NA\n",
        "XLOC_010636 NA\n",
        "XLOC_010637 NA\n",
        "XLOC_010638 NA\n",
        "XLOC_010643 CCT,zf-B_box,C2C2-CO-like\n",
        "XLOC_010673 NA\n",
        "XLOC_010675 NA\n",
        "XLOC_010684 NA\n",
        "XLOC_010687 NA\n",
        "XLOC_010736 NA\n",
        "XLOC_010738 NA\n",
        "XLOC_010750 NA\n",
        "XLOC_010759 NA\n",
        "XLOC_010763 RWP-RK\n",
        "XLOC_010779 NA\n",
        "XLOC_010781 NA\n",
        "XLOC_010788 NA\n",
        "XLOC_010792 NA\n",
        "XLOC_010798 NA\n",
        "XLOC_010824 NA\n",
        "XLOC_010834 NA\n",
        "XLOC_010836 NA\n",
        "XLOC_010838 NA\n",
        "XLOC_010840 NA\n",
        "XLOC_010843 Pericardin_rpt,TRAF,BTB\n",
        "XLOC_010849 NA\n",
        "XLOC_010853 NA\n",
        "XLOC_010883 NA\n",
        "XLOC_010887 NA\n",
        "XLOC_010888 NA\n",
        "XLOC_010891 NA\n",
        "XLOC_010895 NA\n",
        "XLOC_010901 NA\n",
        "XLOC_010912 NA\n",
        "XLOC_010913 NA\n",
        "XLOC_010919 NA\n",
        "XLOC_010924 NA\n",
        "XLOC_010933 NA\n",
        "XLOC_010937 NA\n",
        "XLOC_010938 NA\n",
        "XLOC_010942 NA\n",
        "XLOC_010946 NA\n",
        "XLOC_010966 NA\n",
        "XLOC_010976 NA\n",
        "XLOC_011008 NA\n",
        "XLOC_011057 NA\n",
        "XLOC_011061 NA\n",
        "XLOC_011071 NA\n",
        "XLOC_011074 NA\n",
        "XLOC_011085 NA\n",
        "XLOC_011101 NA\n",
        "XLOC_011105 NA\n",
        "XLOC_011110 NA\n",
        "XLOC_011135 NA\n",
        "XLOC_011140 NA\n",
        "XLOC_011154 NA\n",
        "XLOC_011155 NA\n",
        "XLOC_011156 NA\n",
        "XLOC_011177 NA\n",
        "XLOC_011189 NA\n",
        "XLOC_011216 NA\n",
        "XLOC_011225 NA\n",
        "XLOC_011250 TRAF,DUF6,BTB\n",
        "XLOC_011251 NA\n",
        "XLOC_011260 NA\n",
        "XLOC_011261 NA\n",
        "XLOC_011282 NA\n",
        "XLOC_011288 NA\n",
        "XLOC_011300 NA\n",
        "XLOC_011302 NA\n",
        "XLOC_011318 NA\n",
        "XLOC_011321 NA\n",
        "XLOC_011362 NA\n",
        "XLOC_011379 NA\n",
        "XLOC_011397 NA\n",
        "XLOC_011399 NA\n",
        "XLOC_011403 NA\n",
        "XLOC_011410 NA\n",
        "XLOC_011415 NA\n",
        "XLOC_011436 NA\n",
        "XLOC_011440 NA\n",
        "XLOC_011444 NA\n",
        "XLOC_011446 NA\n",
        "XLOC_011451 NA\n",
        "XLOC_011456 NA\n",
        "XLOC_011478 NA\n",
        "XLOC_011500 NA\n",
        "XLOC_011503 NA\n",
        "XLOC_011539 NA\n",
        "XLOC_011544 NA\n",
        "XLOC_011546 SBP\n",
        "XLOC_011565 NA\n",
        "XLOC_011580 NA\n",
        "XLOC_011583 NA\n",
        "XLOC_011591 NA\n",
        "XLOC_011593 NA\n",
        "XLOC_011595 NA\n",
        "XLOC_011599 NA\n",
        "XLOC_011602 NA\n",
        "XLOC_011611 NA\n",
        "XLOC_011637 NA\n",
        "XLOC_011649 NA\n",
        "XLOC_011663 NA\n",
        "XLOC_011664 NA\n",
        "XLOC_011679 NA\n",
        "XLOC_011692 NA\n",
        "XLOC_011718 NA\n",
        "XLOC_011731 NA\n",
        "XLOC_011738 NA\n",
        "XLOC_011739 NA\n",
        "XLOC_011745 E2F_TDP,E2F-DP,DP\n",
        "XLOC_011747 NA\n",
        "XLOC_011765 NA\n",
        "XLOC_011770 NA\n",
        "XLOC_011781 NA\n",
        "XLOC_011787 NA\n",
        "XLOC_011789 NA\n",
        "XLOC_011793 zf-CCCH,C3H\n",
        "XLOC_011803 NA\n",
        "XLOC_011806 NA\n",
        "XLOC_011840 NA\n",
        "XLOC_011869 NA\n",
        "XLOC_011870 NA\n",
        "XLOC_011882 NA\n",
        "XLOC_011890 NA\n",
        "XLOC_011891 NA\n",
        "XLOC_011892 NA\n",
        "XLOC_011894 NA\n",
        "XLOC_011898 NA\n",
        "XLOC_011904 NA\n",
        "XLOC_011913 NA\n",
        "XLOC_011922 NA\n",
        "XLOC_011924 NA\n",
        "XLOC_011926 NA\n",
        "XLOC_011932 NA\n",
        "XLOC_011948 NA\n",
        "XLOC_011956 NA\n",
        "XLOC_011959 NA\n",
        "XLOC_011975 NA\n",
        "XLOC_011979 NA\n",
        "XLOC_011981 NA\n",
        "XLOC_011987 NA\n",
        "XLOC_011993 NA\n",
        "XLOC_011995 NA\n",
        "XLOC_012006 NA\n",
        "XLOC_012011 NA\n",
        "XLOC_012015 NA\n",
        "XLOC_012028 NA\n",
        "XLOC_012030 NA\n",
        "XLOC_012089 NA\n",
        "XLOC_012114 NA\n",
        "XLOC_012122 NA\n",
        "XLOC_012131 NA\n",
        "XLOC_012137 C2C2-GATA,GATA\n",
        "XLOC_012158 NA\n",
        "XLOC_012163 SET\n",
        "XLOC_012166 NA\n",
        "XLOC_012188 NA\n",
        "XLOC_012191 NA\n",
        "XLOC_012210 NA\n",
        "XLOC_012214 NA\n",
        "XLOC_012226 NA\n",
        "XLOC_012236 NA\n",
        "XLOC_012237 NA\n",
        "XLOC_012243 NA\n",
        "XLOC_012250 NA\n",
        "XLOC_012251 NA\n",
        "XLOC_012293 SBP\n",
        "XLOC_012296 NA\n",
        "XLOC_012311 NA\n",
        "XLOC_012312 NA\n",
        "XLOC_012318 NA\n",
        "XLOC_012322 NA\n",
        "XLOC_012388 NA\n",
        "XLOC_012390 NA\n",
        "XLOC_012392 NA\n",
        "XLOC_012393 NA\n",
        "XLOC_012405 NA\n",
        "XLOC_012406 NA\n",
        "XLOC_012409 NA\n",
        "XLOC_012422 NA\n",
        "XLOC_012432 NA\n",
        "XLOC_012438 NA\n",
        "XLOC_012443 NA\n",
        "XLOC_012467 NA\n",
        "XLOC_012468 NA\n",
        "XLOC_012474 NA\n",
        "XLOC_012481 NA\n",
        "XLOC_012484 NA\n",
        "XLOC_012504 NA\n",
        "XLOC_012512 NA\n",
        "XLOC_012513 NA\n",
        "XLOC_012519 NA\n",
        "XLOC_012525 NA\n",
        "XLOC_012528 NA\n",
        "XLOC_012538 NA\n",
        "XLOC_012547 NA\n",
        "XLOC_012548 PHD,Bromodomain,DDT\n",
        "XLOC_012551 NA\n",
        "XLOC_012570 NA\n",
        "XLOC_012580 NA\n",
        "XLOC_012587 TFIIS,zf-CCCH,C3H\n",
        "XLOC_012589 NA\n",
        "XLOC_012614 NA\n",
        "XLOC_012643 zf-CCCH,C3H\n",
        "XLOC_012662 NA\n",
        "XLOC_012672 NA\n",
        "XLOC_012718 NA\n",
        "XLOC_012719 Helicase_C,Chromo,PHD\n",
        "XLOC_012723 NA\n",
        "XLOC_012733 C2C2-GATA,GATA\n",
        "XLOC_012739 NA\n",
        "XLOC_012745 NA\n",
        "XLOC_012752 NA\n",
        "XLOC_012756 NA\n",
        "XLOC_012757 NA\n",
        "XLOC_012759 NA\n",
        "XLOC_012807 NA\n",
        "XLOC_012808 NA\n",
        "XLOC_012830 NA\n",
        "XLOC_012841 NA\n",
        "XLOC_012846 NA\n",
        "XLOC_012851 NA\n",
        "XLOC_012857 CPP,CXC\n",
        "XLOC_012868 NA\n",
        "XLOC_012871 NA\n",
        "XLOC_012878 NA\n",
        "XLOC_012879 NA\n",
        "XLOC_012887 NA\n",
        "XLOC_012888 NA\n",
        "XLOC_012899 NA\n",
        "XLOC_012914 NA\n",
        "XLOC_012977 NA\n",
        "XLOC_012996 NA\n",
        "XLOC_013008 NA\n",
        "XLOC_013016 NA\n",
        "XLOC_013036 NA\n",
        "XLOC_013037 NA\n",
        "XLOC_013052 NA\n",
        "XLOC_013062 NA\n",
        "XLOC_013071 NA\n",
        "XLOC_013072 NA\n",
        "XLOC_013087 NA\n",
        "XLOC_013088 NA\n",
        "XLOC_013091 NA\n",
        "XLOC_013099 AP2,AP2-EREBP\n",
        "XLOC_013102 NA\n",
        "XLOC_013104 NA\n",
        "XLOC_013108 NA\n",
        "XLOC_013132 NA\n",
        "XLOC_013148 NA\n",
        "XLOC_013151 NA\n",
        "XLOC_013152 NA\n",
        "XLOC_013153 NA\n",
        "XLOC_013157 NA\n",
        "XLOC_013175 NA\n",
        "XLOC_013184 NA\n",
        "XLOC_013213 NA\n",
        "XLOC_013215 NA\n",
        "XLOC_013223 NA\n",
        "XLOC_013226 NA\n",
        "XLOC_013234 NA\n",
        "XLOC_013238 NA\n",
        "XLOC_013242 NA\n",
        "XLOC_013246 NA\n",
        "XLOC_013258 NA\n",
        "XLOC_013265 NA\n",
        "XLOC_013268 NA\n",
        "XLOC_013270 Helicase_C,SNF2_N,Agenet,SNF2\n",
        "XLOC_013278 NA\n",
        "XLOC_013280 NA\n",
        "XLOC_013283 NA\n",
        "XLOC_013286 NA\n",
        "XLOC_013287 NA\n",
        "XLOC_013290 NA\n",
        "XLOC_013297 NA\n",
        "XLOC_013301 NA\n",
        "XLOC_013309 NA\n",
        "XLOC_013317 NA\n",
        "XLOC_013318 NA\n",
        "XLOC_013332 NA\n",
        "XLOC_013339 NA\n",
        "XLOC_013340 NA\n",
        "XLOC_013342 NA\n",
        "XLOC_013343 NA\n",
        "XLOC_013348 NA\n",
        "XLOC_013352 NA\n",
        "XLOC_013357 NA\n",
        "XLOC_013369 NA\n",
        "XLOC_013371 NA\n",
        "XLOC_013377 NA\n",
        "XLOC_013378 NA\n",
        "XLOC_013380 NA\n",
        "XLOC_013382 NA\n",
        "XLOC_013387 NA\n",
        "XLOC_013392 NA\n",
        "XLOC_013402 NA\n",
        "XLOC_013407 NA\n",
        "XLOC_013411 NA\n",
        "XLOC_013412 NA\n",
        "XLOC_013431 NA\n",
        "XLOC_013463 NA\n",
        "XLOC_013467 NA\n",
        "XLOC_013469 NA\n",
        "XLOC_013478 NA\n",
        "XLOC_013492 NA\n",
        "XLOC_013496 NA\n",
        "XLOC_013497 NA\n",
        "XLOC_013515 NA\n",
        "XLOC_013520 NA\n",
        "XLOC_013531 NA\n",
        "XLOC_013532 Myb_DNA-binding,G2-like\n",
        "XLOC_013539 NA\n",
        "XLOC_013551 NA\n",
        "XLOC_013559 NA\n",
        "XLOC_013585 NA\n",
        "XLOC_013586 NA\n",
        "XLOC_013592 NA\n",
        "XLOC_013597 NA\n",
        "XLOC_013601 NA\n",
        "XLOC_013614 NA\n",
        "XLOC_013631 NA\n",
        "XLOC_013657 NA\n",
        "XLOC_013659 NA\n",
        "XLOC_013660 NA\n",
        "XLOC_013661 NA\n",
        "XLOC_013669 NA\n",
        "XLOC_013670 NA\n",
        "XLOC_013681 NA\n",
        "XLOC_013695 NA\n",
        "XLOC_013701 NA\n",
        "XLOC_013744 NA\n",
        "XLOC_013750 NA\n",
        "XLOC_013752 NA\n",
        "XLOC_013764 NA\n",
        "XLOC_013769 GNAT,Acetyltransf_1\n",
        "XLOC_013777 NA\n",
        "XLOC_013783 NA\n",
        "XLOC_013790 NA\n",
        "XLOC_013793 NA\n",
        "XLOC_013794 NA\n",
        "XLOC_013799 NA\n",
        "XLOC_013803 NA\n",
        "XLOC_013811 NA\n",
        "XLOC_013812 NA\n",
        "XLOC_013821 HSF,HSF_DNA-bind\n",
        "XLOC_013850 NA\n",
        "XLOC_013864 NA\n",
        "XLOC_013869 NA\n",
        "XLOC_013871 NA\n",
        "XLOC_013872 NA\n",
        "XLOC_013890 NA\n",
        "XLOC_013892 VARL\n",
        "XLOC_013898 NA\n",
        "XLOC_013903 NA\n",
        "XLOC_013906 NA\n",
        "XLOC_013907 NA\n",
        "XLOC_013917 NA\n",
        "XLOC_013918 NA\n",
        "XLOC_013919 NA\n",
        "XLOC_013933 NA\n",
        "XLOC_013941 NA\n",
        "XLOC_013945 NA\n",
        "XLOC_013948 PHD\n",
        "XLOC_013955 NA\n",
        "XLOC_013960 Myb_DNA-binding,MYB-related\n",
        "XLOC_013971 Jumonji,JmjC\n",
        "XLOC_013973 NA\n",
        "XLOC_013987 NA\n",
        "XLOC_013991 NA\n",
        "XLOC_014001 NA\n",
        "XLOC_014008 NA\n",
        "XLOC_014009 NA\n",
        "XLOC_014010 NA\n",
        "XLOC_014025 NA\n",
        "XLOC_014032 NA\n",
        "XLOC_014037 NA\n",
        "XLOC_014038 NA\n",
        "XLOC_014044 NA\n",
        "XLOC_014064 NA\n",
        "XLOC_014068 NA\n",
        "XLOC_014071 NA\n",
        "XLOC_014072 NA\n",
        "XLOC_014076 NA\n",
        "XLOC_014087 NA\n",
        "XLOC_014093 NA\n",
        "XLOC_014094 NA\n",
        "XLOC_014095 NA\n",
        "XLOC_014099 NA\n",
        "XLOC_014100 NA\n",
        "XLOC_014120 NA\n",
        "XLOC_014123 NA\n",
        "XLOC_014135 NA\n",
        "XLOC_014136 NA\n",
        "XLOC_014138 NA\n",
        "XLOC_014143 NA\n",
        "XLOC_014150 NA\n",
        "XLOC_014167 NA\n",
        "XLOC_014180 NA\n",
        "XLOC_014194 NA\n",
        "XLOC_014195 NA\n",
        "XLOC_014213 NA\n",
        "XLOC_014312 NA\n",
        "XLOC_014317 NA\n",
        "XLOC_014321 PHD,Bromodomain,DDT\n",
        "XLOC_014371 NA\n",
        "XLOC_014376 NA\n",
        "XLOC_014397 NA\n",
        "XLOC_014408 NA\n",
        "XLOC_014411 PHD,Bromodomain,DDT\n",
        "XLOC_014417 NA\n",
        "#cuff DE genes: 1607\n",
        "#up-regulated: 700\n",
        "#down-regulated: 907\n",
        "#novel DE genes: 31\n",
        "#DE genes with match to Chlamy Plant TFDB: 98\n",
        "#DE genes with match to Chlamy Plant TFDB and up reg: 49\n",
        "#DE genes with match to Chlamy Plant TFDB and down reg: 49\n",
        "#DE genes with NO phytozome annotation : 238\n"
       ]
      }
     ],
     "prompt_number": 60
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}