{
 "metadata": {
  "name": "Chlamy_data_analysis"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas\n",
      "import matplotlib.pyplot as plt\n",
      "sys.path.append('/global/homes/a/apratap/dev/eclipse_workspace/python_scripts/lib/');\n",
      "import utils\n",
      "import os\n",
      "import re"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "def create_data_array_from_mapped_files(file_list):\n",
      "    \n",
      "    SUB = 'create_data_array_from_files'\n",
      "    data = []\n",
      "    \n",
      "    for file_name in file_list:\n",
      "        with open(file_name, 'r') as fh:\n",
      "            file_basename = os.path.basename(file_name)\n",
      "            pattern='^.+?experiment\\/(N_.+?)_([AB]).+?accepted_hits_(.+?)\\.bam\\.alignstats$'\n",
      "            matches = re.search(pattern,file_name)\n",
      "            #print '[%s]: Reading data from file file %s' % (SUB,file_basename)\n",
      "            if matches:\n",
      "                #print matches.group(1)\n",
      "                #print matches.group(2)\n",
      "                #print matches.group(3)\n",
      "                #print '----------------------'\n",
      "                condition=matches.group(1)\n",
      "                replicate=matches.group(2)\n",
      "                strand = matches.group(3)\n",
      "            else:\n",
      "                print '[%s]:Warn: Could not find condition and replicate info'\n",
      "                condition = 'NA'\n",
      "                replicate = 'NA' \n",
      "            \n",
      "            \n",
      "            \n",
      "            for (line_num,line) in enumerate(fh):\n",
      "                \n",
      "                #ignore first 3 lines\n",
      "                if line_num <=2:\n",
      "                    continue\n",
      "                    \n",
      "                (variable,value) = line.strip().split()\n",
      "                if variable == '#primary':\n",
      "                    #print variable,value\n",
      "                    \n",
      "                    data.append([condition,replicate,strand,'primary',value])\n",
      "                    #pass\n",
      "        \n",
      "                if variable == '#secondary':\n",
      "                    #print variable,value\n",
      "                    data.append([condition,replicate,strand,'secondary',value])\n",
      "                    \n",
      "                    \n",
      "            #print data\n",
      "    return data\n",
      "    \n",
      "\n",
      "    \n",
      "def create_data_array_from_unmapped_files(file_list):\n",
      "    \n",
      "    SUB = 'create_data_array_from_unmapped_files'\n",
      "    data = []\n",
      "    \n",
      "    for file_name in file_list:\n",
      "        with open(file_name, 'r') as fh:\n",
      "            file_basename = os.path.basename(file_name)\n",
      "            pattern='^.+?experiment\\/(N_.+?)_([AB]).*$'\n",
      "            matches = re.search(pattern,file_name)\n",
      "            #print '[%s]: Reading data from file file %s' % (SUB,file_name)\n",
      "            if matches:\n",
      "                #print matches.group(1)\n",
      "                #print matches.group(2)\n",
      "                #print matches.group(3)\n",
      "                #print '----------------------'\n",
      "                condition=matches.group(1)\n",
      "                replicate=matches.group(2)\n",
      "                strand = 'NA'\n",
      "            else:\n",
      "                print '[%s]:Warn: Could not find condition and replicate info'\n",
      "                condition = 'NA'\n",
      "                replicate = 'NA' \n",
      "            \n",
      "            \n",
      "            \n",
      "            for (line_num,line) in enumerate(fh):\n",
      "                \n",
      "                #ignore first 3 lines\n",
      "                if line_num <=2:\n",
      "                    continue\n",
      "                    \n",
      "                (variable,value) = line.strip().split()\n",
      "         #       print line\n",
      "                if variable == '#unmapped_reads':\n",
      "          #          print variable,value\n",
      "                    data.append([condition,replicate,strand,'unmapped',value])\n",
      "        #print data_per_file\n",
      "    return data\n",
      "\n",
      "        \n",
      "        \n",
      "        \n",
      "#get the list of all the files to aggregate\n",
      "pattern_mapped_reads = 'accepted*sense*alignstats'\n",
      "bam_files_root_path = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment'\n",
      "\n",
      "\n",
      "files = utils.get_FilesList(bam_files_root_path,pattern=pattern_mapped_reads) \n",
      "data_mapped_reads = create_data_array_from_mapped_files(files)\n",
      "print 'Found %d items back' % (len(data_mapped_reads))\n",
      "\n",
      "\n",
      "\n",
      "pattern_unmapped_reads = 'unmapped*alignstats'\n",
      "bam_files_root_path = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment'\n",
      "\n",
      "\n",
      "files = utils.get_FilesList(bam_files_root_path,pattern=pattern_unmapped_reads) \n",
      "data_unmapped_reads = create_data_array_from_unmapped_files(files)\n",
      "print 'Found %d items back' % (len(data_unmapped_reads))\n",
      "\n",
      "\n",
      "\n",
      "#getting clean reads per sample\n",
      "\n",
      "\n",
      "raw_reads ='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/ERCC_mapping/raw_reads_per_sample.txt'\n",
      "\n",
      "data_raw_reads =[ [x.split()[0],x.split()[1],'NA','raw',x.split()[-1]] for (num,x) in enumerate(open(raw_reads,'r')) if num != 0]\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "data = data_mapped_reads + data_unmapped_reads + data_raw_reads\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "df = pandas.DataFrame(data,columns=['condition','replicate','strand','type','count'])\n",
      "\n",
      "\n",
      "print df\n",
      "mapping_summary ='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/ERCC_mapping/mapping_summary.tsv'\n",
      "df.to_csv(mapping_summary,sep='\\t',index=False)\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "[get_FilesList]: Found 36 files at /global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment\n",
        "Found 72 items back\n",
        "[get_FilesList]: Found 18 files at /global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/experiment"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Found 18 items back\n",
        "    condition replicate     strand       type      count\n",
        "0        N_0m         A  antisense    primary   62204046\n",
        "1        N_0m         A  antisense  secondary    7975140\n",
        "2        N_0m         A      sense    primary   61019168\n",
        "3        N_0m         A      sense  secondary   11333047\n",
        "4        N_0m         B      sense    primary   95292456\n",
        "5        N_0m         B      sense  secondary   13373556\n",
        "6        N_0m         B  antisense    primary   98772705\n",
        "7        N_0m         B  antisense  secondary   11621667\n",
        "8       N_10m         A      sense    primary   49608856\n",
        "9       N_10m         A      sense  secondary    9202513\n",
        "10      N_10m         A  antisense    primary   51154038\n",
        "11      N_10m         A  antisense  secondary    6774957\n",
        "12      N_10m         B  antisense    primary   46007007\n",
        "13      N_10m         B  antisense  secondary    5551739\n",
        "14      N_10m         B      sense    primary   44224315\n",
        "15      N_10m         B      sense  secondary    6446127\n",
        "16       N_1h         A      sense    primary   54616787\n",
        "17       N_1h         A      sense  secondary    9159943\n",
        "18       N_1h         A  antisense    primary   56714898\n",
        "19       N_1h         A  antisense  secondary    8471861\n",
        "20       N_1h         B  antisense    primary  138079868\n",
        "21       N_1h         B  antisense  secondary   23681824\n",
        "22       N_1h         B      sense    primary  134258833\n",
        "23       N_1h         B      sense  secondary   28062527\n",
        "24      N_24h         A  antisense    primary   56892492\n",
        "25      N_24h         A  antisense  secondary    7117464\n",
        "26      N_24h         A      sense    primary   52965023\n",
        "27      N_24h         A      sense  secondary    7817425\n",
        "28      N_24h         B  antisense    primary   48089160\n",
        "29      N_24h         B  antisense  secondary    5903421\n",
        "30      N_24h         B      sense    primary   45266617\n",
        "31      N_24h         B      sense  secondary    6386022\n",
        "32       N_2h         A      sense    primary   58712421\n",
        "33       N_2h         A      sense  secondary    9581136\n",
        "34       N_2h         A  antisense    primary   61705347\n",
        "35       N_2h         A  antisense  secondary    8660026\n",
        "36       N_2h         B  antisense    primary   70982335\n",
        "37       N_2h         B  antisense  secondary   11518735\n",
        "38       N_2h         B      sense    primary   67946934\n",
        "39       N_2h         B      sense  secondary   12474355\n",
        "40      N_30m         A      sense    primary   63085088\n",
        "41      N_30m         A      sense  secondary   10430735\n",
        "42      N_30m         A  antisense    primary   65225712\n",
        "43      N_30m         A  antisense  secondary    8865565\n",
        "44      N_30m         B      sense    primary   87083911\n",
        "45      N_30m         B      sense  secondary   12149313\n",
        "46      N_30m         B  antisense    primary   90778614\n",
        "47      N_30m         B  antisense  secondary   11351237\n",
        "48      N_48h         A      sense    primary   45694312\n",
        "49      N_48h         A      sense  secondary    6218094\n",
        "50      N_48h         A  antisense    primary   48676793\n",
        "51      N_48h         A  antisense  secondary    5479428\n",
        "52      N_48h         B      sense    primary   40783220\n",
        "53      N_48h         B      sense  secondary    5805254\n",
        "54      N_48h         B  antisense    primary   42976051\n",
        "55      N_48h         B  antisense  secondary    5178664\n",
        "56       N_6h         A  antisense    primary   62318002\n",
        "57       N_6h         A  antisense  secondary    8397774\n",
        "58       N_6h         A      sense    primary   57754671\n",
        "59       N_6h         A      sense  secondary   10234370\n",
        "60       N_6h         B      sense    primary   67813719\n",
        "61       N_6h         B      sense  secondary   13792577\n",
        "62       N_6h         B  antisense    primary   72929354\n",
        "63       N_6h         B  antisense  secondary   10974171\n",
        "64       N_8h         A      sense    primary   41844914\n",
        "65       N_8h         A      sense  secondary    6693241\n",
        "66       N_8h         A  antisense    primary   45359733\n",
        "67       N_8h         A  antisense  secondary    5371446\n",
        "68       N_8h         B  antisense    primary   48345895\n",
        "69       N_8h         B  antisense  secondary    6085482\n",
        "70       N_8h         B      sense    primary   44445628\n",
        "71       N_8h         B      sense  secondary    7334313\n",
        "72       N_0m         A         NA   unmapped    7291355\n",
        "73       N_0m         B         NA   unmapped    9564216\n",
        "74      N_10m         A         NA   unmapped    6552129\n",
        "75      N_10m         B         NA   unmapped    3876929\n",
        "76       N_1h         A         NA   unmapped    8108125\n",
        "77       N_1h         B         NA   unmapped   30212234\n",
        "78      N_24h         A         NA   unmapped    6310537\n",
        "79      N_24h         B         NA   unmapped    5343857\n",
        "80       N_2h         A         NA   unmapped    8125186\n",
        "81       N_2h         B         NA   unmapped   11182571\n",
        "82      N_30m         A         NA   unmapped    8737593\n",
        "83      N_30m         B         NA   unmapped    8928584\n",
        "84      N_48h         A         NA   unmapped    3715938\n",
        "85      N_48h         B         NA   unmapped    3785309\n",
        "86       N_6h         A         NA   unmapped    7726185\n",
        "87       N_6h         B         NA   unmapped    9148915\n",
        "88       N_8h         A         NA   unmapped    3878563\n",
        "89       N_8h         B         NA   unmapped    4087768\n",
        "90       N_0m         A         NA        raw  142845286\n",
        "91       N_0m         B         NA        raw  220941582\n",
        "92      N_10m         A         NA        raw  122929186\n",
        "93      N_10m         B         NA        raw   99819438\n",
        "94      N_30m         A         NA        raw  151539030\n",
        "95      N_30m         B         NA        raw  194029460\n",
        "96       N_1h         A         NA        raw  129109020\n",
        "97       N_1h         B         NA        raw  334231948\n",
        "98       N_2h         A         NA        raw  156011904\n",
        "99       N_2h         B         NA        raw  163852246\n",
        "100      N_6h         A         NA        raw  138511326\n",
        "101      N_6h         B         NA        raw  162069704\n",
        "102      N_8h         A         NA        raw   95650610\n",
        "103      N_8h         B         NA        raw  100741726\n",
        "104     N_24h         A         NA        raw  124011462\n",
        "105     N_24h         B         NA        raw  109200410\n",
        "106     N_48h         A         NA        raw  104185460\n",
        "107     N_48h         B         NA        raw   93643170\n"
       ]
      }
     ],
     "prompt_number": 128
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "c = '#raw_reads        72352215'\n",
      "print c.strip().split(q)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def create_cuff_gene_ref_gene_loopup_table():\n",
      "    \n",
      "    chlamy_lookup_table_cuff_gene_to_ref_gene = {}\n",
      "    for num,line in enumerate(open('/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/diff_expression/2_with_merged_gtf/cuff_genes_to_ref_gene_names.txt','r')):\n",
      "        line_split=line.strip().split('\\t')\n",
      "        chlamy_lookup_table_cuff_gene_to_ref_gene[line_split[0]] = line_split[1]\n",
      "        \n",
      "    return (chlamy_lookup_table_cuff_gene_to_ref_gene)\n",
      "\n",
      "\n",
      "\n",
      "def get_chlamy_gene_function_info():\n",
      "    #reading the function annotation\n",
      "    chlamy_func_annotation_file='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/Creinhardtii_v5.3_223_phytozome_functional_annotation.csv'\n",
      "    test_data='/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/test.data'\n",
      "\n",
      "    \n",
      "    chlamy_function_info = {}\n",
      "    for num,line in enumerate(open(chlamy_func_annotation_file,'r')):\n",
      "        if num == 0 :\n",
      "            continue\n",
      "        line_split = line.replace('\"','').strip().split('\\t')\n",
      "        \n",
      "        #skipping lines which only have gene name\n",
      "        if len(line_split) < 2:\n",
      "            #print line_split\n",
      "            continue\n",
      "        \n",
      "       #gene name \n",
      "        if line_split[0]:\n",
      "            gene_name = re.sub('\\.t.*','',line_split[0])\n",
      "       #gene common name\n",
      "        if len(line_split) >= 4 and line_split[3]:\n",
      "            gene_common_name=line_split[3]\n",
      "        else:\n",
      "            gene_common_name='NA' \n",
      "        #defline : col2\n",
      "        if line_split[1]:\n",
      "            gene_defline = line_split[1]\n",
      "        else:\n",
      "            gene_defline = 'NA'\n",
      "        #function : col14\n",
      "        \n",
      "        if len(line_split) >= 14 and line_split[13]:\n",
      "            gene_func = line_split[13]\n",
      "        else:\n",
      "            gene_func = 'NA'\n",
      "        \n",
      "        \n",
      "        #print \"%s \\t %s \\t %s \\t %s\" % (gene_name,gene_common_name,gene_defline,gene_func)\n",
      "        \n",
      "        chlamy_function_info[gene_name] = (gene_defline,gene_func)\n",
      "        chlamy_function_info[gene_common_name] = (gene_defline,gene_func)\n",
      "        \n",
      "    return chlamy_function_info\n",
      "\n",
      "\n",
      "\n",
      "def read_cuff_diff_gene_exp_data(cuff_diff_gene_file,base_condition,test_condition):\n",
      "    \n",
      "    cuff_gene_fold_change = {}\n",
      "    for line in open(cuff_diff_gene_file,'r'):\n",
      "        line_split=line.strip().split('\\t')\n",
      "        \n",
      "        if line_split[4].strip() == base_condition and line_split[5].strip() == test_condition:\n",
      "            gene_name = line_split[0].strip()\n",
      "            gene_fold_change = line_split[9].strip()\n",
      "            print gene_name,gene_fold_change\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import sys\n",
      "import re\n",
      "\n",
      "\n",
      "#get the \n",
      "dir()\n",
      "\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 85,
       "text": [
        "['DE_gene_list_0_30m',\n",
        " 'DE_gene_list_0m_10m',\n",
        " 'In',\n",
        " 'Out',\n",
        " '_',\n",
        " '_42',\n",
        " '_81',\n",
        " '_82',\n",
        " '_83',\n",
        " '__',\n",
        " '___',\n",
        " '__builtin__',\n",
        " '__builtins__',\n",
        " '__name__',\n",
        " '__package__',\n",
        " '_dh',\n",
        " '_i',\n",
        " '_i1',\n",
        " '_i10',\n",
        " '_i11',\n",
        " '_i12',\n",
        " '_i13',\n",
        " '_i14',\n",
        " '_i15',\n",
        " '_i16',\n",
        " '_i17',\n",
        " '_i18',\n",
        " '_i19',\n",
        " '_i2',\n",
        " '_i20',\n",
        " '_i21',\n",
        " '_i22',\n",
        " '_i23',\n",
        " '_i24',\n",
        " '_i25',\n",
        " '_i26',\n",
        " '_i27',\n",
        " '_i28',\n",
        " '_i29',\n",
        " '_i3',\n",
        " '_i30',\n",
        " '_i31',\n",
        " '_i32',\n",
        " '_i33',\n",
        " '_i34',\n",
        " '_i35',\n",
        " '_i36',\n",
        " '_i37',\n",
        " '_i38',\n",
        " '_i39',\n",
        " '_i4',\n",
        " '_i40',\n",
        " '_i41',\n",
        " '_i42',\n",
        " '_i43',\n",
        " '_i44',\n",
        " '_i45',\n",
        " '_i46',\n",
        " '_i47',\n",
        " '_i48',\n",
        " '_i49',\n",
        " '_i5',\n",
        " '_i50',\n",
        " '_i51',\n",
        " '_i52',\n",
        " '_i53',\n",
        " '_i54',\n",
        " '_i55',\n",
        " '_i56',\n",
        " '_i57',\n",
        " '_i58',\n",
        " '_i59',\n",
        " '_i6',\n",
        " '_i60',\n",
        " '_i61',\n",
        " '_i62',\n",
        " '_i63',\n",
        " '_i64',\n",
        " '_i65',\n",
        " '_i66',\n",
        " '_i67',\n",
        " '_i68',\n",
        " '_i69',\n",
        " '_i7',\n",
        " '_i70',\n",
        " '_i71',\n",
        " '_i72',\n",
        " '_i73',\n",
        " '_i74',\n",
        " '_i75',\n",
        " '_i76',\n",
        " '_i77',\n",
        " '_i78',\n",
        " '_i79',\n",
        " '_i8',\n",
        " '_i80',\n",
        " '_i81',\n",
        " '_i82',\n",
        " '_i83',\n",
        " '_i84',\n",
        " '_i85',\n",
        " '_i9',\n",
        " '_ih',\n",
        " '_ii',\n",
        " '_iii',\n",
        " '_oh',\n",
        " '_sh',\n",
        " 'chlamy_func_annotation',\n",
        " 'chlamy_func_annotation_file',\n",
        " 'chlamy_function_info',\n",
        " 'chlamy_gene_function',\n",
        " 'create_cuff_gene_ref_gene_loopup_table',\n",
        " 'cuff_gene_diff_exp_file',\n",
        " 'cuff_gene_name',\n",
        " 'defline',\n",
        " 'exit',\n",
        " 'fout',\n",
        " 'function',\n",
        " 'gene_common_name',\n",
        " 'gene_defline',\n",
        " 'gene_func',\n",
        " 'gene_function',\n",
        " 'gene_list',\n",
        " 'gene_name',\n",
        " 'get_chlamy_gene_function_info',\n",
        " 'get_ipython',\n",
        " 'help',\n",
        " 'line',\n",
        " 'line_split',\n",
        " 'lookup_table_cuff_gene_to_chlamy_gene',\n",
        " 'lookup_table_cuff_gene_to_ref_gene',\n",
        " 'num',\n",
        " 'quit',\n",
        " 're',\n",
        " 'read_cuff_diff_gene_exp_data',\n",
        " 'ref_gene_function_info',\n",
        " 'ref_gene_name',\n",
        " 'sys',\n",
        " 'test_data']"
       ]
      }
     ],
     "prompt_number": 85
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print lookup_table_cuff_gene_to_ref_gene.get('XLOC_000035')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        " g2\n"
       ]
      }
     ],
     "prompt_number": 59
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print len(lookup_table_cuff_gene_to_chlamy_gene)\n",
      "print len(ref_gene_function_info)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "14419\n",
        "19034\n"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print chlamy_function_info"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{'Cre01.g000150': ('Zinc-nutrition responsive permease transporter', 'zinc transporter 7 precursor'), 'Cre01.g000050': ('RWP-RK transcription factor', 'RWP-RK domain-containing protein'), 'RWP14': ('RWP-RK transcription factor', 'RWP-RK domain-containing protein'), 'Cre01.g000200': ('NA', 'NA'), 'Cre01.g000100': ('NA', 'NA'), 'ZRT2': ('Zinc-nutrition responsive permease transporter', 'zinc transporter 7 precursor'), 'NA': ('NA', 'NA')}\n"
       ]
      }
     ],
     "prompt_number": 27
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "create_cuff_gene_ref_gene_loopup_table"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 86,
       "text": [
        "<function __main__.create_cuff_gene_ref_gene_loopup_table>"
       ]
      }
     ],
     "prompt_number": 86
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import inspect"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 87
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print inspect.getsource(read_cuff_diff_gene_exp_data)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "def read_cuff_diff_gene_exp_data(cuff_diff_gene_file,base_condition,test_condition):\n",
        "    \n",
        "    cuff_gene_fold_change = {}\n",
        "    for line in open(cuff_diff_gene_file,'r'):\n",
        "        line_split=line.strip().split('\\t')\n",
        "        \n",
        "        if line_split[4].strip() == base_condition and line_split[5].strip() == test_condition:\n",
        "            gene_name = line_split[0].strip()\n",
        "            gene_fold_change = line_split[9].strip()\n",
        "            print gene_name,gene_fold_change\n",
        "\n"
       ]
      }
     ],
     "prompt_number": 90
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}