{
 "metadata": {
  "name": "chlamy_Nitro_data_analysis"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pybedtools\n",
      "import sys\n",
      "import multiprocessing\n",
      "import cStringIO\n",
      "import tempfile\n",
      "import matplotlib.pyplot as plt\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#a function to pass only features of a particular featuretype\n",
      "#this is similar to grep\n",
      "# if inverse_flag is set to True, this is similar to grep -v\n",
      "def featuretype_filter(feature,featuretypes,inverse_flag=False):\n",
      "    for featuretype in featuretypes:\n",
      "        if feature[2] == featuretype:\n",
      "            #match found\n",
      "            if inverse_flag is True:\n",
      "                return False\n",
      "            else:\n",
      "                return True\n",
      "            \n",
      "    #no match found\n",
      "    if inverse_flag is True:\n",
      "        return True\n",
      "    else:\n",
      "        return False\n",
      "\n",
      "\n",
      "#function to subset a feature file selecting the features \n",
      "def subset_featuretypes(bedToolsFile,featuretypes,inverse_flag=False):\n",
      "    result = bedToolsFile.filter(featuretype_filter,featuretypes,inverse_flag=inverse_flag).saveas()\n",
      "    return pybedtools.BedTool(result.fn)\n",
      "\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def parse_gtf_file(gtf_file):\n",
      "    \"\"\"\n",
      "    given a gtf file as input this function will parse and return a single line\n",
      "    is a generator : yield single parsed gtf entry one at a time\n",
      "    \"\"\"\n",
      "    \n",
      "    #sample_gtf_line (split in 3 lines for display purposes)\n",
      "    #chromosome_1    Cufflinks       exon    20356   20687   .       +       .       \n",
      "    #gene_id \"XLOC_000036\"; transcript_id \"TCONS_00000050\"; exon_number \"1\"; gene_name \"g3\"; oId \n",
      "    #\"PAC:26903463\"; contained_in \"TCONS_00000051\"; nearest_ref \"PAC:26903463\"; class_code \"=\"; tss_id \"TSS41\"; p_id \"P3\";\n",
      "\n",
      "    ##just creating an abtract class/container to hold the values for features of a gtf line\n",
      "    class gtf_line_container(object):\n",
      "        \n",
      "        def __init__(self):\n",
      "            self.chr=None\n",
      "            self.start=None\n",
      "            self.end=None\n",
      "            self.transcript_id=None\n",
      "            self.gene_id=None\n",
      "            self.feature_type=None\n",
      "            self.attributes=None\n",
      "    \n",
      "    for num,line in enumerate(open(gtf_file,'r')):\n",
      "        \n",
      "        line_split = line.strip().split('\\t')\n",
      "        \n",
      "        #instantiate the container\n",
      "        gtf_line = gtf_line_container()\n",
      "        \n",
      "        attributes_split=line_split[8].split(';')\n",
      "        #sample attribute split line\n",
      "        #['gene_id \"XLOC_000001\"', ' transcript_id \"TCONS_00000001\"', ' exon_number \"1\"', ' oId \"CUFF.31.1\"', ' class_code \"u\"', ' tss_id \"TSS1\"', '']\n",
      "        \n",
      "        #gene_id\n",
      "        gene_id=attributes_split[0].split(' ')[1].replace('\"','')\n",
      "        \n",
      "        #transcrpit_id\n",
      "        transcript_id=attributes_split[1].split(' ')[2].replace('\"','')\n",
      "        \n",
      "        \n",
      "        gtf_line.chr=line_split[0].strip()\n",
      "        gtf_line.start=line_split[3].strip()\n",
      "        gtf_line.end=line_split[4].strip()\n",
      "        gtf_line.strand=line_split[6].strip()\n",
      "        gtf_line.feature_type=line_split[2].strip()\n",
      "        gtf_line.gene_id = gene_id\n",
      "        gtf_line.transcript_id = transcript_id\n",
      "        gtf_line.attributes = line_split[8].strip()\n",
      "        yield gtf_line\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "def summarize_cuffmerge_output(merged_gtf_file,summarize_by,outFile=None):\n",
      "    '''\n",
      "    given a merged.gtf file produced by cuffmerge this method can\n",
      "    summarize it by gene or transcript level\n",
      "    **PS: assumption is that for each gene the exons are reported, sorted by position**\n",
      "    variable : summarize_by can take the value of gene,transcript\n",
      "    '''\n",
      "\n",
      "    features = {}\n",
      "    class feature(object):\n",
      "        def __init__(self):\n",
      "            self.gene_id = None\n",
      "            self.transcript_id = None\n",
      "            self.start = None\n",
      "            self.end = None\n",
      "            self.strand = None\n",
      "            self.exon_starts = []\n",
      "            self.exon_ends = []\n",
      "            self.attributes = None\n",
      "            \n",
      "    #default for each feature\n",
      "    last_feature_id = None\n",
      "    feature_start = 0;\n",
      "    feature_chromosome = 'NA';\n",
      "    feature_end = 0;\n",
      "    feature_strand = 0;\n",
      "    feature_attributes='NA'\n",
      "    \n",
      "    \n",
      "    #setting the right feture name\n",
      "    if summarize_by == 'gene':\n",
      "        feature_name = 'gene_id'\n",
      "    elif summarize_by == 'transcript':\n",
      "        feature_name = 'transcript_id'\n",
      "    else: \n",
      "        sys.stderr('Error : not a recognized summarize_by id %s ' % summarize_by)\n",
      "        sys.exit(2)\n",
      "    \n",
      "    \n",
      "    for gtf_line in parse_gtf_file(merged_gtf_file):\n",
      "            #interesting peice of logic\n",
      "            #depending on the user entered summarize_by varibale either the transcript_id or gene_id is used\n",
      "            this_feature_id=eval('gtf_line.%s' % feature_name)\n",
      "        \n",
      "            #initialization\n",
      "            if ( last_feature_id is None):\n",
      "                last_feature_id = this_feature_id\n",
      "                feature_start = gtf_line.start\n",
      "                feature_end   = gtf_line.end\n",
      "                feature_strand = gtf_line.strand\n",
      "                feature_chromosome = gtf_line.chr\n",
      "                feature_attributes = gtf_line.attributes\n",
      "                continue\n",
      "            \n",
      "            #if we see the same feature, just update the end\n",
      "            if ( last_feature_id == this_feature_id):\n",
      "                feature_end   = gtf_line.end\n",
      "                feature_attributes = gtf_line.attributes\n",
      "                \n",
      "            else:\n",
      "                #create the new feature summary object before updating to the new feature\n",
      "                feature_summary = feature()\n",
      "                feature_summary.feature_id = last_feature_id\n",
      "                feature_summary.start = feature_start\n",
      "                feature_summary.end = feature_end\n",
      "                feature_summary.strand = feature_strand\n",
      "                feature_summary.chr = feature_chromosome\n",
      "                feature_summary.attributes = feature_attributes\n",
      "                \n",
      "                #store the feature summary object in a dict\n",
      "                features[last_feature_id] = feature_summary\n",
      "                \n",
      "                ##now update and store the new feature\n",
      "                last_feature_id = this_feature_id\n",
      "                feature_start = gtf_line.start\n",
      "                feature_end   = gtf_line.end\n",
      "                feature_strand = gtf_line.strand\n",
      "                feature_chromosome = gtf_line.chr\n",
      "                feature_attributes = gtf_line.attributes\n",
      "    \n",
      "    \n",
      "    feature_summary = feature()\n",
      "    feature_summary.feature_id = last_feature_id\n",
      "    feature_summary.start = feature_start\n",
      "    feature_summary.end = feature_end\n",
      "    feature_summary.strand = feature_strand\n",
      "    feature_summary.chr = feature_chromosome\n",
      "    feature_summary.attributes = feature_attributes\n",
      "                \n",
      "    #store the feature summary object in a dict\n",
      "    features[last_feature_id] = feature_summary\n",
      "    \n",
      "    #write the summarized file if needed\n",
      "    if outFile is not None:\n",
      "        fout = open(outFile,'w')\n",
      "        for each_feature_id in features:\n",
      "            feature = features[each_feature_id]\n",
      "            gtf_line ='%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n' %  (feature.chr,'cufflinks',summarize_by,feature.start,feature.end,\n",
      "                                                                               '.',feature.strand,'.',feature.attributes)\n",
      "            fout.write(gtf_line)\n",
      "        print 'Created file %s summarized by %s of len %d' % (outFile,summarize_by,len(features))\n",
      "        fout.close()\n",
      "        return (features)\n",
      "        \n",
      "    print 'Created features dict, summarized by %s of len %d' % (summarize_by,len(features))\n",
      "    return (features)\n",
      " "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#ref chlamy v5 transcriptome\n",
      "chlamy_ref_transcriptome = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/annotation/Chlamy_v5/Creinhardtii_v5.3_223_gene.gff3'\n",
      "\n",
      "#Chlamy v5 chr sizes\n",
      "chlamy_chr_sizes = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/reference/Chlamy_V5/C_reinhardtii.110311_Chlroplast_mitochondira/chlamy_v5_chr_sizes.txt'\n",
      "\n",
      "\n",
      "#rna_seq_assembly\n",
      "rna_seq_assembly_file = '/global/projectb/projectdirs/PI/Chlamy_Chia_Lin/Chlamy_Nitro_deprivation_Project/Chlamydomonas_reinhardtii_4A+_404151/assembly/assembly_CWong_modified_cufflinks/merged.gtf'\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#RNA-Seq genes\n",
      "temp_rna_seq_genes_fh = tempfile.NamedTemporaryFile(delete=False)\n",
      "rna_seq_genes = summarize_cuffmerge_output(rna_seq_assembly_file,summarize_by='gene',outFile=temp_rna_seq_genes_fh.name)\n",
      "rna_seq_genes = pybedtools.BedTool(temp_rna_seq_genes_fh.name)\n",
      "\n",
      "#RNA-Seq Transcripts\n",
      "temp_rna_seq_transcripts_fh = tempfile.NamedTemporaryFile(delete=False)\n",
      "rna_seq_transcripts = summarize_cuffmerge_output(rna_seq_assembly_file,summarize_by='transcript',outFile=temp_rna_seq_transcripts_fh.name)\n",
      "rna_seq_transcripts = pybedtools.BedTool(temp_rna_seq_transcripts_fh.name)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#available ref trasncriptome\n",
      "cleaned_ref_transcriptome = pybedtools.BedTool(chlamy_ref_transcriptome).remove_invalid().saveas()\n",
      "\n",
      "#cufflinks completed RNA-Seq assembly\n",
      "rna_seq_assembly = pybedtools.BedTool(rna_seq_assembly_file)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#pool = multiprocessing.Pool()\n",
      "#featuretypes = ( (cleaned_ref_transcriptome,'gene'),(cleaned_ref_transcriptome,'mRNA'))\n",
      "#genes,mRNA = pool.map(subset_featuretypes,featuretype)\n",
      "\n",
      "#ref genes\n",
      "ref_genes = subset_featuretypes(cleaned_ref_transcriptome,['gene']).saveas()\n",
      "#ref transcripts\n",
      "ref_mRNA = subset_featuretypes(cleaned_ref_transcriptome,['mRNA']).saveas()\n",
      "\n",
      "#everything except mRNA and genes feature types\n",
      "ref_exons = subset_featuretypes(cleaned_ref_transcriptome,['gene','mRNA'],inverse_flag=True).saveas()\n",
      "\n",
      "#exons in RNA-Seq model\n",
      "rna_seq_exons = subset_featuretypes(rna_seq_assembly,['exon']).saveas()\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#number of novel genes predicted by RNA-Seq assembly\n",
      "rna_seq_novel_genes = rna_seq_genes.intersect(ref_genes, v=True).saveas()\n",
      "\n",
      "#number of novel antisense genes\n",
      "rna_seq_antisense_novel_genes = rna_seq_genes.intersect(ref_genes, S=True).saveas()\n",
      "\n",
      "#RNA-Seq novel exons\n",
      "rna_seq_novel_exons = rna_seq_exons.intersect(ref_exons,v=True).saveas()\n",
      "\n",
      "#RNA-Seq novel exons inside ref genes\n",
      "rna_seq_novel_exons_inside_ref_genes = rna_seq_novel_exons.intersect(ref_genes).saveas()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#finding the mergers and extensions\n",
      "\n",
      "## find the complement of ref genes\n",
      "ref_genes_complement=ref_genes.complement(g=chlamy_chr_sizes).saveas()\n",
      "##see how many RNA_Seq genes interesect with ref_gene complement \n",
      "temp=rna_seq_genes.intersect(ref_genes_complement,u=True).saveas()\n",
      "## do the coverage of rna_seq genes that intersect with ref_genes\n",
      "## count = 1 should be extensions\n",
      "## count > 1 should be mergers\n",
      "count_rna_seq_genes_with_extension_N_mergers = ref_genes.coverage(temp,s=True,counts=True).saveas()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 200
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print (temp.fn)\n",
      "print len(temp)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/projectb/scratch/apratap/pybedtools.nVvKth.tmp\n",
        "7572"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 196
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "counts_mergers_extensions = [ int(line[9]) for line in count_rna_seq_genes_with_extension_N_mergers ]\n",
      "print len(counts)\n",
      "plt.hist(counts)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "7572\n"
       ]
      }
     ],
     "prompt_number": 201
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}