{
 "metadata": {
  "name": "JGI_runs_quality_summaries"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pymysql\n",
      "conn = pymysql.connect(host='redwood.jgi-psf.org',user='jazz',passwd='jazz234',db='rqc')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cur = conn.cursor()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cur.execute(\"SELECT * from seq_units limit 10\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cur.execute(\"select run_id,seq_unit_name,run_date from seq_units where run_date >= '2012-09-25' AND run_date < '2012-11-26'\")\n",
      "\n",
      "for r in cur.fetchall():\n",
      "    print r"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "#cur.execute(\"\"\" select su.seq_unit_name, su.run_barcode, su.run_id, su.run_date,su.run_section,m2m.analysis_id,\n",
      "#stats.stats_name,stats.stats_value\n",
      "#from seq_units su, m2m_analysis_seq_unit m2m, rqc_analysis_stats stats\n",
      "#where su.run_date >= '2012-01-01' and su.run_date < '2012-03-10' and su.seq_unit_name REGEXP '^.+[0-9]\\.srf$'\n",
      "#and m2m.seq_unit_name = su.seq_unit_name\n",
      "#and m2m.analysis_id = stats.analysis_id \"\"\")\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import MySQLdb\n",
      "import datetime\n",
      "conn = MySQLdb.connect(host='redwood.jgi-psf.org',user='rqc_ro',passwd='rqcro',db='rqc')\n",
      "cur = conn.cursor()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def summarize_sequencing_by_runtype(results, instrument_type):\n",
      "    \n",
      "    if results.get(instrument_type,None) is None:\n",
      "        print 'Instrument type : %s not present in the results' % instrument_type\n",
      "        return (None,None)\n",
      "    else:\n",
      "        dates = []\n",
      "        readcounts = []\n",
      "        \n",
      "        if instrument_type == 'HiSeq':\n",
      "            #separate the control lane\n",
      "            hiseq_control_lane_dates = []\n",
      "            hiseq_control_lane_read_counts = []\n",
      "        \n",
      "        \n",
      "        for (count,fastq) in enumerate(results[instrument_type].keys()):\n",
      "            \n",
      "            run_date = results[instrument_type][fastq]['run_date']\n",
      "            if type(run_date) is not datetime.date:\n",
      "                print ('arg must be a datetime.date, not a %s' % type(run_date))\n",
      "                print instrument_type,fastq,run_date,results[instrument_type][fastq]['read count'],results[instrument_type][fastq]['library_name']\n",
      "                print 'Skipping this row of data'\n",
      "                continue\n",
      "\n",
      "            try:\n",
      "                lane_read_count = int(results[instrument_type][fastq]['read count'])\n",
      "            except:\n",
      "                print 'error %s ' % results[instrument_type][fastq]['read count']\n",
      "                print instrument_type,fastq,run_date,results[instrument_type][fastq]['read count'],results[instrument_type][fastq]['library_name']\n",
      "                print 'Skipping this row of data'\n",
      "                continue\n",
      "            \n",
      "            if instrument_type == 'HiSeq' and (results[instrument_type][fastq]['library_name'] == 'PhiX' or results[instrument_type][fastq]['library_name'] == 'GAZP' ):\n",
      "                #print instrument_type,fastq,run_date,results[instrument_type][fastq]['read count'],results[instrument_type][fastq]['library_name']\n",
      "                hiseq_control_lane_dates.append(run_date)\n",
      "                hiseq_control_lane_read_counts.append(lane_read_count)\n",
      "                continue\n",
      "                \n",
      "            \n",
      "            \n",
      "            \n",
      "            dates.append(run_date)\n",
      "            readcounts.append(lane_read_count)\n",
      "    print 'Processed %d fastq for instrument %s' % (count+1,instrument_type)\n",
      "    \n",
      "    if instrument_type == 'HiSeq':\n",
      "        return (dates,readcounts,hiseq_control_lane_dates,hiseq_control_lane_read_counts)\n",
      "    else:\n",
      "        return (dates,readcounts)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "user_start_date = \"2012-07-01\"\n",
      "user_end_date = \"2012-09-30\"\n",
      "fmt_start_date = datetime.datetime.strptime(user_start_date.replace(\"'\",''),'%Y-%m-%d').date()\n",
      "fmt_end_date = datetime.datetime.strptime(user_end_date.replace(\"'\",''),'%Y-%m-%d').date()\n",
      "\n",
      "print 'formatted_start_date : %s ' % fmt_start_date\n",
      "print 'formatted_end_date : %s ' % fmt_end_date\n",
      "\n",
      "\n",
      "sql_query = \"\"\"\n",
      "select DISTINCT \n",
      "sps.physical_run_unit_id,spr.physical_run_unit_id,spr.physical_run_id\n",
      ",sps.dt_begin\n",
      ",srun.physical_run_type,srun.platform_name,srun.instrument_type,srun.instrument_name,srun.read1_length,srun.read2_length\n",
      ",ssu.library_name\n",
      ",ssf.file_name\n",
      ",spr.run_section\n",
      ",ras.stats_name,ras.stats_value\n",
      "from \n",
      "sdm.physical_run_unit_status_history sps\n",
      ",sdm.physical_run_unit spr\n",
      ",sdm.physical_run srun\n",
      ",sdm.sdm_seq_unit ssu\n",
      ",sdm.sdm_seq_unit_file ssf\n",
      ",rqc.m2m_analysis_seq_unit rm2m\n",
      ",rqc.rqc_analysis_stats ras\n",
      "where sps.dt_begin >= '%s' and sps.dt_begin < '%s'\n",
      "and sps.physical_run_unit_status_id = 3\n",
      "and sps.physical_run_unit_id = spr.physical_run_unit_id\n",
      "and spr.physical_run_id = srun.physical_run_id\n",
      "and spr.physical_run_unit_id = ssu.physical_run_unit_id\n",
      "and ssu.sdm_seq_unit_id = ssf.sdm_seq_unit_id\n",
      "and ssf.file_type = 'FASTQ'\n",
      "and ssf.file_name REGEXP '^.+[0-9]\\.fastq.gz'\n",
      "and REPLACE (ssf.file_name, '.fastq.gz','.srf') = rm2m.seq_unit_name\n",
      "and rm2m.analysis_id = ras.analysis_id\n",
      "and ras.stats_name = 'read count'\n",
      "\"\"\" % (fmt_start_date,fmt_end_date+datetime.timedelta(days=1))\n",
      "\n",
      "#debug\n",
      "#print sql_query\n",
      "\n",
      "cur.execute(sql_query)\n",
      "results = {}\n",
      "\n",
      "for (count_rows,row) in enumerate(cur.fetchall()):\n",
      "    \n",
      "    #first 3 cols returned by sql are for debugging purposes only\n",
      "    debug_part = row[:3]\n",
      "    data = row[3:]\n",
      "    \n",
      "    #the order of the following is based on sql query\n",
      "    run_date = data[0]\n",
      "    physical_run_type = data[1]\n",
      "    platform = data[2]\n",
      "    instrument_type = data[3]\n",
      "    instrument_name = data[4]\n",
      "    read1_length = data[5]\n",
      "    read2_length = data[6]\n",
      "    library_name = data[7]\n",
      "    seq_unit = data[8]\n",
      "    seq_unit_lane = data[9]\n",
      "    fastq_rqc_stats_name = data[10]\n",
      "    fastq_rqc_stats_value =  data[11]\n",
      "    \n",
      "    \n",
      "    #print run_date,physical_run_type,platform,instrument_type,instrument_name,library_name,seq_unit\n",
      "    \n",
      "    #register new instrument type\n",
      "    if results.get(instrument_type, None) is None:\n",
      "        results[instrument_type] = {}\n",
      "    \n",
      "    #register new fastq\n",
      "    if results[instrument_type].get(seq_unit, None) is None:\n",
      "        results[instrument_type][seq_unit] = {}\n",
      "        results[instrument_type][seq_unit]['run_date'] = run_date.date()\n",
      "        results[instrument_type][seq_unit]['physical_run_type'] = physical_run_type\n",
      "        results[instrument_type][seq_unit]['platform'] = platform\n",
      "        results[instrument_type][seq_unit]['instrument_type'] = instrument_type\n",
      "        results[instrument_type][seq_unit]['instrument_name'] = instrument_name\n",
      "        results[instrument_type][seq_unit]['library_name'] = library_name\n",
      "        results[instrument_type][seq_unit]['lane'] = seq_unit_lane\n",
      "        results[instrument_type][seq_unit][fastq_rqc_stats_name] = fastq_rqc_stats_value\n",
      "    else:\n",
      "        results[instrument_type][seq_unit][fastq_rqc_stats_name] = fastq_rqc_stats_value\n",
      "    \n",
      "\n",
      "count_rows += 1\n",
      "print 'processed %d rows from mysql results' % count_rows\n",
      "(hiseq_dates,hiseq_lanes_read_counts,hiseq_control_lane_dates,hiseq_control_lane_read_counts)=summarize_sequencing_by_runtype(results,'HiSeq')\n",
      "(miseq_dates,miseq_lanes_read_counts)=summarize_sequencing_by_runtype(results,'MiSeq')\n",
      "(ga2seq_dates,ga2seq_lanes_read_counts)=summarize_sequencing_by_runtype(results,'GA2')\n",
      "(ga2xseq_dates,ga2xseq_lanes_read_counts)=summarize_sequencing_by_runtype(results,'GAIIx')\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "    \n",
      "    \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%hist\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import matplotlib.pyplot as plt\n",
      "from matplotlib.font_manager import FontProperties\n",
      "fig = plt.figure(figsize=(14,12))\n",
      "fig.suptitle('JGI - Sequencing Summary (read counts) \\n Q: %s-%s -- %s-%s' % (fmt_start_date.strftime(\"%B\"),fmt_start_date.year,fmt_end_date.strftime(\"%B\"),fmt_end_date.year), fontsize=20)\n",
      "\n",
      "#TBD\n",
      "#determine the number of rows in plot, based on the number of plots\n",
      "\n",
      "\n",
      "\n",
      "#hiseq\n",
      "ax1 = plt.subplot2grid((3,4),(0,0),colspan=4)\n",
      "plt.plot_date(hiseq_dates,hiseq_lanes_read_counts,color='#33A02C')\n",
      "plt.title(\"HiSeq's Output\")\n",
      "plt.ylabel('#reads per lane(in 100 millions)')\n",
      "plt.plot_date(hiseq_control_lane_dates,hiseq_control_lane_read_counts,color='red',label=\"PhIX-lane\")\n",
      "plt.plot\n",
      "fontP = FontProperties()\n",
      "fontP.set_size('small')\n",
      "plt.legend(loc='upper left',prop=fontP)\n",
      "ax2 = plt.subplot2grid((3,4),(2,0),colspan=2)\n",
      "plt.hist(hiseq_lanes_read_counts,bins=50,color='#33A02C')\n",
      "plt.hist(hiseq_control_lane_read_counts,bins=50,color='red',label=\"PhiX\")\n",
      "plt.xlabel('Hi-Seq read count histogram',fontsize=14)\n",
      "plt.legend(loc='upper left')\n",
      "\n",
      "\n",
      "#MiSeq\n",
      "ax7 = plt.subplot2grid((3,4),(1,0),colspan=4,sharex=ax1)\n",
      "plt.plot_date(x=miseq_dates,y=miseq_lanes_read_counts,color='#A6CEE3')\n",
      "plt.title(\"MiSeq's Output\")\n",
      "plt.ylabel('#reads per lane(in 10 millions)')\n",
      "plt.xlabel('Time Line',fontsize=12)\n",
      "ax8 = plt.subplot2grid((3,4),(2,2),colspan=2)\n",
      "plt.hist(miseq_lanes_read_counts,bins=50,color='#A6CEE3')\n",
      "plt.xlabel('Mi-Seq read count histogram',fontsize=14)\n",
      "\n",
      "\n",
      "\n",
      "#GAIIx\n",
      "#ax3 = plt.subplot2grid((4,4),(1,0),colspan=3,sharex=ax1)\n",
      "#plt.plot_date(x=ga2xseq_dates,y=ga2xseq_lanes_read_counts,color='#B2DF8A')\n",
      "#plt.title(\"GAIIx's Output\")\n",
      "#ax4 = plt.subplot2grid((4,4),(1,3),colspan=1)\n",
      "#plt.hist(ga2xseq_lanes_read_counts,bins=50,color='#B2DF8A')\n",
      "\n",
      "'''\n",
      "#GA2\n",
      "ax5 = plt.subplot2grid((3,4),(1,0),colspan=4,sharex=ax1)\n",
      "plt.plot_date(x=ga2seq_dates,y=ga2seq_lanes_read_counts,color='#1F78B4')\n",
      "plt.title(\"GAII Output\")\n",
      "ax6 = plt.subplot2grid((3,4),(2,2),colspan=2)\n",
      "plt.hist(ga2seq_lanes_read_counts,bins=50,color='#1F78B4')\n",
      "'''\n",
      "\n",
      "\n",
      "out_file_name = 'JGI_sequencing_summary_%s-%s__%s-%s.png' % (fmt_start_date.strftime(\"%B\"),fmt_start_date.year,fmt_end_date.strftime(\"%B\"),fmt_end_date.year)\n",
      "plt.savefig(out_file_name)\n",
      "\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql_query = \"\"\"select ssu.sdm_seq_unit_id, ssf.sdm_seq_unit_id\n",
      "#ssu.physical_run_unit_id,ssu.sdm_seq_unit_status_id\n",
      ",ssf.file_name,ssu.parent_sdm_seq_unit_id\n",
      ",ssu1.library_name as parent_library_name, ssf1.file_name as parent_lib_file_name\n",
      ",ssu.library_name as lib_name, ssu.seq_barcode,ssu.is_multiplexed\n",
      "#,ssu.physical_run_unit_id,spu.physical_run_unit_id\n",
      ",srun.physical_run_type,srun.platform_name,srun.instrument_type,srun.instrument_name,srun.read1_length,srun.read2_length\n",
      ",rm2m.analysis_id\n",
      ",rstats.stats_name,rstats.stats_value\n",
      "\n",
      "from \n",
      "sdm.sdm_seq_unit ssu\n",
      "\n",
      "INNER JOIN sdm.sdm_seq_unit_file ssf\n",
      "ON ssf.sdm_seq_unit_id = ssu.sdm_seq_unit_id\n",
      "\n",
      "INNER JOIN sdm.physical_run_unit spu\n",
      "ON ssu.physical_run_unit_id = spu.physical_run_unit_id\n",
      "\n",
      "INNER JOIN sdm.physical_run srun\n",
      "on spu.physical_run_id = srun.physical_run_id\n",
      "\n",
      "LEFT JOIN rqc.m2m_analysis_seq_unit rm2m\n",
      "on REPLACE (ssf.file_name, '.fastq.gz','.srf') = rm2m.seq_unit_name\n",
      "\n",
      "LEFT JOIN rqc.rqc_analysis_stats rstats\n",
      "on rm2m.analysis_id = rstats.analysis_id\n",
      "\n",
      "LEFT JOIN sdm.sdm_seq_unit ssu1\n",
      "on ssu.parent_sdm_seq_unit_id = ssu1.sdm_seq_unit_id\n",
      "\n",
      "LEFT JOIN sdm.sdm_seq_unit_file ssf1\n",
      "on ssu1.sdm_seq_unit_id = ssf1.sdm_seq_unit_id\n",
      "\n",
      "and ssf.file_type = 'FASTQ'\n",
      "and ssf1.file_type = 'FASTQ'\n",
      "\n",
      "where ssu.parent_sdm_seq_unit_id in ( \n",
      "\t\t\t\t\tselect sdm_seq_unit_id from sdm.sdm_seq_unit \n",
      "\t\t\t\t\twhere  library_name  IN (\"BTYS\")\n",
      "\t\t\t\t    )\n",
      "\n",
      "or ssu.library_name in (\"BTYS\")\n",
      "\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "cur.execute(sql_query)\n",
      "\n",
      "data = {}\n",
      "\n",
      "for row in cur.fetchall():\n",
      "    \n",
      "    sdm_seq_unit_id = row[0]\n",
      "    #ssf_sdm_seq_unit_id = row[1].strip() #skipping this, mainly for debugging purpose\n",
      "    seq_unit_name = row[2].strip()\n",
      "    parent_sdm_seq_unit_id = row[3]\n",
      "    parent_lib_name = row[4]\n",
      "    parent_lib_seq_unit_name = row[5]\n",
      "    lib_name = row[6].strip()\n",
      "    barcode = row[7]\n",
      "    is_multiplexed = row[8]\n",
      "    physical_run_type = row[9].strip()\n",
      "    platform_name = row[10].strip()\n",
      "    instrument_type = row[11].strip()\n",
      "    instrument_name = row[12].strip()\n",
      "    read_1_length = row[13]\n",
      "    read_2_length = row[14]\n",
      "    rqc_analysis_id = row[15]\n",
      "    rqc_stats_name = row[16]\n",
      "    rqc_stats_value = row[17]\n",
      "    \n",
      "    \n",
      "    if seq_unit_name.find('FAIL') != -1:\n",
      "        continue\n",
      "    \n",
      "    insert_record_at = None\n",
      "    child_lib_name = None\n",
      "    child_seq_unit_name = None\n",
      "\n",
      "    #not a multiplexed library\n",
      "    if parent_lib_name is None and lib_name is not None:\n",
      "        parent_lib_name = lib_name\n",
      "        parent_lib_seq_unit_name = seq_unit_name\n",
      "    \n",
      "    #multiplexed library\n",
      "    elif parent_lib_name is not None and lib_name is not None:\n",
      "        child_lib_name = lib_name\n",
      "        child_seq_unit_name = seq_unit_name\n",
      "        \n",
      "    #by now we have the relation, parent and child lib and fastq name\n",
      "    \n",
      "    \n",
      "    \n",
      "    if parent_lib_name is None:  #sanity check\n",
      "        #most likely probem\n",
      "        print 'Parent lib None'\n",
      "        \n",
      "    else:\n",
      "        \n",
      "        #new parent library\n",
      "        if data.get(parent_lib_name,None) is None:\n",
      "            data[parent_lib_name] = {}\n",
      "            data[parent_lib_name]['libs'] = {}\n",
      "            data[parent_lib_name]['seq_units'] = {}\n",
      "            data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name] ={}\n",
      "            insert_record_at =  data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name] \n",
      "        \n",
      "        #insert the child under appropriate parent\n",
      "        if child_lib_name is not None:\n",
      "            if data[parent_lib_name]['libs'].get(child_lib_name,None) is None:\n",
      "                data[parent_lib_name]['libs'][child_lib_name] = {}\n",
      "                data[parent_lib_name]['libs'][child_lib_name]['seq_units'] = {}\n",
      "                data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name] = {}\n",
      "                insert_record_at =  data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name]\n",
      "            \n",
      "    \n",
      "    \n",
      "    \n",
      "      \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 98
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pprint\n",
      "\n",
      "    \n",
      "pprint.pprint(data)\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "{'BTYS': {'libs': {'AUCY': {'seq_units': {'6104.1.36425.CCGTCC.fastq.gz': {}}},\n",
        "                   'AUNN': {'seq_units': {'6104.1.36425.CAGATC.fastq.gz': {}}},\n",
        "                   'AUNO': {'seq_units': {'6104.1.36425.CTTGTA.fastq.gz': {}}},\n",
        "                   'AUNP': {'seq_units': {'6104.1.36425.ATCACG.fastq.gz': {}}},\n",
        "                   'AUNS': {'seq_units': {'6104.1.36425.ACTTGA.fastq.gz': {}}},\n",
        "                   'AUNT': {'seq_units': {'6104.1.36425.AGTTCC.fastq.gz': {}}},\n",
        "                   'AUNU': {'seq_units': {'6104.1.36425.TGACCA.fastq.gz': {}}},\n",
        "                   'AUNW': {'seq_units': {'6104.1.36425.GTAGAG.fastq.gz': {}}},\n",
        "                   'AUNX': {'seq_units': {'6104.1.36425.GTCCGC.fastq.gz': {}}},\n",
        "                   'AUNY': {'seq_units': {'6104.1.36425.GCCAAT.fastq.gz': {}}},\n",
        "                   'AUNZ': {'seq_units': {'6104.1.36425.GTGGCC.fastq.gz': {}}},\n",
        "                   'AUOA': {'seq_units': {'6104.1.36425.ATGTCA.fastq.gz': {}}},\n",
        "                   'AUOB': {'seq_units': {'6104.1.36425.TTAGGC.fastq.gz': {}}},\n",
        "                   'AUOO': {'seq_units': {'6104.1.36425.CGATGT.fastq.gz': {}}},\n",
        "                   'AUOP': {'seq_units': {'6104.1.36425.ACAGTG.fastq.gz': {}}},\n",
        "                   'AUOS': {'seq_units': {'6104.1.36425.GGCTAC.fastq.gz': {}}},\n",
        "                   'AUOT': {'seq_units': {'6104.1.36425.GATCAG.fastq.gz': {}}},\n",
        "                   'AUOU': {'seq_units': {'6104.1.36425.GTGAAA.fastq.gz': {}}},\n",
        "                   'AUOW': {'seq_units': {'6104.1.36425.TAGCTT.fastq.gz': {}}},\n",
        "                   'AUPA': {'seq_units': {'6104.1.36425.AGTCAA.fastq.gz': {}}},\n",
        "                   'UNKNOWN': {'seq_units': {'6104.1.36425.UNKNOWN.fastq.gz': {}}}},\n",
        "          'seq_units': {'6104.1.36425.fastq.gz': {}}}}\n"
       ]
      }
     ],
     "prompt_number": 99
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "      \n",
      "    \n",
      "     ''' \n",
      "        #new child of new parent fastq\n",
      "        if data.get(parent_sdm_seq_unit_id,None) is None:\n",
      "            #new_parent_fastq\n",
      "            data[parent_sdm_seq_unit_id] = {}\n",
      "            data[parent_sdm_seq_unit_id]['childs'] = {}\n",
      "            #new parent's new child fastq\n",
      "            data[parent_sdm_seq_unit_id]['childs'][sdm_seq_unit_id] = {}\n",
      "        \n",
      "        #new child of existing parent fastq\n",
      "        elif data[parent_sdm_seq_unit_id].get(sdm_seq_unit_id, None) is None:\n",
      "                data[parent_sdm_seq_unit_id]['childs'][sdm_seq_unit_id] = {}\n",
      "                \n",
      "        else:\n",
      "            pass\n",
      "            \n",
      "            #data[parent_sdm_seq_unit_id][sdm_seq_unit_id] = {}\n",
      "    \n",
      "        \n",
      "    \n",
      "        data[sdm_seq_unit_id]['lib_name'] = lib_name\n",
      "            data[sdm_seq_unit_id]['barcode'] = barcode\n",
      "            data[sdm_seq_unit_id]['seq_unit_name'] = seq_unit_name\n",
      "            data[sdm_seq_unit_id]['is_multiplexed'] = is_multiplexed\n",
      "        \n",
      "    \n",
      "    \n",
      "    \n",
      "    data[sdm_seq_unit_id]['physical_run_type'] = physical_run_type\n",
      "            data[sdm_seq_unit_id]['platform_name'] = platform_name\n",
      "            data[sdm_seq_unit_id]['instrument_type'] = instrument_type\n",
      "            data[sdm_seq_unit_id]['read_1_length'] = read_1_length\n",
      "            data[sdm_seq_unit_id]['read_2_length'] = read_2_length\n",
      "            data[sdm_seq_unit_id]['rqc_analysis_id'] = rqc_analysis_id\n",
      "            data[sdm_seq_unit_id]['childs'] = {}\n",
      "\n",
      "\n",
      "\n",
      "        data[sdm_seq_unit_id]['file_name'] = seq_unit_name \n",
      "        data[sdm_seq_unit_id]['lib_name'] = lib_name\n",
      "        data[sdm_seq_unit_id]['barcode'] = barcode\n",
      "        data[sdm_seq_unit_id]['seq_unit_name'] = seq_unit_name\n",
      "        data[sdm_seq_unit_id]['is_multiplexed'] = is_multiplexed\n",
      "        data[sdm_seq_unit_id]['physical_run_type'] = physical_run_type\n",
      "        data[sdm_seq_unit_id]['platform_name'] = platform_name\n",
      "        data[sdm_seq_unit_id]['instrument_type'] = instrument_type\n",
      "        data[sdm_seq_unit_id]['read_1_length'] = read_1_length\n",
      "        data[sdm_seq_unit_id]['read_2_length'] = read_2_length\n",
      "        data[sdm_seq_unit_id]['rqc_analysis_id'] = rqc_analysis_id\n",
      "\n",
      "    \n",
      "        #new fastq\n",
      "        data[sdm_seq_unit_id] = {}\n",
      "        data[sdm_seq_unit_id]['file_name'] = seq_unit_name \n",
      "        data[sdm_seq_unit_id]['lib_name'] = lib_name\n",
      "        data[sdm_seq_unit_id]['barcode'] = barcode\n",
      "        data[sdm_seq_unit_id]['seq_unit_name'] = seq_unit_name\n",
      "        data[sdm_seq_unit_id]['is_multiplexed'] = is_multiplexed\n",
      "        data[sdm_seq_unit_id]['physical_run_type'] = physical_run_type\n",
      "        data[sdm_seq_unit_id]['platform_name'] = platform_name\n",
      "        data[sdm_seq_unit_id]['instrument_type'] = instrument_type\n",
      "        data[sdm_seq_unit_id]['read_1_length'] = read_1_length\n",
      "        data[sdm_seq_unit_id]['read_2_length'] = read_2_length\n",
      "        data[sdm_seq_unit_id]['rqc_analysis_id'] = rqc_analysis_id\n",
      "        #data[sdm_seq_unit_id]['rqc_stats_name'] = rqc_stats_name\n",
      "        #data[sdm_seq_unit_id]['rqc_stats_value'] = rqc_stats_value\n",
      "           \n",
      "    else:\n",
      "        pass\n",
      "        \n",
      "    '''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print seq_unit_name"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "seq_unit_name.find('Fsds')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}