{
 "metadata": {
  "name": "JGI_runs_quality_summaries"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pymysql\n",
      "conn = pymysql.connect(host='redwood.jgi-psf.org',user='jazz',passwd='jazz234',db='rqc')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cur = conn.cursor()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cur.execute(\"SELECT * from seq_units limit 10\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cur.execute(\"select run_id,seq_unit_name,run_date from seq_units where run_date >= '2012-09-25' AND run_date < '2012-11-26'\")\n",
      "\n",
      "for r in cur.fetchall():\n",
      "    print r"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "#cur.execute(\"\"\" select su.seq_unit_name, su.run_barcode, su.run_id, su.run_date,su.run_section,m2m.analysis_id,\n",
      "#stats.stats_name,stats.stats_value\n",
      "#from seq_units su, m2m_analysis_seq_unit m2m, rqc_analysis_stats stats\n",
      "#where su.run_date >= '2012-01-01' and su.run_date < '2012-03-10' and su.seq_unit_name REGEXP '^.+[0-9]\\.srf$'\n",
      "#and m2m.seq_unit_name = su.seq_unit_name\n",
      "#and m2m.analysis_id = stats.analysis_id \"\"\")\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import MySQLdb\n",
      "import datetime\n",
      "conn = MySQLdb.connect(host='redwood.jgi-psf.org',user='rqc_ro',passwd='rqcro',db='rqc')\n",
      "cur = conn.cursor()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 93
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def summarize_sequencing_by_runtype(results, instrument_type):\n",
      "    \n",
      "    if results.get(instrument_type,None) is None:\n",
      "        print 'Instrument type : %s not present in the results' % instrument_type\n",
      "        return (None,None)\n",
      "    else:\n",
      "        dates = []\n",
      "        readcounts = []\n",
      "        \n",
      "        if instrument_type == 'HiSeq':\n",
      "            #separate the control lane\n",
      "            hiseq_control_lane_dates = []\n",
      "            hiseq_control_lane_read_counts = []\n",
      "        \n",
      "        \n",
      "        for (count,fastq) in enumerate(results[instrument_type].keys()):\n",
      "            \n",
      "            run_date = results[instrument_type][fastq]['run_date']\n",
      "            if type(run_date) is not datetime.date:\n",
      "                print ('arg must be a datetime.date, not a %s' % type(run_date))\n",
      "                print instrument_type,fastq,run_date,results[instrument_type][fastq]['read count'],results[instrument_type][fastq]['library_name']\n",
      "                print 'Skipping this row of data'\n",
      "                continue\n",
      "\n",
      "            try:\n",
      "                lane_read_count = int(results[instrument_type][fastq]['read count'])\n",
      "            except:\n",
      "                print 'error %s ' % results[instrument_type][fastq]['read count']\n",
      "                print instrument_type,fastq,run_date,results[instrument_type][fastq]['read count'],results[instrument_type][fastq]['library_name']\n",
      "                print 'Skipping this row of data'\n",
      "                continue\n",
      "            \n",
      "            if instrument_type == 'HiSeq' and (results[instrument_type][fastq]['library_name'] == 'PhiX' or results[instrument_type][fastq]['library_name'] == 'GAZP' ):\n",
      "                #print instrument_type,fastq,run_date,results[instrument_type][fastq]['read count'],results[instrument_type][fastq]['library_name']\n",
      "                hiseq_control_lane_dates.append(run_date)\n",
      "                hiseq_control_lane_read_counts.append(lane_read_count)\n",
      "                continue\n",
      "                \n",
      "            \n",
      "            \n",
      "            \n",
      "            dates.append(run_date)\n",
      "            readcounts.append(lane_read_count)\n",
      "    print 'Processed %d fastq for instrument %s' % (count+1,instrument_type)\n",
      "    \n",
      "    if instrument_type == 'HiSeq':\n",
      "        return (dates,readcounts,hiseq_control_lane_dates,hiseq_control_lane_read_counts)\n",
      "    else:\n",
      "        return (dates,readcounts)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "user_start_date = \"2012-07-01\"\n",
      "user_end_date = \"2012-09-30\"\n",
      "fmt_start_date = datetime.datetime.strptime(user_start_date.replace(\"'\",''),'%Y-%m-%d').date()\n",
      "fmt_end_date = datetime.datetime.strptime(user_end_date.replace(\"'\",''),'%Y-%m-%d').date()\n",
      "\n",
      "print 'formatted_start_date : %s ' % fmt_start_date\n",
      "print 'formatted_end_date : %s ' % fmt_end_date\n",
      "\n",
      "\n",
      "sql_query = \"\"\"\n",
      "select DISTINCT \n",
      "sps.physical_run_unit_id,spr.physical_run_unit_id,spr.physical_run_id\n",
      ",sps.dt_begin\n",
      ",srun.physical_run_type,srun.platform_name,srun.instrument_type,srun.instrument_name,srun.read1_length,srun.read2_length\n",
      ",ssu.library_name\n",
      ",ssf.file_name\n",
      ",spr.run_section\n",
      ",ras.stats_name,ras.stats_value\n",
      "from \n",
      "sdm.physical_run_unit_status_history sps\n",
      ",sdm.physical_run_unit spr\n",
      ",sdm.physical_run srun\n",
      ",sdm.sdm_seq_unit ssu\n",
      ",sdm.sdm_seq_unit_file ssf\n",
      ",rqc.m2m_analysis_seq_unit rm2m\n",
      ",rqc.rqc_analysis_stats ras\n",
      "where sps.dt_begin >= '%s' and sps.dt_begin < '%s'\n",
      "and sps.physical_run_unit_status_id = 3\n",
      "and sps.physical_run_unit_id = spr.physical_run_unit_id\n",
      "and spr.physical_run_id = srun.physical_run_id\n",
      "and spr.physical_run_unit_id = ssu.physical_run_unit_id\n",
      "and ssu.sdm_seq_unit_id = ssf.sdm_seq_unit_id\n",
      "and ssf.file_type = 'FASTQ'\n",
      "and ssf.file_name REGEXP '^.+[0-9]\\.fastq.gz'\n",
      "and REPLACE (ssf.file_name, '.fastq.gz','.srf') = rm2m.seq_unit_name\n",
      "and rm2m.analysis_id = ras.analysis_id\n",
      "and ras.stats_name = 'read count'\n",
      "\"\"\" % (fmt_start_date,fmt_end_date+datetime.timedelta(days=1))\n",
      "\n",
      "#debug\n",
      "#print sql_query\n",
      "\n",
      "cur.execute(sql_query)\n",
      "results = {}\n",
      "\n",
      "for (count_rows,row) in enumerate(cur.fetchall()):\n",
      "    \n",
      "    #first 3 cols returned by sql are for debugging purposes only\n",
      "    debug_part = row[:3]\n",
      "    data = row[3:]\n",
      "    \n",
      "    #the order of the following is based on sql query\n",
      "    run_date = data[0]\n",
      "    physical_run_type = data[1]\n",
      "    platform = data[2]\n",
      "    instrument_type = data[3]\n",
      "    instrument_name = data[4]\n",
      "    read1_length = data[5]\n",
      "    read2_length = data[6]\n",
      "    library_name = data[7]\n",
      "    seq_unit = data[8]\n",
      "    seq_unit_lane = data[9]\n",
      "    fastq_rqc_stats_name = data[10]\n",
      "    fastq_rqc_stats_value =  data[11]\n",
      "    \n",
      "    \n",
      "    #print run_date,physical_run_type,platform,instrument_type,instrument_name,library_name,seq_unit\n",
      "    \n",
      "    #register new instrument type\n",
      "    if results.get(instrument_type, None) is None:\n",
      "        results[instrument_type] = {}\n",
      "    \n",
      "    #register new fastq\n",
      "    if results[instrument_type].get(seq_unit, None) is None:\n",
      "        results[instrument_type][seq_unit] = {}\n",
      "        results[instrument_type][seq_unit]['run_date'] = run_date.date()\n",
      "        results[instrument_type][seq_unit]['physical_run_type'] = physical_run_type\n",
      "        results[instrument_type][seq_unit]['platform'] = platform\n",
      "        results[instrument_type][seq_unit]['instrument_type'] = instrument_type\n",
      "        results[instrument_type][seq_unit]['instrument_name'] = instrument_name\n",
      "        results[instrument_type][seq_unit]['library_name'] = library_name\n",
      "        results[instrument_type][seq_unit]['lane'] = seq_unit_lane\n",
      "        results[instrument_type][seq_unit][fastq_rqc_stats_name] = fastq_rqc_stats_value\n",
      "    else:\n",
      "        results[instrument_type][seq_unit][fastq_rqc_stats_name] = fastq_rqc_stats_value\n",
      "    \n",
      "\n",
      "count_rows += 1\n",
      "print 'processed %d rows from mysql results' % count_rows\n",
      "(hiseq_dates,hiseq_lanes_read_counts,hiseq_control_lane_dates,hiseq_control_lane_read_counts)=summarize_sequencing_by_runtype(results,'HiSeq')\n",
      "(miseq_dates,miseq_lanes_read_counts)=summarize_sequencing_by_runtype(results,'MiSeq')\n",
      "(ga2seq_dates,ga2seq_lanes_read_counts)=summarize_sequencing_by_runtype(results,'GA2')\n",
      "(ga2xseq_dates,ga2xseq_lanes_read_counts)=summarize_sequencing_by_runtype(results,'GAIIx')\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "    \n",
      "    \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import matplotlib.pyplot as plt\n",
      "from matplotlib.font_manager import FontProperties\n",
      "fig = plt.figure(figsize=(14,12))\n",
      "fig.suptitle('JGI - Sequencing Summary (read counts) \\n Q: %s-%s -- %s-%s' % (fmt_start_date.strftime(\"%B\"),fmt_start_date.year,fmt_end_date.strftime(\"%B\"),fmt_end_date.year), fontsize=20)\n",
      "\n",
      "#TBD\n",
      "#determine the number of rows in plot, based on the number of plots\n",
      "\n",
      "\n",
      "\n",
      "#hiseq\n",
      "ax1 = plt.subplot2grid((3,4),(0,0),colspan=4)\n",
      "plt.plot_date(hiseq_dates,hiseq_lanes_read_counts,color='#33A02C')\n",
      "plt.title(\"HiSeq's Output\")\n",
      "plt.ylabel('#reads per lane(in 100 millions)')\n",
      "plt.plot_date(hiseq_control_lane_dates,hiseq_control_lane_read_counts,color='red',label=\"PhIX-lane\")\n",
      "plt.plot\n",
      "fontP = FontProperties()\n",
      "fontP.set_size('small')\n",
      "plt.legend(loc='upper left',prop=fontP)\n",
      "ax2 = plt.subplot2grid((3,4),(2,0),colspan=2)\n",
      "plt.hist(hiseq_lanes_read_counts,bins=50,color='#33A02C')\n",
      "plt.hist(hiseq_control_lane_read_counts,bins=50,color='red',label=\"PhiX\")\n",
      "plt.xlabel('Hi-Seq read count histogram',fontsize=14)\n",
      "plt.legend(loc='upper left')\n",
      "\n",
      "\n",
      "#MiSeq\n",
      "ax7 = plt.subplot2grid((3,4),(1,0),colspan=4,sharex=ax1)\n",
      "plt.plot_date(x=miseq_dates,y=miseq_lanes_read_counts,color='#A6CEE3')\n",
      "plt.title(\"MiSeq's Output\")\n",
      "plt.ylabel('#reads per lane(in 10 millions)')\n",
      "plt.xlabel('Time Line',fontsize=12)\n",
      "ax8 = plt.subplot2grid((3,4),(2,2),colspan=2)\n",
      "plt.hist(miseq_lanes_read_counts,bins=50,color='#A6CEE3')\n",
      "plt.xlabel('Mi-Seq read count histogram',fontsize=14)\n",
      "\n",
      "\n",
      "\n",
      "#GAIIx\n",
      "#ax3 = plt.subplot2grid((4,4),(1,0),colspan=3,sharex=ax1)\n",
      "#plt.plot_date(x=ga2xseq_dates,y=ga2xseq_lanes_read_counts,color='#B2DF8A')\n",
      "#plt.title(\"GAIIx's Output\")\n",
      "#ax4 = plt.subplot2grid((4,4),(1,3),colspan=1)\n",
      "#plt.hist(ga2xseq_lanes_read_counts,bins=50,color='#B2DF8A')\n",
      "\n",
      "'''\n",
      "#GA2\n",
      "ax5 = plt.subplot2grid((3,4),(1,0),colspan=4,sharex=ax1)\n",
      "plt.plot_date(x=ga2seq_dates,y=ga2seq_lanes_read_counts,color='#1F78B4')\n",
      "plt.title(\"GAII Output\")\n",
      "ax6 = plt.subplot2grid((3,4),(2,2),colspan=2)\n",
      "plt.hist(ga2seq_lanes_read_counts,bins=50,color='#1F78B4')\n",
      "'''\n",
      "\n",
      "\n",
      "out_file_name = 'JGI_sequencing_summary_%s-%s__%s-%s.png' % (fmt_start_date.strftime(\"%B\"),fmt_start_date.year,fmt_end_date.strftime(\"%B\"),fmt_end_date.year)\n",
      "plt.savefig(out_file_name)\n",
      "\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sql_query = \"\"\"select ssu.sdm_seq_unit_id, ssf.sdm_seq_unit_id\n",
      "#ssu.physical_run_unit_id,ssu.sdm_seq_unit_status_id\n",
      ",ssf.file_name,ssu.parent_sdm_seq_unit_id\n",
      ",ssu1.library_name as parent_library_name, ssf1.file_name as parent_lib_file_name\n",
      ",ssu.library_name as lib_name, ssu.seq_barcode,ssu.is_multiplexed\n",
      "#,ssu.physical_run_unit_id,spu.physical_run_unit_id\n",
      ",srun.physical_run_type,srun.platform_name,srun.instrument_type,srun.instrument_name,srun.read1_length,srun.read2_length\n",
      ",rm2m.analysis_id\n",
      ",rstats.stats_name,rstats.stats_value\n",
      "\n",
      "from \n",
      "sdm.sdm_seq_unit ssu\n",
      "\n",
      "INNER JOIN sdm.sdm_seq_unit_file ssf\n",
      "ON ssf.sdm_seq_unit_id = ssu.sdm_seq_unit_id\n",
      "\n",
      "INNER JOIN sdm.physical_run_unit spu\n",
      "ON ssu.physical_run_unit_id = spu.physical_run_unit_id\n",
      "\n",
      "INNER JOIN sdm.physical_run srun\n",
      "on spu.physical_run_id = srun.physical_run_id\n",
      "\n",
      "LEFT JOIN rqc.m2m_analysis_seq_unit rm2m\n",
      "on REPLACE (ssf.file_name, '.fastq.gz','.srf') = rm2m.seq_unit_name\n",
      "\n",
      "LEFT JOIN rqc.rqc_analysis_stats rstats\n",
      "on rm2m.analysis_id = rstats.analysis_id\n",
      "\n",
      "LEFT JOIN sdm.sdm_seq_unit ssu1\n",
      "on ssu.parent_sdm_seq_unit_id = ssu1.sdm_seq_unit_id\n",
      "\n",
      "LEFT JOIN sdm.sdm_seq_unit_file ssf1\n",
      "on ssu1.sdm_seq_unit_id = ssf1.sdm_seq_unit_id\n",
      "\n",
      "and ssf.file_type = 'FASTQ'\n",
      "and ssf1.file_type = 'FASTQ'\n",
      "\n",
      "where ssu.parent_sdm_seq_unit_id in ( \n",
      "\t\t\t\t\tselect sdm_seq_unit_id from sdm.sdm_seq_unit \n",
      "\t\t\t\t\twhere  library_name  IN (\"BTYS\")\n",
      "\t\t\t\t    )\n",
      "\n",
      "or ssu.library_name in (\"BTYS\")\n",
      "\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "cur.execute(sql_query)\n",
      "\n",
      "data = {}\n",
      "\n",
      "for count,row in enumerate(cur.fetchall()):\n",
      "    \n",
      "    \n",
      "    if count == 200:\n",
      "        pass\n",
      "    \n",
      "    \n",
      "    sdm_seq_unit_id = row[0]\n",
      "    #ssf_sdm_seq_unit_id = row[1].strip() #skipping this, mainly for debugging purpose\n",
      "    seq_unit_name = row[2].strip()\n",
      "    parent_sdm_seq_unit_id = row[3]\n",
      "    parent_lib_name = row[4]\n",
      "    parent_lib_seq_unit_name = row[5]\n",
      "    lib_name = row[6].strip()\n",
      "    barcode = row[7]\n",
      "    is_multiplexed = row[8]\n",
      "    physical_run_type = row[9].strip()\n",
      "    platform_name = row[10].strip()\n",
      "    instrument_type = row[11].strip()\n",
      "    instrument_name = row[12].strip()\n",
      "    read_1_length = row[13]\n",
      "    read_2_length = row[14]\n",
      "    rqc_analysis_id = row[15]\n",
      "    rqc_stats_name = row[16]\n",
      "    rqc_stats_value = row[17]\n",
      "    \n",
      "    \n",
      "    if seq_unit_name.find('FAIL') != -1:\n",
      "        continue\n",
      "    \n",
      "    insert_record_at = None\n",
      "    child_lib_name = None\n",
      "    child_seq_unit_name = None\n",
      "    \n",
      "    #not a multiplexed library\n",
      "    if parent_lib_name is None and lib_name is not None:\n",
      "        parent_lib_name = lib_name\n",
      "        parent_lib_seq_unit_name = seq_unit_name\n",
      "    \n",
      "    #multiplexed library\n",
      "    elif parent_lib_name is not None and lib_name is not None:\n",
      "        child_lib_name = lib_name\n",
      "        child_seq_unit_name = seq_unit_name\n",
      "        \n",
      "    #by now we have the relation, parent and child lib and fastq name\n",
      "    \n",
      "    \n",
      "    \n",
      "    if parent_lib_name is None:  #sanity check\n",
      "        #most likely probem\n",
      "        print 'Parent lib None'\n",
      "        \n",
      "    else:\n",
      "        \n",
      "        #new parent library\n",
      "        if data.get(parent_lib_name,None) is None:\n",
      "            data[parent_lib_name] = {}\n",
      "            data[parent_lib_name]['libs'] = {}\n",
      "            data[parent_lib_name]['seq_units'] = {}\n",
      "            data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name] = {}\n",
      "            insert_record_at =  data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name]\n",
      "        elif data[parent_lib_name]['seq_units'].get(parent_lib_seq_unit_name,None) is None:\n",
      "            data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name] = {}\n",
      "            insert_record_at =  data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name]\n",
      "        else:\n",
      "            insert_record_at =  data[parent_lib_name]['seq_units'][parent_lib_seq_unit_name]\n",
      "        \n",
      "        #insert the child under appropriate parent\n",
      "        if child_lib_name is not None:\n",
      "            if data[parent_lib_name]['libs'].get(child_lib_name,None) is None:\n",
      "                data[parent_lib_name]['libs'][child_lib_name] = {}\n",
      "                data[parent_lib_name]['libs'][child_lib_name]['seq_units'] = {}\n",
      "                data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name] = {}\n",
      "                insert_record_at =  data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name]\n",
      "            elif data[parent_lib_name]['libs'][child_lib_name]['seq_units'].get(child_seq_unit_name,None) is None:\n",
      "                data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name] = {}\n",
      "                insert_record_at =  data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name]\n",
      "            else:\n",
      "                insert_record_at =  data[parent_lib_name]['libs'][child_lib_name]['seq_units'][child_seq_unit_name]\n",
      "        \n",
      "        \n",
      "        if insert_record_at is None:\n",
      "            print 'Insert Record Error for row'\n",
      "            print row\n",
      "        \n",
      "        else:\n",
      "            \n",
      "            if insert_record_at.get('rqc_stats',None) is None:\n",
      "                insert_record_at['rqc_stats'] = {}\n",
      "            \n",
      "            if rqc_stats_name is None:\n",
      "                continue\n",
      "            \n",
      "            insert_record_at['file_name'] = seq_unit_name \n",
      "            insert_record_at['lib_name'] = lib_name\n",
      "            insert_record_at['barcode'] = barcode\n",
      "            insert_record_at['seq_unit_name'] = seq_unit_name\n",
      "            insert_record_at['is_multiplexed'] = is_multiplexed\n",
      "            insert_record_at['physical_run_type'] = physical_run_type\n",
      "            insert_record_at['platform_name'] = platform_name\n",
      "            insert_record_at['instrument_type'] = instrument_type\n",
      "            insert_record_at['read_1_length'] = read_1_length\n",
      "            insert_record_at['read_2_length'] = read_2_length\n",
      "            insert_record_at['parent_lib_seq_unit_name'] = parent_lib_seq_unit_name\n",
      "            insert_record_at['rqc_analysis_id'] = rqc_analysis_id\n",
      "            insert_record_at['rqc_stats'][rqc_stats_name]=rqc_stats_value\n",
      "        \n",
      "        \n",
      "        \n",
      "    \n",
      "    \n",
      "    \n",
      "      \n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 206
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pprint\n",
      "\n",
      "\n",
      "\n",
      "def summarize(data):\n",
      "    \n",
      "    \n",
      "    #parent library summary\n",
      "    for parent_lib in data:\n",
      "        parent_lib_total_reads = 0\n",
      "        \n",
      "        for seq_unit_name in data[parent_lib]['seq_units']:\n",
      "            \n",
      "            \n",
      "            parent_seq_unit_stats = data[parent_lib]['seq_units'][seq_unit_name]\n",
      "            #pprint.pprint(parent_seq_unit_stats)\n",
      "            parent_lib_read_count = int(parent_seq_unit_stats['rqc_stats']['read count'])\n",
      "            parent_lib_read_count = parent_seq_unit_stats['rqc_stats']['read count']\n",
      "            parent_lib_read_1_length = int(parent_seq_unit_stats['read_1_length'])\n",
      "            parent_lib_read_2_length = int(parent_seq_unit_stats['read_2_length'])\n",
      "            parent_lib_read1_q20_len = parent_seq_unit_stats['rqc_stats']['read q20 read1']\n",
      "            parent_lib_read2_q20_len = parent_seq_unit_stats['rqc_stats']['read q20 read2']\n",
      "            parent_lib_percent_rRNA = parent_seq_unit_stats['rqc_stats']['illumina read percent contamination rrna']\n",
      "            parent_lib_percent_contamination = float(parent_seq_unit_stats['rqc_stats']['illumina read percent contamination artifact']) + float(parent_seq_unit_stats['rqc_stats']['illumina read percent contamination contaminants'])\n",
      "            parent_lib_read_count_artifacts_filtered =  int(round(float(parent_lib_read_count)*( (int(100)-parent_lib_percent_contamination)/ 100)) )\n",
      "                \n",
      "            \n",
      "            #parent_lib_raw_based_in_gb =   \n",
      "            parent_lib_raw_based_in_gb =  round ( ( ( float(parent_lib_read_count) * float(parent_lib_read_1_length) ) / 1000000000),2)\n",
      "            parent_lib_cleaned_bases_in_gb = round( ( parent_lib_read_count_artifacts_filtered * float(parent_lib_read_1_length) / 1000000000),2)\n",
      "                \n",
      "                \n",
      "            print '%s \\t %s \\t %s \\t %s \\t %s \\t %s' % (parent_lib,'None',seq_unit_name,parent_lib_raw_based_in_gb,\n",
      "                                                        parent_lib_cleaned_bases_in_gb,parent_lib_percent_contamination)\n",
      "                \n",
      "            \n",
      "            #print parent_lib,'None',seq_unit_name,parent_lib_read_count,parent_lib_read1_q20_len,parent_lib_read2_q20_len\n",
      "            \n",
      "            #break\n",
      "        \n",
      "        for child_lib in data[parent_lib]['libs']:\n",
      "            total_per_child_lib = 0\n",
      "            \n",
      "            child_lib_read_count = 0\n",
      "            child_lib_read_count_artifact_filtered = 0\n",
      "            child_lib_raw_bases_in_gb = 0\n",
      "            child_lib_cleaned_bases_in_gb = 0\n",
      "            child_lib_percent_contamination = []\n",
      "            child_lib_percent_rRNA = []\n",
      "            \n",
      "            for child_seq_unit in data[parent_lib]['libs'][child_lib]['seq_units']:\n",
      "                child_seq_unit_stats = data[parent_lib]['libs'][child_lib]['seq_units'][child_seq_unit]\n",
      "                \n",
      "                child_lib_su_read_count = int(child_seq_unit_stats['rqc_stats']['read count'])\n",
      "                child_lib_su_read1_q20_len = child_seq_unit_stats['rqc_stats']['read q20 read1']\n",
      "                child_lib_su_read2_q20_len = child_seq_unit_stats['rqc_stats']['read q20 read2']\n",
      "                child_lib_su_read_1_length = int(child_seq_unit_stats['read_1_length'])\n",
      "                child_lib_su_read_2_length = int(child_seq_unit_stats['read_2_length'])\n",
      "                child_lib_su_percent_rRNA = float(child_seq_unit_stats['rqc_stats']['illumina read percent contamination rrna'])\n",
      "                child_lib_su_percent_contamination = round( float(child_seq_unit_stats['rqc_stats']['illumina read percent contamination artifact']) + float(child_seq_unit_stats['rqc_stats']['illumina read percent contamination contaminants']),2)\n",
      "                child_lib_su_read_count_artifacts_filtered =  int(round(float(child_lib_su_read_count)*( (int(100)-child_lib_su_percent_contamination)/ 100)) )\n",
      "                \n",
      "                child_lib_su_raw_based_in_gb = round( ( child_lib_su_read_count * float(child_lib_su_read_1_length) / 1000000000),2)\n",
      "                child_lib_su_cleaned_bases_in_gb = round( ( child_lib_su_read_count_artifacts_filtered * float(child_lib_su_read_1_length) / 1000000000),2)\n",
      "                \n",
      "                \n",
      "                child_lib_read_count += child_lib_su_read_count\n",
      "                child_lib_read_count_artifact_filtered += child_lib_su_read_count_artifacts_filtered\n",
      "                child_lib_raw_bases_in_gb += child_lib_su_raw_based_in_gb\n",
      "                child_lib_cleaned_bases_in_gb += child_lib_su_cleaned_bases_in_gb\n",
      "                child_lib_percent_contamination.append(child_lib_su_percent_contamination)\n",
      "                child_lib_percent_rRNA.append(child_lib_su_percent_rRNA)\n",
      "                \n",
      "                \n",
      "                '''\n",
      "                print '%s \\t %s \\t %s \\t %s \\t %s \\t %s \\t %s \\t %s \\t %s' % (parent_lib,child_lib,child_seq_unit,\n",
      "                                                                  round((child_lib_su_read_count/1000000.0),2),round((child_lib_su_read_count_artifacts_filtered/1000000.0),2),\n",
      "                                                                  child_lib_su_raw_based_in_gb,child_lib_su_cleaned_bases_in_gb,\n",
      "                                                                  child_lib_su_percent_contamination,child_lib_su_percent_rRNA)\n",
      "                '''       \n",
      "            \n",
      "            \n",
      "            print '%s \\t %s \\t %s \\t %s \\t %s \\t %s \\t %s \\t %s' % (parent_lib,child_lib,\n",
      "                                                                  round((child_lib_read_count/1000000.0),2),round((child_lib_read_count_artifact_filtered/1000000.0),2),\n",
      "                                                                  child_lib_raw_bases_in_gb,child_lib_cleaned_bases_in_gb,\n",
      "                                                              \n",
      "                                                              \n",
      "                                                                round(sum(child_lib_percent_contamination)/float(len(child_lib_percent_contamination)),2),\n",
      "                                                                round(sum(child_lib_percent_rRNA)/len(child_lib_percent_rRNA),2)\n",
      "                                                                ) \n",
      "        \n",
      "        \n",
      "\n",
      "\n",
      "    #print [ seq_unit, for parent_lib in data  for seq_unit in data[parent_lib]['seq_units'].keys() for x in data[parent_lib]['seq_units'][seq_unit] ]\n",
      "       \n",
      "            \n",
      "    \n",
      "    \n",
      "    \n",
      "summarize(data)\n",
      "\n",
      "#pprint.pprint(data)\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "BTYS \t None \t 6104.3.36427.fastq.gz \t 35.74 \t 34.21 \t 4.27917\n",
        "BTYS \t None \t 6104.1.36425.fastq.gz \t 36.03 \t 34.63 \t 3.90524\n",
        "BTYS \t None \t 6104.2.36426.fastq.gz \t 35.48 \t 33.93 \t 4.38042\n",
        "BTYS \t AUPA \t 44.45 \t 43.84 \t 6.66 \t 6.57 \t 1.38 \t 18.15\n",
        "BTYS \t AUOA \t 32.93 \t 32.2 \t 4.95 \t 4.84 \t 2.22 \t 18.12\n",
        "BTYS \t AUOB \t 43.09 \t 42.17 \t 6.46 \t 6.33 \t 2.14 \t 11.34\n",
        "BTYS \t AUOW \t 39.21 \t 38.65 \t 5.89 \t 5.8 \t 1.41 \t 10.09\n",
        "BTYS \t AUNW \t 31.84 \t 30.54 \t 4.77 \t 4.58 \t 4.09 \t 33.45\n",
        "BTYS \t AUOO \t 33.98 \t 33.17 \t 5.1 \t 4.98 \t 2.37 \t 21.3\n",
        "BTYS \t UNKNOWN \t 26.0 \t 18.21 \t 3.9 \t 2.74 \t 29.96 \t 13.52\n",
        "BTYS \t AUNN \t 28.07 \t 25.87 \t 4.2 \t 3.87 \t 7.84 \t 12.46\n",
        "BTYS \t AUNO \t 36.71 \t 35.92 \t 5.51 \t 5.39 \t 2.16 \t 13.93\n",
        "BTYS \t AUOS \t 32.47 \t 31.75 \t 4.87 \t 4.76 \t 2.22 \t 27.43\n",
        "BTYS \t AUOP \t 15.89 \t 15.26 \t 2.39 \t 2.29 \t 3.95 \t 11.59\n",
        "BTYS \t AUNT \t 38.15 \t 37.44 \t 5.72 \t 5.62 \t 1.85 \t 28.73\n",
        "BTYS \t AUNU \t 30.99 \t 29.91 \t 4.65 \t 4.48 \t 3.48 \t 10.74\n",
        "BTYS \t AUOU \t 36.81 \t 36.23 \t 5.53 \t 5.44 \t 1.59 \t 28.52\n",
        "BTYS \t AUOT \t 35.64 \t 34.67 \t 5.35 \t 5.2 \t 2.73 \t 11.87\n",
        "BTYS \t AUNX \t 35.88 \t 35.23 \t 5.38 \t 5.29 \t 1.8 \t 23.15\n",
        "BTYS \t AUNY \t 28.98 \t 26.38 \t 4.35 \t 3.96 \t 8.98 \t 41.04\n",
        "BTYS \t AUNZ \t 35.08 \t 34.52 \t 5.26 \t 5.18 \t 1.6 \t 14.05\n",
        "BTYS \t AUNS \t 28.83 \t 28.03 \t 4.32 \t 4.2 \t 2.78 \t 13.21\n",
        "BTYS \t AUCY \t 49.05 \t 47.8 \t 7.35 \t 7.17 \t 2.55 \t 26.38\n",
        "BTYS \t AUNP \t 31.05 \t 27.41 \t 4.65 \t 4.11 \t 11.71 \t 14.0\n"
       ]
      }
     ],
     "prompt_number": 254
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print sum(child_lib_percent_contamination)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'child_lib_percent_contamination' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-230-f5da114f4a87>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[1;32mprint\u001b[0m \u001b[0msum\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchild_lib_percent_contamination\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[1;31mNameError\u001b[0m: name 'child_lib_percent_contamination' is not defined"
       ]
      }
     ],
     "prompt_number": 230
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pprint.pprint(parent_seq_unit_stats)\n",
      "\n",
      "\n",
      "'''\n",
      "                ,child_lib_read_count_artifacts_filtered,\n",
      "                       ,cleaned_bases_in_gb\n",
      "                       ,child_lib_read1_q20_len,child_lib_read2_q20_len\n",
      "                       )\n",
      "            \n",
      "    '''\n",
      "        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'parent_seq_unit_stats' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-91-3d128e148ee5>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mpprint\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mparent_seq_unit_stats\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[1;31mNameError\u001b[0m: name 'parent_seq_unit_stats' is not defined"
       ]
      }
     ],
     "prompt_number": 91
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import math\n",
      "print (round(3.349384934,2))\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "3.35\n"
       ]
      }
     ],
     "prompt_number": 160
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "s=[1,2,3,4,5]\n",
      "sum(s)/len(s)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 216,
       "text": [
        "3"
       ]
      }
     ],
     "prompt_number": 216
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "'''\n",
      "6104.3.36427.fastq.gz {'platform_name': 'Illumina', 'instrument_type': 'HiSeq', 'file_name': '6104.3.36427.fastq.gz', \n",
      "'barcode': None, 'is_multiplexed': 1, 'physical_run_type': 'Paired-End Indexed', 'read_1_length': 150L, \n",
      "'seq_unit_name': '6104.3.36427.fastq.gz', \n",
      "\n",
      "        'rqc_stats': \n",
      "                    {'read bwa aligned duplicate percent': '23.3', 'demultiplex stats plot': \n",
      "                     '6104.3.36427.index_sequence_detection.png', 'base Q5': '91', 'read Q25': '89.65', \n",
      "                     'read tax species of refseq.microbial': '137', 'read Q20': '94.70', 'base Q20': '88', 'base Q25': '85', \n",
      "                       'read GC std': '9.40', 'illumina read percent contamination phix': '0', 'read N frequence': '0.24', \n",
      "                    'read qual pos plot merged': '6104.3.36427.s0.01.qrpt.png', \n",
      "                    'read tophit file of nt': 'megablast.reads.fa.v.nt.FmLD2a10p90E30JFfTITW45.parsed.tophit', \n",
      "                    'read screend': '2382583', 'read artifact percent': '0.0', 'illumina read percent contamination artifact': '3.00606', \n",
      "                    'read Q10': '98.62', 'read q20 read2': '137', 'read level': 'completed', 'read artifact': '0', \n",
      "                    'read base quality stats plot': '6104.3.36427.s0.01.fastq.base_qual.stats.png', \n",
      "                    'read 21mer window sliding plot': '6104.3.36427.s0.01.21mer.qual.png', 'read base count': '35743708800', \n",
      "                    'base C15': '317828752', 'base Q30': '78', 'base C10': '319645201', 'read bwa aligned duplicate': '555945', \n",
      "                    'read qhist plot': '6104.3.36427.s0.01.fastq.qhist.png', 'base Q15': '90', 'read count': '238291392', \n",
      "                    'read 20mer uniqueness text': '6104.3.36427.merSampler.m20.e25000', 'read matching hits of nt': '728799', \n",
      "                    'illumina read percent contamination mitochondrion': '1.66504', 'read qual pos qrpt 1': '6104.3.36427.s0.01.r1.qrpt', \n",
      "                    'read qual pos qrpt 2': '6104.3.36427.s0.01.r2.qrpt', 'base C25': '302227827', \n",
      "                    'read base percentage plot 1': '6104.3.36427.s0.01.r1.fastq.base.stats.Npercent.png', \n",
      "                    'read base percentage plot 2': '6104.3.36427.s0.01.r2.fastq.base.stats.Npercent.png', \n",
      "                    'illumina read percent contamination contaminants': '1.27311', 'read GC text hist': '6104.3.36427.s0.01.gc.hist',\n",
      "                    'fake read qual pos plot': '6104.3.36427.s0.01.index.qrpt.png', \n",
      "                    'read q20 read1': '134', \n",
      "                    'read taxlist file of nt': 'megablast.reads.fa.v.nt.FmLD2a10p90E30JFfTITW45.parsed.taxlist', \n",
      "                    'read length 2': '150', 'Q30 bases Q score mean': '36.2917', 'read top hits of nt': '7604', \n",
      "                    'read level MEGAN: Taxa Distribution of contigs vs. NT': '6104.3.36427.taxa.jpg', \n",
      "                    'read 20mer uniqueness plot': '6104.3.36427.merSampler.m20.e25000.png', \n",
      "                    'demultiplex stats': '6104.3.36427.demultiplex_stats', 'read GC mean': '47.99', \n",
      "                    'read matching hits of refseq.microbial': '12590', 'base C30': '278972589', \n",
      "                    'read 21mer window sliding text data': '6104.3.36427.s0.01.21mer.qual', \n",
      "                    'demultiplex filtered stats': '6104.3.36427.filtered_demultiplex_stats', \n",
      "                    'read parsed file of nt': 'megablast.reads.fa.v.nt.FmLD2a10p90E30JFfTITW45.parsed', \n",
      "                    'read taxlist file of refseq.microbial': 'megablast.reads.fa.v.refseq.microbial.FmLD2a10p90E30JFfTITW45.parsed.taxlist', \n",
      "                    'base Q10': '90', 'read length 1': '150', 'read base count plot 2': '6104.3.36427.s0.01.r2.fastq.base.stats.png', \n",
      "                    'read tax species of nt': '308', 'read base quality stats': '6104.3.36427.s0.01.fastq.base_qual.stats', \n",
      "                    'read tophit file of refseq.microbial': 'megablast.reads.fa.v.refseq.microbial.FmLD2a10p90E30JFfTITW45.parsed.tophit', \n",
      "                    'illumina read percent contamination fosmid': '0', 'illumina read percent contamination rrna': '19.3655', \n",
      "                    'read level MEGAN: Phylogeny of contigs vs NT filtered by class': '6104.3.36427.phylo2.jpg', \n",
      "                    'read 20mer sample size': '238275000', 'overall bases Q score mean': '31.3062', \n",
      "                    'read level MEGAN: Phylogeny of contigs vs NT': '6104.3.36427.phylo.jpg', \n",
      "                    'read base count plot 1': '6104.3.36427.s0.01.r1.fastq.base.stats.png', \n",
      "                    'illumina read sciclone DNA count file': '6104.3.36427.s0.01.fastq.trim37.GC.standards.well.fa.aln.sam.count.txt', \n",
      "                    'read tagdust coverage cutoff': '100.0', 'Q30 bases Q score std': '3.01621', \n",
      "                    'illumina read percent contamination plastid': '1.64137', \n",
      "                    'read qhist text': '6104.3.36427.s0.01.fastq.qhist', \n",
      "                    'read parsed file of refseq.microbial': 'megablast.reads.fa.v.refseq.microbial.FmLD2a10p90E30JFfTITW45.parsed', \n",
      "                    'base C20': '310938645', 'read base count text 1': '6104.3.36427.s0.01.r1.fastq.base.stats',\n",
      "                     'read base count text 2': '6104.3.36427.s0.01.r2.fastq.base.stats', 'read Q5': '99.29', 'base C5': '323083905', \n",
      "                    'read 20mer percentage random mers': '44.46', 'Overall bases Q score std': '11.1185', \n",
      "                    'read Q30': '75.78', 'read Q15': '97.32', 'read 20mer percentage starting mers': '32.89', \n",
      "                    'read top hits of refseq.microbial': '1882', 'reads number': '49868', 'read GC plot': '6104.3.36427.s0.01.gc.hist.png', \n",
      "                    'read bwa aligned': '2382583', 'illumina read percent contamination ecoli combined': '0.0771012'}, \n",
      "                    'lib_name': 'BTYS', 'rqc_analysis_id': 67175L, 'read_2_length': 150L}\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "      \n",
      "    \n",
      "     ''' \n",
      "        #new child of new parent fastq\n",
      "        if data.get(parent_sdm_seq_unit_id,None) is None:\n",
      "            #new_parent_fastq\n",
      "            data[parent_sdm_seq_unit_id] = {}\n",
      "            data[parent_sdm_seq_unit_id]['childs'] = {}\n",
      "            #new parent's new child fastq\n",
      "            data[parent_sdm_seq_unit_id]['childs'][sdm_seq_unit_id] = {}\n",
      "        \n",
      "        #new child of existing parent fastq\n",
      "        elif data[parent_sdm_seq_unit_id].get(sdm_seq_unit_id, None) is None:\n",
      "                data[parent_sdm_seq_unit_id]['childs'][sdm_seq_unit_id] = {}\n",
      "                \n",
      "        else:\n",
      "            pass\n",
      "            \n",
      "            #data[parent_sdm_seq_unit_id][sdm_seq_unit_id] = {}\n",
      "    \n",
      "        \n",
      "    \n",
      "        data[sdm_seq_unit_id]['lib_name'] = lib_name\n",
      "            data[sdm_seq_unit_id]['barcode'] = barcode\n",
      "            data[sdm_seq_unit_id]['seq_unit_name'] = seq_unit_name\n",
      "            data[sdm_seq_unit_id]['is_multiplexed'] = is_multiplexed\n",
      "        \n",
      "    \n",
      "    \n",
      "    \n",
      "    data[sdm_seq_unit_id]['physical_run_type'] = physical_run_type\n",
      "            data[sdm_seq_unit_id]['platform_name'] = platform_name\n",
      "            data[sdm_seq_unit_id]['instrument_type'] = instrument_type\n",
      "            data[sdm_seq_unit_id]['read_1_length'] = read_1_length\n",
      "            data[sdm_seq_unit_id]['read_2_length'] = read_2_length\n",
      "            data[sdm_seq_unit_id]['rqc_analysis_id'] = rqc_analysis_id\n",
      "            data[sdm_seq_unit_id]['childs'] = {}\n",
      "\n",
      "\n",
      "\n",
      "        data[sdm_seq_unit_id]['file_name'] = seq_unit_name \n",
      "        data[sdm_seq_unit_id]['lib_name'] = lib_name\n",
      "        data[sdm_seq_unit_id]['barcode'] = barcode\n",
      "        data[sdm_seq_unit_id]['seq_unit_name'] = seq_unit_name\n",
      "        data[sdm_seq_unit_id]['is_multiplexed'] = is_multiplexed\n",
      "        data[sdm_seq_unit_id]['physical_run_type'] = physical_run_type\n",
      "        data[sdm_seq_unit_id]['platform_name'] = platform_name\n",
      "        data[sdm_seq_unit_id]['instrument_type'] = instrument_type\n",
      "        data[sdm_seq_unit_id]['read_1_length'] = read_1_length\n",
      "        data[sdm_seq_unit_id]['read_2_length'] = read_2_length\n",
      "        data[sdm_seq_unit_id]['rqc_analysis_id'] = rqc_analysis_id\n",
      "\n",
      "    \n",
      "        #new fastq\n",
      "        data[sdm_seq_unit_id] = {}\n",
      "        data[sdm_seq_unit_id]['file_name'] = seq_unit_name \n",
      "        data[sdm_seq_unit_id]['lib_name'] = lib_name\n",
      "        data[sdm_seq_unit_id]['barcode'] = barcode\n",
      "        data[sdm_seq_unit_id]['seq_unit_name'] = seq_unit_name\n",
      "        data[sdm_seq_unit_id]['is_multiplexed'] = is_multiplexed\n",
      "        data[sdm_seq_unit_id]['physical_run_type'] = physical_run_type\n",
      "        data[sdm_seq_unit_id]['platform_name'] = platform_name\n",
      "        data[sdm_seq_unit_id]['instrument_type'] = instrument_type\n",
      "        data[sdm_seq_unit_id]['read_1_length'] = read_1_length\n",
      "        data[sdm_seq_unit_id]['read_2_length'] = read_2_length\n",
      "        data[sdm_seq_unit_id]['rqc_analysis_id'] = rqc_analysis_id\n",
      "        #data[sdm_seq_unit_id]['rqc_stats_name'] = rqc_stats_name\n",
      "        #data[sdm_seq_unit_id]['rqc_stats_value'] = rqc_stats_value\n",
      "           \n",
      "    else:\n",
      "        pass\n",
      "        \n",
      "    '''"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print seq_unit_name"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "seq_unit_name.find('Fsds')\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}