{
 "metadata": {
  "name": "rna_pet_linker_cleaning"
 }, 
 "nbformat": 2, 
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "from Bio.Seq import Seq", 
      "from Bio import SeqIO", 
      "import string", 
      "import os"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 80
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "linker='CTGCTGTACCGTACATCCGCCTTGGCCGTACAGCAG'", 
      "adapter_3_prime='GAGCTAGTTCTGGAGTTTTTTTTTTTTTTTT'", 
      "adapter_5_prime='TCCTACCTGGAGAACATGAGGCAGCCAG'", 
      "", 
      "", 
      "", 
      "def revcomp(sequence):", 
      "    revcomp_seq = sequence[::-1]", 
      "    trans_table = string.maketrans('AGCTagct','TCGATCGA') ", 
      "    revcomp_seq = revcomp_seq.translate(trans_table)", 
      "    return revcomp_seq", 
      "", 
      "", 
      "def get_kmers_list(sequence,kmer_len):", 
      "    kmers_list = []", 
      "    sequence = sequence.strip().upper()", 
      "    seq_revcomp = revcomp(sequence)", 
      "    #print 'Seq:%s\\nRev:%s \\n' % (sequence,seq_revcomp)", 
      "    ", 
      "    #generate list of k-mers for sequqence", 
      "    [ kmers_list.append(sequence[x:x+kmer_len]) for x in xrange(len(sequence)-kmer_len+1)]", 
      "        ", 
      "    #generate list of k-mers for rev comp sequqence", 
      "    [ kmers_list.append(seq_revcomp[x:x+kmer_len]) for x in xrange(len(sequence)-kmer_len+1)]", 
      "    return  kmers_list", 
      "", 
      "", 
      "", 
      "", 
      "def get_kmers_dict(sequence,kmer_len):", 
      "    kmers_dict = {}", 
      "    sequence = sequence.strip().upper()", 
      "    seq_revcomp = revcomp(sequence)", 
      "    #print 'Seq:%s\\nRev:%s \\n' % (sequence,seq_revcomp)", 
      "    ", 
      "    #generate dict of k-mers for sequqence", 
      "    [ kmers_dict.update( {sequence[x:x+kmer_len]:1} )  for x in xrange(len(sequence)-kmer_len+1)]", 
      "        ", 
      "    #generate dict of k-mers for rev comp sequqence", 
      "    [ kmers_dict.update( {seq_revcomp[x:x+kmer_len]:1} ) for x in xrange(len(sequence)-kmer_len+1)]", 
      "    return  kmers_dict", 
      "", 
      "", 
      "", 
      "", 
      "def get_kmers_dict_with_one_error(sequence,kmer_len):", 
      "    kmers_dict = {}", 
      "    sequence = sequence.strip().upper()", 
      "    seq_revcomp = revcomp(sequence)", 
      "    #print 'Seq:%s\\nRev:%s \\n' % (sequence,seq_revcomp)", 
      "    ", 
      "    #generate dict of k-mers for sequqence with one error", 
      "    [ kmers_dict.update( mutate_kmer_with_one_error( sequence[x:x+kmer_len] ) )  for x in xrange(len(sequence)-kmer_len+1) ]", 
      "        ", 
      "    #generate dict of k-mers for rev comp sequqence with one error", 
      "    [ kmers_dict.update( mutate_kmer_with_one_error( seq_revcomp[x:x+kmer_len] ) ) for x in xrange(len(sequence)-kmer_len+1)]", 
      "    return  kmers_dict", 
      "", 
      "", 
      "", 
      "def do_kmer_match(query,kmers_list,kmer_len):", 
      "    query=query.upper()", 
      "    for x in xrange(len(query)-kmer_len+1):", 
      "        if query[x:x+kmer_len] in kmers_list:", 
      "            match_position = x + 1", 
      "            return (match_position,query[x:x+kmer_len])", 
      "    return (None,None)", 
      "", 
      "", 
      "", 
      "###kmer based matching using subject kmers as dictionary", 
      "###test method", 
      "def do_kmer_dict_match(query,kmers_dict,kmer_len):", 
      "    query=query.upper()", 
      "    for x in xrange(len(query)-kmer_len+1):", 
      "        if kmers_dict.get(query[x:x+kmer_len],None) is not None:", 
      "            match_position = x + 1 ", 
      "            return (match_position,query[x:x+kmer_len])", 
      "    return (None,None)", 
      "    ", 
      "", 
      "", 
      "", 
      "def mutate_kmer_with_one_error(kmer):", 
      "    SUB='gen_kmers_with_one_error'", 
      "    ", 
      "    temp_dict = {}", 
      "    for x in xrange(len(kmer)):", 
      "        #print x", 
      "        prefix = kmer[:x]", 
      "        suffix = kmer[x+1:]", 
      "        temp_dict.update([ (prefix+base+suffix,1) for base in 'ACGTN'])", 
      "    #print temp_dict", 
      "    #print (len(temp_dict))", 
      "    return temp_dict"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 101
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "kmers=get_kmers_dict('AGCT',4)", 
      "", 
      "kmers=get_kmers_dict_with_one_error('AGCT',4)", 
      "", 
      "print kmers", 
      "", 
      "print do_kmer_dict_match('aaaaaaaaaaaaaaaaaaa',kmers,4)", 
      "", 
      "#import collections", 
      "#c = collections.Counter()", 
      "", 
      "#c=check('AGCT')"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 82
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "############################", 
      "### Testing kmer match using list and dict and with one error in kmer matching", 
      "### performance looks similar", 
      "############################", 
      "", 
      "", 
      "from timeit import Timer", 
      "import timeit", 
      "kmers=get_kmers_list('ACGTGATCGTATTACGATCGTAGCTGATCGTAGCGATCGTACTGAC',14)", 
      "t = Timer(lambda: do_kmer_match('ATCGTATTACGATCGTAGCTGATCG',kmers,14))", 
      "print 'using list kmer match time %s' % t.timeit(number=1000000)", 
      "", 
      "", 
      "#converting the kmers list to dict", 
      "kmers_dict=get_kmers_dict('ACGTGATCGTATTACGATCGTAGCTGATCGTAGCGATCGTACTGAC',14)", 
      "t = Timer(lambda: do_kmer_dict_match('ATCGTATTACGATCGTAGCTGATCG',kmers_dict,14))", 
      "print 'using dict kmer match time %s' % t.timeit(number=1000000)", 
      "", 
      "# kmers with one error and using dict method to match", 
      "kmers_dict=get_kmers_dict_with_one_error('ACGTGATCGTATTACGATCGTAGCTGATCGTAGCGATCGTACTGAC',14)", 
      "t = Timer(lambda: do_kmer_dict_match('ATCGTATTACGATCGTAGCTGATCG',kmers_dict,14))", 
      "print 'using dict kmer(with one error) match time %s' % t.timeit(number=1000000)"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 83
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "###############", 
      "###tested the IO time for SeqIO and native method of reading a fastq", 
      "###############", 
      "", 
      "", 
      "#sample file", 
      "sample_fastq='/house/groupdirs/genetic_analysis/rna/projects/Escherichia_coli_K-12_MG1655_405338/CSBZ_Mplexed_library/CSBU/linker_removal/test/test_40_reads.fastq'", 
      "", 
      "#print check(sample_fastq)", 
      "t= Timer(lambda: check(sample_fastq))", 
      "print t.timeit(number=10)", 
      "", 
      "", 
      "", 
      "def check1(file_name):", 
      "    count = 0", 
      "    for seq_record in SeqIO.parse(file_name,\"fastq\"):", 
      "        count +=1", 
      "    return count", 
      "#print check1(sample_fastq)", 
      "t= Timer(lambda: check1(sample_fastq))", 
      "print t.timeit(number=10)", 
      ""
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 84
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "##checking the k-mer overlap between 3' and 5' adapter and linker", 
      "", 
      "", 
      "linker='CTGCTGTACCGTACATCCGCCTTGGCCGTACAGCAG'", 
      "adapter_3_prime='GAGCTAGTTCTGGAGTTTTTTTTTTTTTTTT'", 
      "adapter_5_prime='TCCTACCTGGAGAACATGAGGCAGCCAG'", 
      "", 
      "", 
      "print \"Len of linker is \", len(linker)", 
      "print \"Len of 3' adapter is \", len(adapter_3_prime)", 
      "print \"Len of 5' adapter is \", len(adapter_5_prime)", 
      "", 
      "", 
      "", 
      "#comparing linker to 3 prime adapter", 
      "for x in xrange(10,30,2):", 
      "    kmer_len = x", 
      "    kmers=get_kmers_list(linker,kmer_len)", 
      "    print do_kmer_match(adapter_3_prime,kmers,kmer_len)", 
      "", 
      "#comparing linker to 5 prime adapter", 
      "for x in xrange(10,30,2):", 
      "    kmer_len = x", 
      "    kmers=get_kmers_list(linker,kmer_len)", 
      "    print do_kmer_match(adapter_5_prime,kmers,kmer_len)", 
      "", 
      "", 
      "#comparing 3_prime adater to 5 prime adapter", 
      "for x in xrange(10,30,2):", 
      "    kmer_len = x", 
      "    kmers=get_kmers_list(linker,kmer_len)", 
      "    print do_kmer_match(adapter_5_prime,kmers,kmer_len)"
     ], 
     "language": "python", 
     "outputs": []
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "def write_read_pair_to_fastq(*args):", 
      "    \"\"\"takes the read pair information and write to fastq", 
      "    \"\"\"", 
      "    ", 
      "    if (len(args) != 10):", 
      "        print '[%s]: Error -> number of variables passed not equal to 10'", 
      "        return False", 
      "    else:", 
      "        ", 
      "        fh_read_1            = args[0]", 
      "        read_1_seq           = args[2]", 
      "        read_1_qual          = args[3]", 
      "        read_1_direction     = args[4]", 
      "        fh_read_2            = args[5]", 
      "        read_2_seq           = args[7]", 
      "        read_2_qual          = args[8]", 
      "        read_2_direction     = args[9]", 
      "        ", 
      "        ", 
      "        read_1_header        = args[1].replace('/1','_dir=' + read_1_direction + '/1')", 
      "        read_2_header        = args[6].replace('/2','_dir=' + read_2_direction + '/2')", 
      "        ", 
      "        read_1_full  = '%s\\n%s\\n%s\\n%s\\n' % (read_1_header,read_1_seq,'+',read_1_qual)", 
      "        read_2_full  = '%s\\n%s\\n%s\\n%s\\n' % (read_2_header,read_2_seq,'+',read_2_qual)", 
      "        ", 
      "        fh_read_1.write(read_1_full)", 
      "        fh_read_2.write(read_2_full)", 
      "        ", 
      "    ", 
      "        ", 
      "        ", 
      "        ", 
      "        ", 
      "        ", 
      "        ", 
      "def get_min_number(*numbers):", 
      "    \"\"\"", 
      "    given a list/tuple with numbers and None", 
      "    return the minimum number", 
      "    None is not used", 
      "    \"\"\"", 
      "    ", 
      "    return min(x for x in (numbers) if x is not None)", 
      "        ", 
      "    ", 
      "def debug(*vars):", 
      "    ", 
      "    ", 
      "    read_1_header = vars[0]", 
      "    read_1_seq = vars[1]", 
      "    read_1_5_prime_match = vars[2]", 
      "    read_1_3_prime_match = vars[3]", 
      "    read_1_linker_match = vars[4]", 
      "    read_1_5_prime_kmer = vars[5]", 
      "    read_1_3_prime_kmer = vars[6]", 
      "    read_1_linker_kmer  = vars[7]", 
      "    ", 
      "    ", 
      "    read_2_header = vars[8]", 
      "    read_2_seq = vars[9]", 
      "    read_2_5_prime_match = vars[10]", 
      "    read_2_3_prime_match = vars[11]", 
      "    read_2_linker_match = vars[12]", 
      "    read_2_5_prime_kmer = vars[13]", 
      "    read_2_3_prime_kmer =vars[14]", 
      "    read_2_linker_kmer  = vars[15]", 
      "    result              = vars[16]", 
      "    ", 
      "", 
      "    print \"Read 1: %s \" % read_1_header", 
      "    print \"Read 1: %s \" % read_1_seq", 
      "    print \"Read 1: 5'-> %s , 3'->%s, linker-> %s\"      % (read_1_5_prime_match,read_1_3_prime_match, read_1_linker_match)", 
      "    print \"Read 1: 5' Kmer : %s, 3' Kmer: %s, Linker kmer : %s \" % (read_1_5_prime_kmer,read_1_3_prime_kmer,read_1_linker_kmer)", 
      "    ", 
      "    print \"Read 2: %s \" % read_2_header", 
      "    print \"Read 2: %s \" % read_2_seq", 
      "    print \"Read 2: 5'-> %s , 3'->%s, linker-> %s\"      % (read_2_5_prime_match,read_2_3_prime_match, read_2_linker_match)", 
      "    print \"Read 2: 5' Kmer : %s, 3' Kmer: %s, Linker kmer : %s \" % (read_2_5_prime_kmer,read_2_3_prime_kmer,read_2_linker_kmer)", 
      "    print \"Result: %s\" % result", 
      "    print '----------------------------'  ", 
      "    ", 
      "", 
      "    "
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 146
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "", 
      "", 
      "def classify_reads(fastq_file,kmers_5_prime,kmers_3_prime,kmer_len,fh_read_1,fh_read_2,verbose=0,min_trimmed_read_length=25):", 
      "    '''", 
      "       Read a paired fastq file and ", 
      "       and bin the reads as coming 5' or 3'", 
      "    '''", 
      "    ", 
      "    SUB='classify_reads'", 
      "    ", 
      "    #various counters", 
      "    count = 0", 
      "    count_trimer_pairs = 0", 
      "    count_ambiguous_pairs = 0", 
      "    count_pairs_with_no_adapter = 0", 
      "    count_pairs_with_no_adapter_but_linker_found = 0", 
      "    count_pairs_with_full_dir_determined = 0", 
      "    count_pairs_with_partial_dir_determined =0", 
      "    count_read_pair_lt_min_length = 0", 
      "    ", 
      "    trim_position_read_1 = 0", 
      "    trim_position_read_2 = 0", 
      "    ", 
      "    ", 
      "    fh = open(fastq_file,'r')", 
      "    for line in fh:", 
      "        ", 
      "        if (count % 1000000 == 0) and (count != 0):", 
      "            print '[%s]: processed %d million pairs' % (SUB, count/1000000)", 
      "        ", 
      "        ", 
      "        if not line.startswith('@'):", 
      "            print 'File %s doesnt seem to be paired fastq formatted ' % file_name", 
      "        else:", 
      "            read_1_header  = line.strip()", 
      "            read_1_seq      = fh.next().strip()", 
      "            fh.next() #skip quality header", 
      "            read_1_qual     = fh.next().strip()", 
      "            ", 
      "            read_2_header  = fh.next().strip()", 
      "            read_2_seq      = fh.next().strip()", 
      "            fh.next() #skip quality header", 
      "            read_2_qual     = fh.next().strip()", 
      "            ", 
      "            ", 
      "            ##do the kmer based matching : binning reads", 
      "            ", 
      "            #read1", 
      "            (read_1_5_prime_match,read_1_5_prime_kmer) = do_kmer_dict_match(read_1_seq,kmers_5_prime,kmer_len)", 
      "            (read_1_3_prime_match,read_1_3_prime_kmer) = do_kmer_dict_match(read_1_seq,kmers_3_prime,kmer_len)", 
      "            (read_1_linker_match,read_1_linker_kmer)   = do_kmer_dict_match(read_1_seq,kmers_linker,kmer_len)", 
      "            ", 
      "            #read2", 
      "            (read_2_5_prime_match,read_2_5_prime_kmer) = do_kmer_dict_match(read_2_seq,kmers_5_prime,kmer_len)", 
      "            (read_2_3_prime_match,read_2_3_prime_kmer) = do_kmer_dict_match(read_2_seq,kmers_3_prime,kmer_len)", 
      "            (read_2_linker_match,read_2_linker_kmer)   = do_kmer_dict_match(read_2_seq,kmers_linker,kmer_len)", 
      "            ", 
      "            #increment the read_pair count by 1", 
      "            count += 1", 
      "            ", 
      "        ", 
      "            ############", 
      "            #conditions", 
      "            ############", 
      "            ", 
      "            #if a read has all three 5 and 3 prime adapter and linker", 
      "            if ( ", 
      "                read_1_5_prime_match is not None", 
      "                and ", 
      "                read_1_3_prime_match is not None", 
      "                and ", 
      "                read_1_linker_match is not None ) \\", 
      "                or \\", 
      "                ( ", 
      "                 read_2_5_prime_match is not None", 
      "                 and ", 
      "                 read_2_3_prime_match is not None ", 
      "                 and ", 
      "                 read_2_linker_match is not None", 
      "                ):", 
      "                ", 
      "                count_trimer_pairs += 1", 
      "                result = '[%s]: Trimer pair ' % (SUB)", 
      "                ", 
      "            ", 
      "            #if read 1 or read 2 match the same direction", 
      "            elif ( read_1_5_prime_match and read_2_5_prime_match  is not None) or( read_1_3_prime_match and read_2_3_prime_match  is not None):", 
      "                count_ambiguous_pairs += 1", 
      "                result = '[%s]: Ambiguous pair ' % (SUB)", 
      "                ", 
      "", 
      "                ", 
      "            #if either adapter is not found on read 1 or 2", 
      "            elif (", 
      "                  read_1_5_prime_match is None  ", 
      "                  and ", 
      "                  read_1_3_prime_match is None ", 
      "                  and ", 
      "                  read_2_5_prime_match is None ", 
      "                  and ", 
      "                  read_2_3_prime_match is None", 
      "                  ):", 
      "                #but if linker is found", 
      "                if ( read_1_linker_match or read_2_linker_match):", 
      "                    count_pairs_with_no_adapter_but_linker_found +=1", 
      "                    result = '[%s]:  No adapter pair, linker found ' % (SUB)", 
      "                    ", 
      "                else:", 
      "                    count_pairs_with_no_adapter +=1", 
      "                    result = '[%s]:  No adapter/linker pair ' % (SUB)", 
      "                    ", 
      "            #if direction could be determined in some way ", 
      "            else:", 
      "                ", 
      "                trim_position_read_1= get_min_number(read_1_5_prime_match, read_1_3_prime_match, read_1_linker_match,len(read_1_seq))", 
      "                trim_position_read_2= get_min_number(read_2_5_prime_match, read_2_3_prime_match, read_2_linker_match,len(read_2_seq))", 
      "                ", 
      "                trimmed_read_1_seq = read_1_seq[:trim_position_read_1 - 1]  #0-based coordinate system again", 
      "                trimmed_read_2_seq = read_2_seq[:trim_position_read_2 - 1]  #0-based coordinate system again", 
      "                    ", 
      "                trimmed_read_1_qual = read_1_qual[:trim_position_read_1 - 1] #0-based coordinate system again", 
      "                trimmed_read_2_qual = read_2_qual[:trim_position_read_2 - 1] #0-based coordinate system again", 
      "                ", 
      "                ", 
      "                ##if the trimmed read pair is greater than the min_trimmed_read_length", 
      "                if ( min(trim_position_read_1,trim_position_read_2) >= min_trimmed_read_length ):", 
      "         ", 
      "                    #read 1 is from 3 prime and read 2 is from 5 prime", 
      "                    if ( read_1_3_prime_match and read_2_5_prime_match ):", 
      "                            count_pairs_with_full_dir_determined += 1", 
      "                            result = '[%s]:  Full dir deter : trim_length read1:%d, read 2:%d ' % (SUB,trim_position_read_1,trim_position_read_2)", 
      "                            read_1_direction = '3_prime'", 
      "                            read_2_direction = '5_prime'", 
      "                            write_read_pair_to_fastq( fh_read_1, read_1_header, trimmed_read_1_seq, trimmed_read_1_qual, read_1_direction, ", 
      "                                                      fh_read_2, read_2_header, trimmed_read_2_seq, trimmed_read_2_qual, read_2_direction )", 
      "                    ", 
      "                    #read 1 is from 5 prime and read 2 is from 3 prime", 
      "                    elif ( read_1_5_prime_match  and read_2_3_prime_match ) :", 
      "                        count_pairs_with_full_dir_determined += 1", 
      "                        result = '[%s]:  Full dir deter : trim_length read1:%d, read 2:%d ' % (SUB,trim_position_read_1,trim_position_read_2)", 
      "                        read_1_direction = '5_prime'", 
      "                        read_2_direction = '3_prime'", 
      "                        write_read_pair_to_fastq( fh_read_1, read_1_header, trimmed_read_1_seq, trimmed_read_1_qual, read_1_direction, ", 
      "                                                  fh_read_2, read_2_header, trimmed_read_2_seq, trimmed_read_2_qual, read_2_direction )", 
      "                            ", 
      "                    #if direction could be determined for one read in a pair", 
      "                    elif (read_1_5_prime_match or read_1_3_prime_match or  read_2_5_prime_match or read_2_3_prime_match is not None):", 
      "                        count_pairs_with_partial_dir_determined += 1", 
      "                        result = '[%s]:  partial direction trim_length read1:%s, read 2:%s ' % (SUB,trim_position_read_1,trim_position_read_2)", 
      "                            ", 
      "                        ", 
      "                        #if read 1 is from 5 prime", 
      "                        if (read_1_5_prime_match):", 
      "                            read_1_direction = '5_prime'", 
      "                            read_2_direction = '3_prime'", 
      "                         ", 
      "                            ", 
      "                        #if read 1 is from 3 prime", 
      "                        elif (read_1_3_prime_match):", 
      "                            read_1_direction = '3_prime'", 
      "                            read_2_direction = '5_prime'", 
      "                        ", 
      "                        ", 
      "                        #if read 2 is from 5 prime", 
      "                        if (read_2_5_prime_match):", 
      "                            read_1_direction = '3_prime'", 
      "                            read_2_direction = '5_prime'", 
      "                                            ", 
      "                        #if read 2 is from 3 prime", 
      "                        elif (read_2_3_prime_match):", 
      "                            read_1_direction = '5_prime'", 
      "                            read_2_direction = '3_prime'", 
      "                                                ", 
      "                        ", 
      "                        write_read_pair_to_fastq( fh_read_1, read_1_header, trimmed_read_1_seq, trimmed_read_1_qual, read_1_direction, ", 
      "                                                  fh_read_2, read_2_header, trimmed_read_2_seq, trimmed_read_2_qual, read_2_direction )", 
      "                        ", 
      "                    ", 
      "                                                    ", 
      "                    else:", 
      "                        result = \"[%s]: Can't determine the pair condition \" % (SUB)", 
      "                        debug( ", 
      "                              read_1_header, read_1_seq, read_1_5_prime_match, read_1_3_prime_match,read_1_linker_match,", 
      "                              read_1_5_prime_kmer, read_1_3_prime_kmer,read_1_linker_kmer,", 
      "                              read_2_header, read_2_seq, read_2_5_prime_match, read_2_3_prime_match,read_2_linker_match,", 
      "                              read_2_5_prime_kmer, read_2_3_prime_kmer,read_2_linker_kmer,", 
      "                              result", 
      "                              );", 
      "                else:", 
      "                    count_read_pair_lt_min_length += 1", 
      "                      ", 
      "                                                        ", 
      "            if verbose:", 
      "                debug( ", 
      "                      read_1_header, read_1_seq, read_1_5_prime_match, read_1_3_prime_match,read_1_linker_match,", 
      "                      read_1_5_prime_kmer, read_1_3_prime_kmer,read_1_linker_kmer,", 
      "                      read_2_header, read_2_seq, read_2_5_prime_match, read_2_3_prime_match,read_2_linker_match,", 
      "                      read_2_5_prime_kmer, read_2_3_prime_kmer,read_2_linker_kmer,", 
      "                      result", 
      "                      )", 
      "                      ", 
      "     ", 
      "             ", 
      "    #print summary", 
      "    print '[%s]: kmer_len %s' % (SUB,kmer_len)", 
      "    print '[%s]: Pairs %s ' % (SUB,count)", 
      "    print '[%s]: Trimer  %s' % (SUB, count_trimer_pairs)", 
      "    print '[%s]: Ambiguous  %s' % (SUB, count_ambiguous_pairs)", 
      "    print '[%s]: No Adapter %s' % (SUB, count_pairs_with_no_adapter)", 
      "    print '[%s]: No Adapter but linker %s' % (SUB, count_pairs_with_no_adapter_but_linker_found)", 
      "    print '[%s]: Good Pairs less than min threshold %s' % ( SUB, count_read_pair_lt_min_length)", 
      "    print '[%s]: Full dir determined %s' % ( SUB,count_pairs_with_full_dir_determined)", 
      "    print '[%s]: Partial dir determined %s' % ( SUB, count_pairs_with_partial_dir_determined)", 
      "    print '------------------------------------------'", 
      "    print \"\\n\\n\"", 
      "    ", 
      "    ##END Function", 
      "    ", 
      "    ", 
      "    ", 
      "", 
      "base_analysis_dir = '/house/groupdirs/genetic_analysis/rna/projects/Escherichia_coli_K-12_MG1655_405338/CSBZ_Mplexed_library/CSBU/linker_removal/'", 
      "#sample file", 
      "sample_fastq='test/test_200k_reads.fastq'", 
      "production_fastq='CSBU.2390.4.1914.GATCAG_qtrim_jartifacts.rRNA_cleaned.fastq'", 
      "fastq_path=base_analysis_dir + production_fastq", 
      "dirname = os.path.dirname(fastq_path)", 
      "    ", 
      "\"\"\"", 
      "kmer_len=24", 
      "kmers_5_prime=get_kmers_list(adapter_5_prime,kmer_len)", 
      "kmers_3_prime=get_kmers_list(adapter_3_prime,kmer_len)", 
      "kmers_linker=get_kmers_list(linker,kmer_len)", 
      "classify_reads(fastq_path,kmers_5_prime,kmers_3_prime,kmer_len)", 
      "\"\"\"", 
      "", 
      "             ", 
      "verbosity=0", 
      "min_trimmed_read_length = 15", 
      "", 
      "for x in xrange(14,16,2):", 
      "    kmer_len=x", 
      "    outfile_read_1 = dirname + '/cleaned_directional_read_1_kmer_'+ str(kmer_len) +'.fastq'", 
      "    outfile_read_2 = dirname + '/cleaned_directional_read_2_kmer_'+ str(kmer_len)  +'.fastq'", 
      "    fh_read_1 = open(outfile_read_1,'w')", 
      "    fh_read_2 = open(outfile_read_2,'w')", 
      "    ", 
      "    kmers_5_prime = get_kmers_dict_with_one_error(adapter_5_prime,kmer_len)", 
      "    kmers_3_prime = get_kmers_dict_with_one_error(adapter_3_prime,kmer_len)", 
      "    kmers_linker  = get_kmers_dict_with_one_error(linker,kmer_len)", 
      "    ", 
      "    t = Timer(lambda: classify_reads(fastq_path,kmers_5_prime,kmers_3_prime,kmer_len,fh_read_1,fh_read_2,verbose=verbosity, min_trimmed_read_length=min_trimmed_read_length))", 
      "    print 'Took %3.2f time' % t.timeit(number=1)", 
      "    fh_read_1.close()", 
      "    fh_read_2.close()"
     ], 
     "language": "python", 
     "outputs": [
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "[classify_reads]: processed 1 million pairs", 
        "[classify_reads]: processed 2 million pairs"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "[classify_reads]: processed 3 million pairs"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "[classify_reads]: processed 4 million pairs"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "[classify_reads]: kmer_len 14"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "[classify_reads]: Pairs 4551722 ", 
        "[classify_reads]: Trimer  2696976", 
        "[classify_reads]: Ambiguous  302301", 
        "[classify_reads]: No Adapter 66510", 
        "[classify_reads]: No Adapter but linker 6239", 
        "[classify_reads]: Good Pairs less than min threshold 641250", 
        "[classify_reads]: Full dir determined 149868", 
        "[classify_reads]: Partial dir determined 688578", 
        "------------------------------------------", 
        "", 
        "", 
        "", 
        "Took 880.27 time"
       ]
      }
     ], 
     "prompt_number": 154
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "####generate histogram of trimmed read lengths", 
      "", 
      "import numpy as np", 
      "import matplotlib.pyplot as plot", 
      "", 
      "", 
      "random_numbers = np.random.random(200)", 
      "print random_numbers", 
      "plot.hist(random_numbers)"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 142
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "import socket", 
      "print socket.gethostname()"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 143
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [], 
     "language": "python", 
     "outputs": []
    }
   ]
  }
 ]
}