{
 "metadata": {
  "name": "rna_pet_clustering"
 }, 
 "nbformat": 2, 
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "import subprocess", 
      "import shlex", 
      "", 
      "def index_bam(bamFile):", 
      "    SUB = 'index_bam'", 
      "    args = ['samtools','index',bamFile]", 
      "    return_code = subprocess.check_call(args) ", 
      "    if return_code == 0:", 
      "        print '[%s]: Created index for %s' % (SUB,os.path.basename(bamFile))"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 2
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "import pysam", 
      "import os", 
      "", 
      "", 
      "from subprocess import call", 
      "", 
      "def split_bam_to_sense_antisense(bamFile, force_run=False):", 
      "    ", 
      "    '''given an input bam file only with mate pair direction reads", 
      "       and color tag YC:Z:green (5 prime) or YC:Z:orange( 3 prime)", 
      "       the function will split it into two bam files", 
      "       1. reads capturing cDNA ends on the sense strand", 
      "       2. reads capturing cDNA ends on the antisense strand", 
      "    '''", 
      "    ", 
      "    SUB = 'split_bam_to_sense_antisense'", 
      "    ", 
      "    print '[%s]:opening bam %s for reading ' % (SUB,bamFile)", 
      "    ", 
      "    ", 
      "    bamFile_bai = bamFile + '.bai'", 
      "    if not os.path.exists(bamFile_bai):", 
      "        index_bam(bamFile)", 
      "    ", 
      "    ", 
      "    ", 
      "    input_bamFile_handle = pysam.Samfile(bamFile,'rb')", 
      "    base_dir = os.path.dirname(bamFile) or '.'", 
      "    file_prefix=os.path.basename(bamFile).replace('.bam','')", 
      "    ", 
      "    sense_bamFile     = base_dir+'/'+file_prefix+'_sense.bam'", 
      "    antisense_bamFile = base_dir+'/'+file_prefix+'_antisense.bam';", 
      "    ", 
      "    #if ( 1 and 1 ):", 
      "    if ( os.path.exists(sense_bamFile) and os.path.exists(antisense_bamFile) and ( force_run == 'False') ):", 
      "        print '[%s]: Expected output  %s  and  %s  already present \\n no processing required' % (SUB,os.path.basename(sense_bamFile_path), os.path.basename(antisense_bamFile_path))", 
      "        return (sense_bamFile_path, antisense_bamFile_path)", 
      "    ", 
      "    sense_bamFile_handle = pysam.Samfile(sense_bamFile,'wb',template=input_bamFile_handle)", 
      "    antisense_bamFile_handle = pysam.Samfile(antisense_bamFile,'wb',template=input_bamFile_handle)", 
      "    ", 
      "    for read in input_bamFile_handle.fetch():", 
      "        color  = read.tags[-1][1]  #last entry in the list of tuples and then extract the second element of tuple", 
      "        ", 
      "        if color not in ('orange','green'):", 
      "            print 'Warning : color %s is not orange/green' % color", 
      "            continue", 
      "        ", 
      "        if read.is_reverse:  #read maps in the opposite direction", 
      "            if color == 'green': ##read is coming from sense strand", 
      "                #print 'Found %s color' % color", 
      "                sense_bamFile_handle.write(read)", 
      "            elif color == 'orange':", 
      "                #print 'Found %s color' % color", 
      "                antisense_bamFile_handle.write(read)", 
      "    ", 
      "    #read maps in the same direction as reference 5 prime to 3 prime", 
      "        else:", 
      "            if color == 'orange':  ##read is coming from the sense strand", 
      "                sense_bamFile_handle.write(read)", 
      "        ", 
      "            elif color == 'green': ##read is coming from antisense strand", 
      "                antisense_bamFile_handle.write(read)", 
      "    print 'Created sense bam %s \\n Create antisense bam %s \\n\\n' % (os.path.basename(sense_bamFile),os.path.basename(antisense_bamFile))", 
      "    ", 
      "    ", 
      "    #print 'Ready to index %s and %s files' % (os.path.basename(sense_bamFile),os.path.basename(antisense_bamFile))", 
      "    ", 
      "    #for some reason this doesnt work inside this method", 
      "    # works fine outside this function", 
      "    #index_bam(sense_bamFile)", 
      "    #index_bam(antisense_bamFile_path)", 
      "    ", 
      "    return(sense_bamFile,antisense_bamFile)"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 3
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "def remove_low_coverage_reads(bamFile,coverage=2):", 
      "    ", 
      "    SUB ='remove_low_coverage_reads'", 
      "    ", 
      "    base_path = os.path.dirname(bamFile)", 
      "    bamFile_name = os.path.basename(bamFile)", 
      "    bamFile_handle = pysam.Samfile(bamFile,'rb')", 
      "    references = bamFile_handle.references", 
      "    out_bamFile = base_path + '/' + bamFile_name.replace('.bam','') + '_mincov_'+str(coverage)+'.bam'", 
      "    ", 
      "    ", 
      "    if os.path.exists(out_bamFile):", 
      "        print 'Expected output %s already present \\n no processed required' % (out_bamFile)", 
      "        return out_bamFile", 
      "    ", 
      "    ", 
      "    out_bamFile_handle = pysam.Samfile(out_bamFile,'wb',template=bamFile_handle)", 
      "    count_num_reads_low_coverage = 0;", 
      "    ", 
      "    ", 
      "    for chromosome in references:", 
      "        low_coverage_read_headers = {}", 
      "        ", 
      "        for pileupcol in bamFile_handle.pileup(chromosome):", 
      "           if pileupcol.n <= coverage and pileupcol.n > 0:", 
      "                for read in pileupcol.pileups:", 
      "                    low_coverage_read_headers[read.alignment.qname] = 1", 
      "        ", 
      "        #if len(low_coverage_read_headers.keys()) > 0:", 
      "            #print low_coverage_read_headers.keys()", 
      "            #print ", 
      "        ", 
      "        for read in bamFile_handle.fetch(chromosome):", 
      "            if ( low_coverage_read_headers.get(read.qname,'None') == 'None'):", 
      "                out_bamFile_handle.write(read)", 
      "            else:", 
      "                count_num_reads_low_coverage += 1", 
      "        ", 
      "        os.system(\"samtools index out_bamFile\")", 
      "    print 'Wrote file %s, found and skipped %d low coverage reads' % (os.path.basename(out_bamFile), count_num_reads_low_coverage)", 
      "    ", 
      "    #doesn't work inside the ", 
      "    #index_bam(out_bamFile)", 
      "    ", 
      "    return out_bamFile"
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 4
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "def plot_bamCoverage(bamFile,min_coverage=2):", 
      "    ", 
      "    SUB ='plot_bamCoverage'", 
      "    ", 
      "    base_path = os.path.dirname(bamFile)", 
      "    bamFile_name = os.path.basename(bamFile)", 
      "    bamFile_handle = pysam.Samfile(bamFile,'rb')", 
      "    references = bamFile_handle.references", 
      "    out_plot_name = base_path + '/' + bamFile_name.replace('.bam','') + '_mincov_'+str(min_coverage)+'.png'", 
      "", 
      "    if os.path.exists(out_plot_name):", 
      "        print 'Expected output %s already present \\n no processed required' % (out_plot_name)", 
      "        return out_plot_name", 
      "    ", 
      "    nuc_coverage = []", 
      "    for chromosome in references:", 
      "        for pileupcol in bamFile_handle.pileup(chromosome):", 
      "           if pileupcol.n > min_coverage:", 
      "                nuc_coverage.append(pileupcol.n)", 
      "    ", 
      "    #print 'Max : %d' %(max(coverage))", 
      "    plot(nuc_coverage,'-')", 
      "    ", 
      "", 
      "#plot_bamCoverage(antisense_bamFile)", 
      "                "
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 5
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "##clustering attempt", 
      "from scipy.spatial import distance", 
      "import numpy as np", 
      "from scipy.spatial import distance", 
      "from sklearn.cluster import DBSCAN", 
      "from sklearn import metrics", 
      "", 
      "", 
      "def cluster_points_DBSCAN(data_numpy_array):", 
      "    #eucledian distance calculation", 
      "    D = distance.pdist(data_numpy_array)", 
      "    S = distance.squareform(D)", 
      "    H = 1 - S/np.max(S)", 
      "    db = DBSCAN().fit(H, eps=0.95)", 
      "    ", 
      "    core_samples_indices = db.core_sample_indices_", 
      "    labels = db.labels_", 
      "    components = db.components_", 
      "    ", 
      "    n_clusters_ = len(set(labels)) - ( 1 if -1 in labels else 0)", 
      "    ", 
      "    print 'Num of clusters %d' % n_clusters_", 
      "", 
      "    #f = open('check.bed', 'w')", 
      "    ", 
      "    ", 
      "    ", 
      "    for k in set(labels):", 
      "        index_class_members = [ index[0] for index in np.argwhere(labels == k ) ]", 
      "        data = data_numpy_array[index_class_members]", 
      "        mean_start = mean(data[:,0])", 
      "        mean_end = mean(data[:,1]) ", 
      "        print '%d \\t %d : shape %s \\n' % (mean_start, mean_end, data.shape)", 
      "        print \"=======================\"", 
      "    ", 
      "    ", 
      "    "
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 6
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [
      "import numpy as np", 
      "", 
      "def cluster_local(data):", 
      "    ", 
      "    cluster_bucket = []", 
      "    last_start = 0", 
      "    last_end = 0", 
      "    temp_bin = [];", 
      "    ", 
      "    for start,end in data:", 
      "        print 'last_start: %d , last_end : %d , cstart : %d , cend : %d' % (last_start,last_end,start,end)", 
      "        ##error condition", 
      "        if start < last_start:  ", 
      "            print 'Error : current start %d < last_start %d ' % (start, last_start)", 
      "        ##start condition", 
      "        if last_start == 0 and last_end == 0:", 
      "            last_start = start", 
      "            last_end = end", 
      "            temp_bin.append((start,end))", 
      "            continue", 
      "            ", 
      "        #main condition", 
      "        if ( ", 
      "            ( (start == last_start) or ( start >= last_start+1)  ) ", 
      "            and ", 
      "            ( (end == last_end+1)  or (end == last_end-1)   or (end == last_end) ) ", 
      "            ):", 
      "            temp_bin.append((start,end))", 
      "            last_start = start", 
      "            last_end = end", 
      "            print 'Current temp bin values %s' % temp_bin", 
      "        else:", 
      "            cluster_bucket.append(temp_bin)", 
      "            print 'Flushing out the temp_bin'", 
      "            temp_bin = []", 
      "            last_start = start", 
      "            last_end = end", 
      "        ", 
      "            ", 
      "            ", 
      "    ", 
      "", 
      "        ", 
      "    ", 
      "    ", 
      "    ", 
      "    "
     ], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": 7
    }, 
    {
     "cell_type": "code", 
     "collapsed": false, 
     "input": [
      "import pysam", 
      "import sys", 
      "import collections", 
      "from time import sleep", 
      "", 
      "def cluster_pet_points_per_chromosome(bamFile):", 
      "    ", 
      "    dir = os.path.dirname(bamFile)", 
      "    ", 
      "    bamFile_handle = pysam.Samfile(bamFile,'rb')", 
      "    bamFile_handle_mate = pysam.Samfile(bamFile,'rb')  #for fetching the mate without disturbing the main file handle", 
      "    ", 
      "    references = bamFile_handle.references", 
      "    ", 
      "    count = 0;", 
      "    for chromosome in references:", 
      "        points = []", 
      "        cnt = collections.Counter()", 
      "        ", 
      "        for read in bamFile_handle.fetch(chromosome):", 
      "            ''' below is a very customized code block", 
      "                should only work on the mate paired reads for RNA-PET", 
      "                NOT TO BE GENERALIZED", 
      "            '''", 
      "            try:", 
      "                #read_mate = bamFile_handle_mate.mate(read)", 
      "                ", 
      "                if read.is_read1:", 
      "                    if read.is_reverse: #start of transcript is read.pos and end is read.mpos+length of read 2 ", 
      "                        #print 'Read 1 is rev ',read.isize", 
      "                        #print 'pos %d mpos %d' % (read.pos,read.mpos)", 
      "                        start = read.pos", 
      "                        end = read.pos + read.isize", 
      "                        #end = read_mate.pos + read_mate.alen", 
      "                        points.append((start,end))", 
      "                        cnt[(start,end)] +=1", 
      "                        out_str = '%d \\t %d\\n' % (start,end)", 
      "                        #fh.write(out_str)", 
      "                        #print 'start %d end %d' % (start,end)", 
      "                    else:", 
      "                        #print 'Read 1 is fow ',read.isize", 
      "                        #print 'pos %d mpos %d' % (read.pos,read.mpos)", 
      "                        #start = read_mate.pos", 
      "                        start = read.mpos", 
      "                        end = read.mpos + (read.isize * -1)", 
      "                        points.append((start,end))", 
      "                        cnt[(start,end)] +=1", 
      "                        out_str = '%d \\t %d\\n' % (start,end)", 
      "                        #fh.write(out_str)", 
      "                        #print 'start %d end %d' % (start,end)", 
      "            except ValueError:", 
      "                print 'No mate found for read %s' % read.qname", 
      "                continue", 
      "            ", 
      "", 
      "            ", 
      "                    ", 
      "        if (  len(points)  > 0):", 
      "            points_array = array(points)", 
      "            print 'Shape of numpy array',points_array.shape", 
      "            print 'Size of list points is %d' % sys.getsizeof(points)", 
      "            print 'Size of numpy array is %d' % sys.getsizeof(points_array) ## will just be a reference so not the actual size", 
      "            print 'Len of counter is %d' % len(cnt)", 
      "            ", 
      "            ", 
      "            #points_array.sort()", 
      "            #print type(cnt)", 
      "            #print cnt", 
      "            ", 
      "            if chromosome == 'chromosome_4':", 
      "                dir = os.path.dirname(bamFile) or '.'", 
      "                temp_file = dir + '/uniq_points_chr_'+chromosome+'.txt'", 
      "                fh = open(temp_file,'w')", 
      "                for start,end in cnt.iterkeys():", 
      "                    out = '%d \\t %d\\n' %(start,end)", 
      "                    fh.write(out)", 
      "            ", 
      "            ", 
      "            values = [ key for key in cnt.itervalues()]", 
      "            ", 
      "            pylab.plot(values)", 
      "            pylab.ylabel('#count')", 
      "            pylab.xlabel('uniq-start-end-pair')", 
      "            pylab.title(os.path.basename(bamFile))", 
      "            ", 
      "            fig_file = dir + '/sense.png'", 
      "            pylab.savefig(fig_file)", 
      "            ", 
      "            #dir(cnt)", 
      "            #print points_array", 
      "            #cluster_local(points_array)", 
      "            #cluster_points_DBSCAN(values)", 
      "            #print points_array'''", 
      "", 
      "            ", 
      "            ", 
      "", 
      "sample_bamFile1='/house/homedirs/a/apratap/playground/GIS_RNA_SEQ/pet_RNA/combined_analysis/Chlamy_V5/pet_clustering_test/test_mate_pair_sorted.bam'", 
      "sample_bamFile2='/house/homedirs/a/apratap/playground/GIS_RNA_SEQ/pet_RNA/combined_analysis/Chlamy_V5/pet_clustering_test/mate_pairs_chr4.bam'", 
      "full_bamFile='/house/homedirs/a/apratap/playground/GIS_RNA_SEQ/pet_RNA/combined_analysis/Chlamy_V5/bwa/mate_pairs_sorted.bam'", 
      "", 
      "# step 1", 
      "sense_bamFile, antisense_bamFile = split_bam_to_sense_antisense(sample_bamFile2, force_run=True)", 
      "index_bam(sense_bamFile)", 
      "index_bam(antisense_bamFile)", 
      "", 
      "#step 2", 
      "filtered_bamFile = remove_low_coverage_reads(sense_bamFile)", 
      "sleep(1)", 
      "index_bam(filtered_bamFile)", 
      "", 
      "plot_bamCoverage(filtered_bamFile)", 
      "", 
      "#cluster_pet_points_per_chromosome(filtered_bamFile)", 
      "", 
      "", 
      "", 
      "#cluster_pet_points_per_chromosome(antisense_bamFile)", 
      "#cluster_pet_points_per_chromosome(sample_bamFile2)"
     ], 
     "language": "python", 
     "outputs": [
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "[split_bam_to_sense_antisense]:opening bam /house/homedirs/a/apratap/playground/GIS_RNA_SEQ/pet_RNA/combined_analysis/Chlamy_V5/pet_clustering_test/mate_pairs_chr4.bam for reading ", 
        "Created sense bam mate_pairs_chr4_sense.bam ", 
        " Create antisense bam mate_pairs_chr4_antisense.bam ", 
        ""
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "[index_bam]: Created index for mate_pairs_chr4_sense.bam"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "[index_bam]: Created index for mate_pairs_chr4_antisense.bam"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "Expected output /house/homedirs/a/apratap/playground/GIS_RNA_SEQ/pet_RNA/combined_analysis/Chlamy_V5/pet_clustering_test/mate_pairs_chr4_sense_mincov_2.bam already present ", 
        " no processed required", 
        "[index_bam]: Created index for mate_pairs_chr4_sense_mincov_2.bam"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        "", 
        "Max : 8014"
       ]
      }, 
      {
       "output_type": "stream", 
       "stream": "stdout", 
       "text": [
        ""
       ]
      }
     ], 
     "prompt_number": 8
    }, 
    {
     "cell_type": "code", 
     "collapsed": true, 
     "input": [], 
     "language": "python", 
     "outputs": [], 
     "prompt_number": "&nbsp;"
    }
   ]
  }
 ]
}