{
 "metadata": {
  "name": "rna_spike_in_report_sciclone"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "from pandas import *\n",
      "import os, fnmatch\n",
      "import re\n",
      "import matplotlib.pyplot as plt\n",
      "import numpy as np\n",
      "import pandas\n",
      "from matplotlib.table import Table\n",
      "\n",
      "##turning off the html printing of tables\n",
      "set_printoptions(notebook_repr_html = True)\n",
      "set_printoptions(max_rows=500)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "well_rowIndex_to_rowName = { 1:'A', 2:'B', 3:'C', 4:'D', 5:'E', 6:'F', 7:'G', 8:'H'}\n",
      "well_rowName_to_rowIndex = {'A':1,'B':2,'C':3, 'D':4,'E':5, 'F':6, 'G':7, 'H':8}\n",
      "\n",
      "\n",
      "def index_num_to_wellName(index_num):\n",
      "    m=re.search('(IT0+)(.*)',index_num)\n",
      "    if m:\n",
      "        index_num = int(m.group(2))\n",
      "        if index_num > 96:  ##current plate can hav max of 96 wells\n",
      "            return 'NA'\n",
      "        quotient,remainder = divmod(index_num,8)\n",
      "        #print index_num, quotient, remainder\n",
      "        \n",
      "        ###quotient is X-axis (horizontal line on plate) 1-12\n",
      "        ###remainder : well name (vertical line on plate)  A-H\n",
      "        if remainder != 0:\n",
      "            colIndex = quotient + 1\n",
      "            rowName = well_rowIndex_to_rowName.get(remainder)\n",
      "            wellName='%s%d' % (rowName,colIndex)\n",
      "        else:\n",
      "            colIndex = quotient\n",
      "            rowName = well_rowIndex_to_rowName.get(8)\n",
      "            wellName = '%s%d' % (rowName,colIndex)\n",
      "\n",
      "    return wellName\n",
      "\n",
      "\n",
      "def is_well_spiked(wellName):\n",
      "    '''\n",
      "    given a well name, A1, B8 etc \n",
      "    the script returns whether the well is spiked or not (True/ False)\n",
      "    '''\n",
      "    match = re.search('^([A-H]{1})(\\d+?)$',wellName.strip())\n",
      "    if match:\n",
      "        row_name = match.group(1)\n",
      "        column = int(match.group(2))\n",
      "        #print 'row: %s, col : %s' % (row_name,column)\n",
      "        if column > 12:\n",
      "            print 'Columns number should be <= 12 : %s' % column\n",
      "            return 'NA'\n",
      "        \n",
      "        #converting the row and col to 0 based values\n",
      "        row = well_rowName_to_rowIndex[row_name] - 1\n",
      "        column = int(column) -1 \n",
      "        \n",
      "        #main logic in the next line\n",
      "        # 0 based system\n",
      "        #when row is even all the even columns in the row have spikes\n",
      "        #when row is odd all the odd columns in the row have spikes\n",
      "        spiked = [True, False]\n",
      "        index = [column % 2 , (column+1) % 2 ][ row % 2 ]\n",
      "        #print 'row: %s, col : %s, Spiked: %s' % (row,column,spiked[index])\n",
      "        return spiked[index]\n",
      "    else:\n",
      "        print 'No match found for well name %s' % wellName\n",
      "        return 'NA'\n",
      "\n",
      "\n",
      "def get_plate_index_info_df(refresh = False):\n",
      "    '''\n",
      "    TODO\n",
      "    '''\n",
      "    \n",
      "    JGI_indexes = '/global/homes/a/apratap/databases/plate_indexes/JGI_index_info.txt'\n",
      "    final_index_info_file = '/global/homes/a/apratap/databases/plate_indexes/plate_indexInfo.txt'\n",
      "    \n",
      "    if refresh:    \n",
      "        ##get the standard 96 well plate index map\n",
      "        df_index_well_map = pandas.read_csv(JGI_indexes,names =['index_num','index'], sep='\\t')\n",
      "        \n",
      "        ##adding the well location\n",
      "        df_index_well_map['well'] = df_index_well_map['index_num'].map(index_num_to_wellName)\n",
      "        df_index_well_map['row']= df_index_well_map['well'].map(lambda x: int(well_rowName_to_rowIndex.get(x[0])) )\n",
      "        df_index_well_map['col']= df_index_well_map['well'].map(lambda x: int(x[1:3]) )\n",
      "        df_index_well_map['Spiked']= df_index_well_map['well'].map(is_well_spiked)\n",
      "        df_index_well_map.to_csv(final_index_info_file,sep='\\t',index=False)\n",
      "    \n",
      "    return pandas.read_csv(final_index_info_file, sep='\\t',header=0)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df_index_well_map = get_plate_index_info_df(refresh=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "\"\"\"\n",
      "For a given dir/path ..recursively look for files in the naming format *.bam.readCounts\n",
      "and generate a list of stats for the each ERCC spike for each library\n",
      "\"\"\"\n",
      "\n",
      "def get_FilesList(path,pattern):\n",
      "    \"\"\"\n",
      "    traverse a dir recursively and find files with full paths\n",
      "    that match the given pattern\n",
      "    \"\"\"\n",
      "    SUB='get_FilesList'\n",
      "    found_files = []\n",
      "    for root,dirs,files in os.walk(path):\n",
      "        for basename in files:\n",
      "            if fnmatch.fnmatch(basename,pattern):\n",
      "                filename = str(os.path.join(root,basename))\n",
      "                filename.strip()\n",
      "                found_files.append(filename.strip())\n",
      "    print '[%s]: Found %d files at %s' % ( SUB,len(found_files),path)\n",
      "    return found_files\n",
      "                \n",
      "    \n",
      "def create_data_array(file_list):\n",
      "    data = []\n",
      "    #out_stats_file = path+'spike_in_stats.txt'\n",
      "    #out_fh = open(out_stats_file,'w')\n",
      "    \n",
      "    for file_name in file_list:\n",
      "        #print file_name\n",
      "        with open(file_name, 'r') as fh:\n",
      "            file_blob = fh.readlines();\n",
      "            file_basename = os.path.basename(file_name)\n",
      "            file_prefix = file_basename.replace('.bam.readCounts','')\n",
      "            #file_blob = file_blob[2:-3]\n",
      "            \n",
      "            for line in file_blob:\n",
      "                each_list = line.split('\\t')\n",
      "                spike_seq = each_list[0].strip()\n",
      "                read_count = int(each_list[1].strip())\n",
      "                percentage = float(each_list[2].replace('%',''))\n",
      "                \n",
      "                if spike_seq == '*':\n",
      "                    read_count = int(0)\n",
      "                    percentage = float(0.0)\n",
      "                #finding the index from the fastq name\n",
      "                m = re.search('(.+?)\\..+?([AGCT]{6,})',file_prefix)\n",
      "                if m:\n",
      "                    library_name = m.group(1)\n",
      "                    index_in_fastq = m.group(2)\n",
      "                    #print index_in_fastq\n",
      "                    temp_tuple=(library_name,file_prefix,spike_seq,read_count,percentage,index_in_fastq)\n",
      "                    #print temp_tuple\n",
      "                    data.append(temp_tuple)\n",
      "    print 'Read %d files' % (len(file_list))\n",
      "    return data\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 23
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def sciclone_plate_plot(data,method,plot_title=None,fmt='{:.4f}', bkg_colors=['yellow','white']):\n",
      "    \"\"\"\n",
      "    given a numpy array generated from create_data_array{function}\n",
      "    create a plate level plot\n",
      "    \"\"\"\n",
      "    \n",
      "    plt.close()\n",
      "    \n",
      "    fig=plt.figure(figsize=(14,12))\n",
      "\n",
      "    ax= plt.subplot(2,1,1,)\n",
      "    plt.title(plot_title,verticalalignment='center')\n",
      "    #subplot(2,1,1)\n",
      "    #fig(num=None,figsize=(8,6),dpi=100,facecolor='w',edgecolor='k')\n",
      "    ax.set_axis_off()\n",
      "    #ax.set_figsize=(12,10)\n",
      "    tb = Table(ax,bbox=[0,0,1,1])\n",
      "    \n",
      "    nrows, ncols = data.shape\n",
      "    width, height = 1.0/ncols, 1.0/nrows\n",
      "    \n",
      "    \n",
      "    #calculating the sd and mean for spikewells\n",
      "    values_all_spiked_wells =[]\n",
      "    #based on 0 index\n",
      "    #even rows have even spiked in cols\n",
      "    values_spiked_wells_even_rows = data[::2,::2].flatten() \n",
      "    # odd rows have odd spiked in cols\n",
      "    values_spiked_wells_odd_rows = data[1::2,1::2].flatten()\n",
      "    #print values_spiked_wells_odd_rows\n",
      "    values_all_spiked_wells = np.concatenate( (values_spiked_wells_even_rows, values_spiked_wells_odd_rows), axis=0)\n",
      "    \n",
      "    #removing the NaN's from the array\n",
      "    not_NaN_values_all_spiked_wells = values_all_spiked_wells[ ~np.isnan(values_all_spiked_wells) ]\n",
      "    \n",
      "    \n",
      "    if method == 'read_counts':\n",
      "        not_NaN_values_all_spiked_wells = not_NaN_values_all_spiked_wells / 1000\n",
      "    \n",
      "    #calculating the sd and mean for spikewells\n",
      "    values_all_non_spiked_wells = []\n",
      "    #even rows have odd non-spiked wells\n",
      "    vales_non_spiked_wells_even_rows = data[::2,1::2].flatten()\n",
      "    #odd rows have even non-spiked wells\n",
      "    values_non_spiked_wells_odd_rows = data[1::2,::2].flatten()\n",
      "    values_all_non_spiked_wells = np.concatenate( ( vales_non_spiked_wells_even_rows, values_non_spiked_wells_odd_rows))\n",
      "    \n",
      "    \n",
      "    #removing the NaN's from the array\n",
      "    not_NaN_values_all_non_spiked_wells = values_all_non_spiked_wells[ ~np.isnan(values_all_non_spiked_wells) ]\n",
      "    \n",
      "\n",
      "    mean_all_spiked_wells = np.mean(not_NaN_values_all_spiked_wells)\n",
      "    sd_all_spiked_wells = np.std(not_NaN_values_all_spiked_wells)\n",
      "    \n",
      "    mean_all_non_spiked_wells = np.mean(not_NaN_values_all_non_spiked_wells)\n",
      "    sd_all_non_spiked_wells = np.std(not_NaN_values_all_non_spiked_wells)\n",
      "    \n",
      "    \n",
      "    print 'Spiked Wells %d  : %f +/- %f ' % (len(not_NaN_values_all_spiked_wells),mean_all_spiked_wells,sd_all_spiked_wells)\n",
      "    print 'Non Spiked Wells %d  : %f +/- %f ' % (len(not_NaN_values_all_spiked_wells),mean_all_non_spiked_wells,sd_all_non_spiked_wells)\n",
      "\n",
      "    \n",
      "    for (i,j), val in np.ndenumerate(data):\n",
      "        \n",
      "        ## if  i is even we want even cols in the row to be colored\n",
      "        ## if i is odd we want odd cols in the row to be colored\n",
      "        idx = [ j %2, (j+1) % 2][ i % 2]\n",
      "        color = bkg_colors[idx]\n",
      "\n",
      "        if np.isnan(val):\n",
      "            color = 'grey'\n",
      "        \n",
      "        ##for read percent\n",
      "        ##change the color fo the cell if it is off from the standard values\n",
      "        if method == 'read_percent':\n",
      "            if color == 'yellow':\n",
      "                if val > 3.7:\n",
      "                    color='#377EB8'\n",
      "                if val <= 0.4:\n",
      "                    color='#377EB8'\n",
      "            if color == 'white':\n",
      "                if val >= 0.005:\n",
      "                    color='#E41A1C'\n",
      "        \n",
      "        \n",
      "        ##for read counts        \n",
      "        ##change the color of the cell if it is off from the standard values\n",
      "        if method == 'read_counts':\n",
      "            if color == 'yellow':\n",
      "                if val > 600000:\n",
      "                    color='#377EB8'\n",
      "                if val <= 50000:\n",
      "                    color='#377EB8'\n",
      "            if color == 'white':\n",
      "                if val >= 700:\n",
      "                    color='#E41A1C'\n",
      "                \n",
      "        #\n",
      "        if method == 'read_counts':\n",
      "            fmt='{:.0f}'\n",
      "            \n",
      "        tb.add_cell(i,j,width,height,text=fmt.format(val),loc='center',facecolor=color)\n",
      "        \n",
      "    #col labels\n",
      "    for j,label in enumerate(   range(  1,(ncols+1)  )  ):\n",
      "        #print j, label\n",
      "        #tb.add_cell(-1,j, width, height/2, text=label,loc='center',edgecolor=\"none\",facecolor=\"none\")\n",
      "        tb.add_cell(nrows,j, width, height/2, text=label,loc='center',edgecolor=\"none\",facecolor=\"none\")\n",
      "        \n",
      "    #row labels\n",
      "    for i,label in enumerate(['A','B','C','D','E','F','G','H']):\n",
      "        #print j, label\n",
      "        tb.add_cell(i,-1, width/2, height, text=label,loc='center',edgecolor=\"none\",facecolor=\"none\")\n",
      "    \n",
      "    \n",
      "    plt.text(0,-.02,'Yellow : spiked cells  , *Blue*: outliers')\n",
      "    plt.text(0,-.06,'White  : non-spiked cells , *Red* : outliers')\n",
      "    \n",
      "    spiked_wells_summary =     'Mean %2.4f +/- %2.4f' % (mean_all_spiked_wells,sd_all_spiked_wells)\n",
      "    non_spiked_wells_summary = 'Mean  %2.4f +/- %2.4f' % (mean_all_non_spiked_wells,sd_all_non_spiked_wells)\n",
      "    \n",
      "    ax.add_table(tb)\n",
      "    \n",
      "    plt.subplot(2,2,3)\n",
      "    plt.title('histogram:reads in spiked wells')\n",
      "    if method == 'read_counts':\n",
      "        plt.xlabel('Read Counts (**in thousands**)')\n",
      "    else:\n",
      "        plt.xlabel('Read Percent')\n",
      "    plt.hist(not_NaN_values_all_spiked_wells, color='darkred',edgecolor='w',bins=25)\n",
      "    plt.text(min(not_NaN_values_all_spiked_wells),-1, spiked_wells_summary)\n",
      "    plt.grid()\n",
      "    \n",
      "    \n",
      "    plt.subplot(2,2,4)\n",
      "    plt.title('histogram:reads in non-spiked wells')\n",
      "    if method == 'read_counts':\n",
      "        plt.xlabel('Read Counts')\n",
      "    else:\n",
      "        plt.xlabel('Read Percent')\n",
      "    plt.hist(not_NaN_values_all_non_spiked_wells, color='k',edgecolor='w',bins=25)\n",
      "    plt.text(min(not_NaN_values_all_non_spiked_wells),-3, non_spiked_wells_summary)\n",
      "    plt.grid()\n",
      "\n",
      "    return (not_NaN_values_all_spiked_wells,not_NaN_values_all_non_spiked_wells,fig)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 77
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\n",
      "#def looper(path,pattern):\n",
      "def looper(path,pattern):\n",
      "    \n",
      "    percentreads_all_plates_spiked_wells = []\n",
      "    percentreads_all_plates_non_spiked_wells = []\n",
      "    \n",
      "    readcounts_all_plates_spiked_wells = []\n",
      "    readcounts_all_plates_non_spiked_wells = []\n",
      "    \n",
      "    \n",
      "    for i,each_dir in enumerate(path):\n",
      "        print '%d: %s' % (i+1,each_dir)\n",
      "        file_list = get_FilesList(each_dir,pattern)                \n",
      "        data  = create_data_array(file_list)\n",
      "\n",
      "        #create a pandas data.frame\n",
      "        df = DataFrame(data,columns=['lib','fastq','spike','rcount','percent','index'])\n",
      "        ##merging the well info with the spike-in info\n",
      "        fdf = merge(df,df_index_well_map,on='index', how='outer')\n",
      "        \n",
      "        fdf.to_csv('1.csv')\n",
      "        \n",
      "        ###PLOT 1\n",
      "        ##create a pivot table for read percent per well\n",
      "        plate_spike_percent = pivot_table(fdf,values='percent',rows=['row'],cols=['col'], aggfunc=np.sum, fill_value='NaN')\n",
      "        #print plate_spike_percent\n",
      "        #converting to numpy array\n",
      "        plate_spike_percent = np.array(plate_spike_percent)\n",
      "        ##generating the plate plot\n",
      "        (not_NaN_values_all_spiked_wells,not_NaN_values_all_non_spiked_wells,figure) = sciclone_plate_plot(plate_spike_percent,method='read_percent',plot_title=each_dir)\n",
      "        figure = 'sciclone_stats_read_percent.png'\n",
      "        figure_path = each_dir + '/' + figure\n",
      "        plt.savefig(figure_path)\n",
      "        ##adding all the non NaN values to list\n",
      "        percentreads_all_plates_spiked_wells.extend( not_NaN_values_all_spiked_wells )\n",
      "        percentreads_all_plates_non_spiked_wells.extend(not_NaN_values_all_non_spiked_wells)\n",
      "        \n",
      "        \n",
      "        \n",
      "        ##PLOT 2\n",
      "        ##create a pivot table read **count** per well\n",
      "        plate_spike_readcount = pivot_table(fdf,values='rcount',rows=['row'],cols=['col'], aggfunc=np.sum, fill_value='NaN')\n",
      "        #converting to numpy array\n",
      "        plate_spike_readcount = np.array(plate_spike_readcount)\n",
      "        ##generating the plate plot\n",
      "        (not_NaN_values_all_spiked_wells,not_NaN_values_all_non_spiked_wells,figure) = sciclone_plate_plot(plate_spike_readcount,method='read_counts',plot_title=each_dir)\n",
      "        figure = 'sciclone_stats_read_counts.png'\n",
      "        figure_path = each_dir + '/' + figure\n",
      "        plt.savefig(figure_path)\n",
      "        \n",
      "        \n",
      "        ##adding all the non NaN values to list\n",
      "        readcounts_all_plates_spiked_wells.extend( not_NaN_values_all_spiked_wells )\n",
      "        readcounts_all_plates_non_spiked_wells.extend( not_NaN_values_all_non_spiked_wells )\n",
      "\n",
      "        \n",
      "    return (percentreads_all_plates_spiked_wells,percentreads_all_plates_non_spiked_wells,\n",
      "            readcounts_all_plates_spiked_wells, readcounts_all_plates_non_spiked_wells);\n",
      "        \n",
      "\n",
      "pattern = '*.bam.readCounts'\n",
      "\n",
      "'''\n",
      "path = [\n",
      "        '/house/groupdirs/genetic_analysis/rna/projects/Zea_mays_404736/',\n",
      "        '/house/groupdirs/genetic_analysis/rna/projects/Zea_mays_405317/',\n",
      "        '/house/groupdirs/genetic_analysis/rna/projects/Zea_mays_405318/',\n",
      "        '/house/groupdirs/genetic_analysis/rna/projects/Zea_mays_405319/',\n",
      "        '/house/groupdirs/genetic_analysis/rna/projects/Zea_mays_405320/',\n",
      "        '/house/groupdirs/genetic_analysis/rna/projects/Zea_mays_407386/',\n",
      "       ]\n",
      "'''\n",
      "#path = ['/global/projectb/scratch/apratap/PS-4903/plate_8','/global/projectb/scratch/apratap/PS-4903/plate_9']\n",
      "\n",
      "path = ['/global/projectb/scratch/apratap/PS-5049']\n",
      "\n",
      "(percentreads_all_plates_spiked_wells,percentreads_all_plates_non_spiked_wells,\n",
      " readcounts_all_plates_spiked_wells, readcounts_all_plates_non_spiked_wells )        =   looper(path,pattern)\n",
      "\n",
      "\n",
      "\n",
      "#clf()\n",
      "#close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1: /global/projectb/scratch/apratap/PS-5049\n",
        "[get_FilesList]: Found 91 files at /global/projectb/scratch/apratap/PS-5049"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Read 91 files\n",
        "col      1       2       3       4       5       6       7       8       9       10      11      12\n",
        "row                                                                                                \n",
        "1       NaN  0.0016  0.3707  0.0007  0.6683  0.0004  0.0000  0.0000  3.3307  0.0000  2.8276     NaN\n",
        "2    0.0018  0.6610  0.0002  0.6591  0.0004  1.1427  0.0006  0.0000  0.0058  0.0001  0.0027  1.4227\n",
        "3    3.2981  0.0005  1.2706  0.0004  1.9294  0.0010  0.0000  0.0006  0.0003  0.0002  0.7263  0.0003\n",
        "4    0.0043  0.6980  0.0009  0.3731  0.0013  0.6297  0.0001  1.0028  0.0002  1.1417  0.0004  0.0000\n",
        "5    1.6228  0.0013  0.6958  0.0013  1.4288  0.0000  1.0518  0.0004  1.8444  0.0003  1.4741  0.0000\n",
        "6    0.0014  1.0008  0.0003  0.9301  0.0008  0.0000  0.0013  1.1757  0.0004  0.1626  0.0006  0.0000\n",
        "7    0.6119  0.0013  1.1855  0.0014  0.0000  0.0005  0.0000  0.0007  0.5916  0.0034  0.9852     NaN\n",
        "8       NaN  1.1785  0.0006  0.5322  0.0008  0.9057  0.0000  0.9623  0.0000  0.6202  0.0007     NaN"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Spiked Wells 46  : 0.893759 +/- 0.797034 "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Non Spiked Wells 46  : 0.000931 +/- 0.001135 \n",
        "Spiked Wells 46  : 455.580696 +/- 399.793413 "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "Non Spiked Wells 46  : 481.600000 +/- 462.554184 \n"
       ]
      }
     ],
     "prompt_number": 78
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "############\n",
      "#Summary for strand specificity\n",
      "############\n",
      "\n",
      "\n",
      "#1. split all the sorted mapped bams in sense and antisense\n",
      "# currently there is one method which does it for first strand RNA-Seq reads (dUTP)\n",
      "\n",
      "#2. read all the log files\n",
      "\n",
      "log_files_pattern = '*strand_split.log'\n",
      "#path = '/global/projectb/scratch/apratap/strand_specificity_check/ERCC_mapping'\n",
      "path = '/global/projectb/scratch/apratap/PS-5049'\n",
      "files_list = get_FilesList(path,log_files_pattern)\n",
      "\n",
      "data = []\n",
      "for file in files_list:\n",
      "    #cutsom replacement to get the file name\n",
      "    prefix = os.path.basename(file).replace('_sorted_strand_split.log','')\n",
      "    (lib,run_id,lane,basecall_id,index) = prefix.split('.')\n",
      "        \n",
      "    #counters\n",
      "    num_raw_reads = 0\n",
      "    num_sense_reads = 0\n",
      "    num_antisense_reads = 0\n",
      "    mapped_reads = 0\n",
      "    unmapped_reads = 0\n",
      "    \n",
      "    for line in open(file,'r'):\n",
      "        if line.strip().find('reads input bam') is not -1:\n",
      "            num_raw_reads = line.split(':')[-1].strip()\n",
      "        elif line.strip().find('reads sense bam') is not -1:\n",
      "            num_sense_reads = line.split(':')[-1].strip()\n",
      "        elif line.strip().find('reads antisense bam') is not -1:\n",
      "            num_antisense_reads = line.split(':')[-1].strip()\n",
      "        elif line.strip().find('mapped reads') is not -1:\n",
      "            mapped_reads = line.split(':')[-1].strip()\n",
      "    data.append((lib,prefix,index,int(num_raw_reads),int(num_sense_reads),int(num_antisense_reads),int(mapped_reads)))\n",
      "\n",
      "    \n",
      "#creating the pandas DF\n",
      "df = pandas.DataFrame.from_records(data,columns=('lib','fastq','index','#raw','#sense','#antisense','#mapped'))\n",
      "\n",
      "#join df with the index info info to\n",
      "df = df.merge(df_index_well_map,on='index')\n",
      "#subset the dataframe to keep only spiked indexes\n",
      "df = df[df.Spiked == True ]\n",
      "\n",
      "#calculate the %reads map\n",
      "df['%mapped'] = df.apply(lambda x: round((float(x['#mapped'])/float(x['#raw']) )*100,3),axis=1)\n",
      "\n",
      "#calculate the %reads map that are *SENSE*\n",
      "df['%mapped_sense'] = df[ df['#mapped'] > 10000 ].apply(lambda x: round((float(x['#sense'])/float(x['#mapped']) )*100,3),axis=1)\n",
      "\n",
      "#calculate the %reads map that are  *ANTISENSE*\n",
      "df['%mapped_antisense'] = df[ df['#mapped'] > 10000 ].apply(lambda x: round((float(x['#antisense'])/float(x['#mapped']) )*100,3),axis=1)\n",
      "\n",
      "\n",
      "print df.to_string()\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "[get_FilesList]: Found 91 files at /global/projectb/scratch/apratap/PS-5049\n",
        "     lib                     fastq   index      #raw   #sense  #antisense  #mapped index_num well  row  col Spiked  %mapped  %mapped_sense  %mapped_antisense\n",
        "0   HCAH  HCAH.6441.4.44347.AAACAT  AAACAT  63315946       25           0       25     IT049   A7    1    7   True    0.000            NaN                NaN\n",
        "2   HCBH  HCBH.6441.5.44348.AAATGC  AAATGC  63692058       19           4       23     IT051   C7    3    7   True    0.000            NaN                NaN\n",
        "4   HBXN  HBXN.6441.5.44348.AACTTG  AACTTG  52388586   448714       94577   543291     IT053   E7    5    7   True    1.037         82.592             17.408\n",
        "6   HOSW  HOSW.6441.6.44349.AAGCGA  AAGCGA  29270726        0           0        0     IT055   G7    7    7   True    0.000            NaN                NaN\n",
        "9   HBYH  HBYH.6441.6.44349.ACAAAC  ACAAAC  60599152        2           0        2     IT058   B8    2    8   True    0.000            NaN                NaN\n",
        "10  HBZS  HBZS.6440.1.44331.ACAGTG  ACAGTG  55914208   707330      185644   892974     IT005   E1    5    1   True    1.597         79.211             20.789\n",
        "12  HBYT  HBYT.6441.7.44350.ACCCAG  ACCCAG  31841644   268459       45664   314123     IT060   D8    4    8   True    0.987         85.463             14.537\n",
        "14  HBWT  HBWT.6441.7.44350.ACGATA  ACGATA  51811664   501362       99283   600645     IT062   F8    6    8   True    1.159         83.471             16.529\n",
        "17  HBYY  HBYY.6448.1.44628.AGAAGA  AGAAGA  47911928   408376       46379   454755     IT064   H8    8    8   True    0.949         89.801             10.199\n",
        "18  HCBB  HCBB.6448.1.44628.AGATAG  AGATAG  53229922  1467005      277257  1744262     IT065   A9    1    9   True    3.277         84.105             15.895\n",
        "20  HBZH  HBZH.6448.2.44629.AGCGCT  AGCGCT  54790872      113          15      128     IT067   C9    3    9   True    0.000            NaN                NaN\n",
        "23  HOSP  HOSP.6449.7.44642.AGTTCC  AGTTCC  68485414   537490      136255   673745     IT014   F2    6    2   True    0.984         79.776             20.224\n",
        "24  HBXP  HBXP.6448.2.44629.ATACGG  ATACGG  40550738   619143      116808   735951     IT069   E9    5    9   True    1.815         84.128             15.872\n",
        "26  GTZP  GTZP.6448.3.44630.ATCTAT  ATCTAT  65307542   342426       38372   380798     IT071   G9    7    9   True    0.583         89.923             10.077\n",
        "27  HCAC  HCAC.6440.4.44334.ATGAGC  ATGAGC  82180102   466722       66997   533719     IT026   B4    2    4   True    0.649         87.447             12.553\n",
        "30  HCAG  HCAG.6440.5.44335.CAAAAG  CAAAAG  73483852   233252       36867   270119     IT028   D4    4    4   True    0.368         86.352             13.648\n",
        "32  HOST  HOST.6440.5.44335.CACCGG  CACCGG  68255852   522719      103092   625811     IT030   F4    6    4   True    0.917         83.527             16.473\n",
        "34  HBWG  HBWG.6440.6.44336.CACTCA  CACTCA  45821998   206799       33541   240340     IT032   H4    8    4   True    0.525         86.044             13.956\n",
        "35  HBXG  HBXG.6449.6.44641.CAGATC  CAGATC  47947896   245024       43845   288869     IT007   G1    7    1   True    0.602         84.822             15.178\n",
        "36  HBZB  HBZB.6440.6.44336.CAGGCG  CAGGCG  51333012   292227       45608   337835     IT033   A5    1    5   True    0.658         86.500             13.500\n",
        "38  HBYB  HBYB.6441.1.44344.CATTTT  CATTTT  67495142  1100968      184808  1285776     IT035   C5    3    5   True    1.905         85.627             14.373\n",
        "42  HBYO  HBYO.6448.3.44630.CCGCAA  CCGCAA  65067134       46           3       49     IT074  B10    2   10   True    0.000            NaN                NaN\n",
        "43  HBWN  HBWN.6440.2.44332.CCGTCC  CCGTCC  80632048   794811      141216   936027     IT016   H2    8    2   True    1.161         84.913             15.087\n",
        "45  HBXW  HBXW.6448.4.44631.CGAGAA  CGAGAA  21849204   207083       38190   245273     IT076  D10    4   10   True    1.123         84.430             15.570\n",
        "47  HBZX  HBZX.6441.1.44344.CGGAAT  CGGAAT  62846880   747126      138925   886051     IT037   E5    5    5   True    1.410         84.321             15.679\n",
        "50  HBWC  HBWC.6441.2.44345.CTATAC  CTATAC  60245618       48           6       54     IT039   G5    7    5   True    0.000            NaN                NaN\n",
        "53  HCAP  HCAP.6449.7.44642.CTTGTA  CTTGTA  64831736   387795       58202   445997     IT012   D2    4    2   True    0.688         86.950             13.050\n",
        "54  HBWP  HBWP.6448.4.44631.GAAACC  GAAACC  16962132    24873        2242    27115     IT078  F10    6   10   True    0.160         91.732              8.268\n",
        "57  HCAZ  HCAZ.6440.7.44337.GACGGA  GACGGA  39665700   222518       20227   242745     IT080  H10    8   10   True    0.612         91.667              8.333\n",
        "58  HBWY  HBWY.6440.4.44334.GAGTGG  GAGTGG  57611210   532052      139087   671139     IT023   G3    7    3   True    1.165         79.276             20.724\n",
        "59  HCBN  HCBN.6440.7.44337.GATATA  GATATA  48336238  1159929      184766  1344695     IT081  A11    1   11   True    2.782         86.260             13.740\n",
        "62  HBZO  HBZO.6448.5.44632.GCAAGG  GCAAGG  28267026   172085       29896   201981     IT083  C11    3   11   True    0.715         85.199             14.801\n",
        "65  HBXB  HBXB.6448.5.44632.GCCTTA  GCCTTA  41704056   523303       81658   604961     IT085  E11    5   11   True    1.451         86.502             13.498\n",
        "67  HBWB  HBWB.6448.6.44633.GGCACA  GGCACA  34058350   290168       39976   330144     IT087  G11    7   11   True    0.969         87.891             12.109\n",
        "71  HBZZ  HBZZ.6440.2.44332.GTAGAG  GTAGAG  19255490    59330       10877    70207     IT017   A3    1    3   True    0.365         84.507             15.493\n",
        "73  HBXZ  HBXZ.6440.3.44333.GTGAAA  GTGAAA  51076420   529462      109782   639244     IT019   C3    3    3   True    1.252         82.826             17.174\n",
        "75  HBZU  HBZU.6440.3.44333.GTTTCG  GTTTCG  58572266   341600       59622   401222     IT021   E3    5    3   True    0.685         85.140             14.860\n",
        "76  HBYA  HBYA.6441.2.44345.TAATCG  TAATCG  61504952   589011      103757   692768     IT042   B6    2    6   True    1.126         85.023             14.977\n",
        "78  HCAX  HCAX.6449.6.44641.TAGCTT  TAGCTT  65128330   365747       58333   424080     IT010   B2    2    2   True    0.651         86.245             13.755\n",
        "79  HBZW  HBZW.6441.3.44346.TATAAT  TATAAT  71166138   381778       60526   442304     IT044   D6    4    6   True    0.622         86.316             13.684\n",
        "81  HBWW  HBWW.6441.3.44346.TCCCGA  TCCCGA  82291256       59           2       61     IT046   F6    6    6   True    0.000            NaN                NaN\n",
        "83  HBYW  HBYW.6441.4.44347.TCGGCA  TCGGCA  54758052   442703       47139   489842     IT048   H6    8    6   True    0.895         90.377              9.623\n",
        "84  HCAT  HCAT.6448.6.44633.TGAATG  TGAATG  45886680   551332       90486   641818     IT090  B12    2   12   True    1.399         85.902             14.098\n",
        "87  HBXU  HBXU.6448.7.44634.TGCTGG  TGCTGG  71439978        0           0        0     IT092  D12    4   12   True    0.000            NaN                NaN\n",
        "89  HBYG  HBYG.6440.1.44331.TTAGGC  TTAGGC  29991438   846624      128272   974896     IT003   C1    3    1   True    3.251         86.842             13.158\n",
        "90  HBWO  HBWO.6448.7.44634.TTCGAA  TTCGAA  48344168        0           0        0     IT094  F12    6   12   True    0.000            NaN                NaN"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 71
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ax= df.set_index('lib').sort_index()[['%mapped_sense','%mapped_antisense']].plot(kind='bar',stacked=True,\n",
      "                                                                                color=['#E5F5F9','darkred'],\n",
      "                                                                                yticks=(range(0,101,10))\n",
      "                                                                                )\n",
      "ax.tick_params(axis='x',labelsize=6)\n",
      "ax.set_ylabel('read mapping %')\n",
      "#ax.grid()\n",
      "plt.tight_layout()\n",
      "plt.savefig(path+'/strand_specificity.png',dpi=150)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 72
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p.close()\n",
      "\n",
      "\n",
      "####plotting the read percentage in all the spike in plates\n",
      "\n",
      "figure(figsize=(12,10))\n",
      "subplot(2,1,1)\n",
      "title('Histogram : %reads All Spiked Wells')\n",
      "hist(percentreads_all_plates_spiked_wells,bins=100,color=\"darkred\",edgecolor='white')\n",
      "grid()\n",
      "xlabel('Read Percentage -->')\n",
      "\n",
      "\n",
      "subplot(2,1,2)\n",
      "title('Histogram : %reads All Non Spiked Wells')\n",
      "hist(percentreads_all_plates_non_spiked_wells,bins=100,color=\"darkgreen\",edgecolor='white')\n",
      "grid()\n",
      "xlabel('Read Percentage -->')\n",
      "\n",
      "\n",
      "\n",
      "####plotting the read Counts in all the spike in plates\n",
      "figure(figsize=(12,10))\n",
      "subplot(2,1,1)\n",
      "title('Histogram : read Counts All Spiked Wells')\n",
      "hist(readcounts_all_plates_spiked_wells,bins=100,color=\"darkred\",edgecolor='white')\n",
      "grid()\n",
      "xlabel('Read Counts (in thousands) -->')\n",
      "\n",
      "\n",
      "subplot(2,1,2)\n",
      "title('Histogram : read Counts All Non Spiked Wells')\n",
      "hist(readcounts_all_plates_non_spiked_wells,bins=100,color=\"darkgreen\",edgecolor='white')\n",
      "grid()\n",
      "xlabel('Read Counts -->')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "x = lambda x: int(x) if x and x.isdigit() else None\n",
      "x('d')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 38
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "##removing all the spike sequences \n",
      "fdf_rcount_gt_10 = fdf[fdf.rcount > 10]\n",
      "spikes_grp_by_well = fdf_rcount_gt_10.groupby(['well'])['spike']\n",
      "\n",
      "spikes_grp_by_well.describe()['count'] "
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": true,
     "input": [
      "fdf.to_csv('/house/homedirs/a/apratap/playground/test/spike_in_stats.txt', sep=',',header=True,index=True,index_label='num',na_rep='NaN')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}